{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "tar -xvf /home/yiheng/data/Wagga_run2/basecalled_data/Wagga_run2_albacore202.tar.gz\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "['barcode07', 'barcode08', 'barcode09', 'barcode10', 'barcode11']"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# coding: utf-8\n",
    "\n",
    "# In[16]:\n",
    "\n",
    "import argparse\n",
    "import os\n",
    "import subprocess\n",
    "import datetime\n",
    "\n",
    "\n",
    "# In[17]:\n",
    "\n",
    "sep_barcodes = '07,08,09,10,11'\n",
    "indir = '/home/yiheng/data/Wagga_run2'\n",
    "\n",
    "# This will be a dummy notebook for generating the analysis scripts for Yihengs WGS nanopore data analysis\n",
    "\n",
    "# In[51]:\n",
    "\n",
    "os.getcwd()\n",
    "\n",
    "\n",
    "# In[52]:\n",
    "\n",
    "#stuff you get from argparse\n",
    "#print (\"indir = {}\".format(d))\n",
    "#print (\"sep_barcodes = {}\".format(s))\n",
    "#print (\"threads = {}\".format(t))\n",
    "#print (\"ram = {}\".format(r))\n",
    "#print (\"BLAST_DB_1 = {}\".format(db1))\n",
    "#print (\"BLAST_DB_2 = {}\".format(db2))\n",
    "#BLAST_DB_2 = 'nt'\n",
    "\n",
    "\n",
    "\n",
    "# In[53]:\n",
    "\n",
    "#get the date\n",
    "now = datetime.datetime.now()\n",
    "date = now.strftime(\"%Y%m%d\")\n",
    "\n",
    "\n",
    "# In[54]:\n",
    "\n",
    "#define the path for control DNA sequence for nanolyse\n",
    "CDS_PATH = '/home/yiheng/bio/ncbi/db/DNA_CS.fasta'\n",
    "\n",
    "\n",
    "# In[55]:\n",
    "\n",
    "date\n",
    "\n",
    "\n",
    "# In[56]:\n",
    "\n",
    "if not os.path.exists(indir):\n",
    "    print('INDIR %s does not exsist. Exit!' % indir)\n",
    "    #exit()\n",
    "\n",
    "\n",
    "# In[57]:\n",
    "\n",
    "os.listdir(indir)\n",
    "\n",
    "\n",
    "# In[58]:\n",
    "\n",
    "os.path.abspath(indir)\n",
    "\n",
    "\n",
    "# In[59]:\n",
    "\n",
    "SCRIPT_FOLDER = os.path.join(os.path.abspath(indir), 'scripts')\n",
    "WORKSPACE = os.path.join(os.path.abspath(indir), 'workspace')\n",
    "TRACKING = os.path.join(os.path.abspath(indir), 'tracking')\n",
    "if not os.path.exists(SCRIPT_FOLDER):\n",
    "    os.mkdir(SCRIPT_FOLDER)\n",
    "if not os.path.exists(WORKSPACE):\n",
    "    os.mkdir(WORKSPACE)\n",
    "if not os.path.exists(TRACKING):\n",
    "    os.mkdir(TRACKING)\n",
    "\n",
    "\n",
    "\n",
    "# In[60]:\n",
    "\n",
    "BASECALLFOLDER = os.path.join(os.path.abspath(indir),'basecalled_data' )\n",
    "\n",
    "\n",
    "# In[61]:\n",
    "\n",
    "#not check again if the basecalled data is there\n",
    "if not os.path.exists(BASECALLFOLDER):\n",
    "    print('No basecalled data in %s. Exit!' % indir)\n",
    "    #exit()\n",
    "\n",
    "\n",
    "# In[62]:\n",
    "\n",
    "os.listdir(BASECALLFOLDER)\n",
    "\n",
    "\n",
    "# In[63]:\n",
    "\n",
    "if  len([os.path.join(BASECALLFOLDER, x) for x in os.listdir(BASECALLFOLDER) if x.endswith('tar.gz')]) == 1:\n",
    "    tar_file = [os.path.join(BASECALLFOLDER, x) for x in os.listdir(BASECALLFOLDER) if x.endswith('tar.gz')][0]\n",
    "    runid = tar_file.split('/')[-1].split('.')[0]\n",
    "else:\n",
    "        print('None or mulitiple tar files')\n",
    "\n",
    "\n",
    "# In[64]:\n",
    "\n",
    "#now untar the tar gz file\n",
    "os.chdir(BASECALLFOLDER)\n",
    "unzip_command = 'tar -xvf %s' %(tar_file)\n",
    "print(unzip_command)\n",
    "unzip_command_stderr = subprocess.check_output(unzip_command, shell=True, stderr=subprocess.STDOUT)\n",
    "\n",
    "\n",
    "# In[65]:\n",
    "\n",
    "#now we should have generated the folder with the basecallded data\n",
    "BASECALLED_DATA_FOLDER = os.path.join(BASECALLFOLDER, tar_file.split('.')[0])\n",
    "if not os.path.exists(BASECALLED_DATA_FOLDER):\n",
    "    print(\"Something with the unzipping of the basecalled data went wrong.\")\n",
    "\n",
    "\n",
    "# In[66]:\n",
    "\n",
    "#check the contents in the zipped folder\n",
    "content = ['workspace', 'configuration.cfg', 'sequencing_summary.txt', 'pipeline.log']\n",
    "if not set(os.listdir(BASECALLED_DATA_FOLDER)) == set(content):\n",
    "    print(\"Something with the unzipping of the basecalled data went wrong.\")\n",
    "\n",
    "BASECALLED_DATA_WORKSPACE = os.path.join(BASECALLED_DATA_FOLDER, 'workspace')\n",
    "\n",
    "\n",
    "# In[67]:\n",
    "\n",
    "PASS_FOLDER = os.path.join(BASECALLED_DATA_WORKSPACE, 'pass')\n",
    "os.listdir(PASS_FOLDER)\n",
    "\n",
    "\n",
    "# In[68]:\n",
    "\n",
    "FAIL_FOLDER = os.path.join(BASECALLED_DATA_WORKSPACE, 'fail')\n",
    "os.listdir(FAIL_FOLDER)\n",
    "\n",
    "\n",
    "# In[69]:\n",
    "\n",
    "sep_barcodes_list = sep_barcodes.split(',')\n",
    "\n",
    "\n",
    "# In[70]:\n",
    "\n",
    "# sep_barcodes_list\n",
    "\n",
    "# make a barcode 00 folder\n",
    "# loop throug all folders that are not in the sep_barcode list and combine those\n",
    "# add 00 to the sep_barcode list\n",
    "\n",
    "# sep_barcodes_list\n",
    "# In[71]:\n",
    "\n",
    "#make the barcode00 folder to combine all the misidentified/unclassified barcode\n",
    "if not os.path.exists(os.path.join(FAIL_FOLDER, 'barcode00')):\n",
    "    os.mkdir(os.path.join(FAIL_FOLDER, 'barcode00'))\n",
    "\n",
    "if not os.path.exists(os.path.join(PASS_FOLDER, 'barcode00')):\n",
    "    os.mkdir(os.path.join(PASS_FOLDER, 'barcode00'))\n",
    "\n",
    "\n",
    "# In[72]:\n",
    "\n",
    "['barcode%s' % x for x in sep_barcodes_list]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "# In[73]:\n",
    "\n",
    "#cat all barcode misclassified/unclassified reads in the pass folder to one fastq file and redirect to barcode00 folder\n",
    "\n",
    "if not os.path.isfile('%s/barcode00.fastq' % os.path.join(PASS_FOLDER, 'barcode00')):\n",
    "    for folder in [y.split('/')[-1] for y in os.listdir(PASS_FOLDER)]:\n",
    "        if folder not in ['barcode%s' % x for x in sep_barcodes_list]\\\n",
    "        and not folder == 'barcode00':\n",
    "            tmp_folder = os.path.join(PASS_FOLDER, folder)\n",
    "            #print(folder)\n",
    "            for fastq in [os.path.join(tmp_folder, z) for z in os.listdir(tmp_folder)  if z.endswith('fastq')]:\n",
    "                cat_fastq_command = 'cat %s' % fastq\n",
    "                #print(fastq)\n",
    "        cat_fastq_command +=' >> %s/barcode00.fastq' % os.path.join(PASS_FOLDER, 'barcode00')\n",
    "        print(cat_fastq_command)\n",
    "        cat_fastq_command_stderr = subprocess.check_output(cat_fastq_command, shell=True, stderr=subprocess.STDOUT)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['07', '08', '09', '10', '11']"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sep_barcodes_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#cat all barcode misclassified/unclassified reads in the fail folder to one fastq file and redirect to barcode00 folder\n",
    "if not os.path.isfile('%s/barcode00.fastq' % os.path.join(FAIL_FOLDER, 'barcode00')):\n",
    "    cat_fastq_command = 'cat'\n",
    "    for folder in [y.split('/')[-1] for y in os.listdir(FAIL_FOLDER)]:\n",
    "        if folder not in ['barcode%s' % x for x in sep_barcodes_list]\\\n",
    "        and not folder == 'barcode00':\n",
    "            tmp_folder = os.path.join(FAIL_FOLDER, folder)\n",
    "            #print(folder)\n",
    "            for fastq in [os.path.join(tmp_folder, z) for z in os.listdir(tmp_folder)  if z.endswith('fastq')]:\n",
    "                cat_fastq_command += ' %s' % fastq\n",
    "                #print(fastq)\n",
    "        cat_fastq_command +=' >> %s/barcode00.fastq' % os.path.join(FAIL_FOLDER, 'barcode00')\n",
    "        print(cat_fastq_command)\n",
    "    cat_fastq_command_stderr = subprocess.check_output(cat_fastq_command, shell=True, stderr=subprocess.STDOUT)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
