{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This is a notebook to pull down all the information for figure 3 (ncbi hits). The design is to compair between each sample (barcode) instead of each flowcell as they sequenced the same thing. Therefore I just pulled all the ncbi hits for each flowcell for each barcode and merged them together according to the barcode."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import os\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib\n",
    "from itertools import product\n",
    "import argparse\n",
    "from ete3 import NCBITaxa "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "ncbi = NCBITaxa()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# use argparse to do this\n",
    "#BASEDIR = args.BASEDIR\n",
    "BASEDIR = '/home/yiheng/data/20170617_FAH05432'\n",
    "\n",
    "# here we define the folder name of the dataframe it created by capturing the folder from the BASDIR\n",
    "folder_name = os.path.basename(BASEDIR)\n",
    "column_name = folder_name.split('_')[-1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# first check if the analysis folder is there\n",
    "folder_list = 'analysis  basecalled_data  scripts  tracking  workspace'.split(' ')\n",
    "for x in range(0,folder_list.count('')):\n",
    "    folder_list.remove('')\n",
    "#fix this test\n",
    "if not set(os.listdir(os.path.abspath(BASEDIR))) == set (folder_list):\n",
    "    print(\"Something wrong with basefolder. check it please.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/yiheng/anaconda3/lib/python3.6/site-packages/IPython/core/interactiveshell.py:2717: DtypeWarning: Columns (1,2,3,5,27) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  interactivity=interactivity, compiler=compiler, result=result)\n"
     ]
    }
   ],
   "source": [
    "# get the dataframe there\n",
    "dataframe = os.path.join(BASEDIR, 'analysis', 'summary_df_%s.tab' % folder_name)\n",
    "sum_df = pd.read_csv(dataframe, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['read_id', 'passes_filtering', 'sequence_length_template',\n",
       "       'mean_qscore_template', 'barcode_arrangement', 'barcode_score', 'kit',\n",
       "       'variant', 'pc_survived', 'nl_survived', 'qseqid_rg', 'sseqid_rg',\n",
       "       'evalue_rg', 'length_rg', 'pident_rg', 'nident_rg', 'sacc_rg',\n",
       "       'staxids_rg', 'scomnames_rg', 'read_length_pc_x', 'qseqid_nt',\n",
       "       'sseqid_nt', 'evalue_nt', 'length_nt', 'pident_nt', 'nident_nt',\n",
       "       'sacc_nt', 'staxids_nt', 'scomnames_nt', 'read_length_pc_y'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sum_df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# fills the nan with T/F for easier handling\n",
    "sum_df.sseqid_rg.fillna(False, inplace=True)\n",
    "sum_df.sseqid_nt.fillna(False, inplace=True)\n",
    "# filter out the rg blast hit\n",
    "ntblasthit_reads = sum_df[(sum_df.sseqid_rg == False) & (sum_df.sseqid_nt != False) & (sum_df.passes_filtering == True) & (sum_df.pc_survived == True) & (sum_df.nl_survived == True)]\n",
    "pd.set_option('display.max_columns', None)\n",
    "\n",
    "#Removing unncessary columns from joint_NCBI_df_taxonomy\n",
    "#I hard coded to select as it depends on what information you need\n",
    "for columns in ['passes_filtering', \n",
    "       'mean_qscore_template', 'barcode_score', 'kit',\n",
    "       'variant', 'pc_survived', 'nl_survived', 'qseqid_rg', 'sseqid_rg',\n",
    "       'evalue_rg', 'length_rg', 'pident_rg', 'nident_rg', 'sacc_rg',\n",
    "       'staxids_rg', 'scomnames_rg', 'read_length_pc_x', 'qseqid_nt',\n",
    "       'sseqid_nt', 'evalue_nt', 'length_nt', 'pident_nt', 'nident_nt',\n",
    "       'sacc_nt','scomnames_nt', 'read_length_pc_y']:\n",
    "    del ntblasthit_reads[columns] "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "###########This block for barcodes order\n",
    "# now everything left is what we want to plot out\n",
    "# arrange different barcodes\n",
    "# REMEMBER: This need to be manual check:\n",
    "# barcode01 is Pst infected sample, barcode02 is Zymo sample, 03 is nill sample, 04 is co-infection sample and 05 is Pyre sample\n",
    "new_barcodes = ['barcode01', 'barcode02', 'barcode03', 'barcode04', 'barcode05']\n",
    "if column_name == 'run1':\n",
    "    ntblasthit_barcode01 = ntblasthit_reads[ntblasthit_reads.barcode_arrangement == 'barcode02']\n",
    "    ntblasthit_barcode02 = ntblasthit_reads[ntblasthit_reads.barcode_arrangement == 'barcode06']\n",
    "    ntblasthit_barcode03 = ntblasthit_reads[ntblasthit_reads.barcode_arrangement == 'barcode04']\n",
    "    ntblasthit_barcode04 = ntblasthit_reads[ntblasthit_reads.barcode_arrangement == 'barcode05']\n",
    "    ntblasthit_barcode05 = ntblasthit_reads[ntblasthit_reads.barcode_arrangement == 'barcode03']\n",
    "    ntblasthit_barcode01.barcode_arrangement = ntblasthit_barcode01.barcode_arrangement.replace(to_replace='barcode02', value='barcode01')\n",
    "    ntblasthit_reads.drop(ntblasthit_reads.index[ntblasthit_reads.barcode_arrangement.str.contains('barcode07')], inplace = True)\n",
    "    ntblasthit_barcode01.barcode_arrangement = ntblasthit_barcode01.barcode_arrangement.replace(to_replace='barcode03', value='barcode07')\n",
    "    ntblasthit_barcode01.barcode_arrangement = ntblasthit_barcode01.barcode_arrangement.replace(to_replace='barcode06', value='barcode02')\n",
    "    ntblasthit_barcode01.barcode_arrangement = ntblasthit_barcode01.barcode_arrangement.replace(to_replace='barcode04', value='barcode03')\n",
    "    ntblasthit_barcode01.barcode_arrangement = ntblasthit_barcode01.barcode_arrangement.replace(to_replace='barcode05', value='barcode04')\n",
    "    ntblasthit_barcode01.barcode_arrangement = ntblasthit_barcode01.barcode_arrangement.replace(to_replace='barcode07', value='barcode05')\n",
    "\n",
    "elif column_name == 'run2':\n",
    "    # first delete the rows that contains the barcode we are going to replace\n",
    "    # I think maybe better to first adjust the df to get rid of all the mis/unclassified reads so the df get smaller\n",
    "    # and we do not need those reads for plotting\n",
    "    \n",
    "    ntblasthit_reads.drop(ntblasthit_reads.index[ntblasthit_reads.barcode_arrangement.str.contains('barcode01')], inplace = True)\n",
    "    ntblasthit_reads.drop(ntblasthit_reads.index[ntblasthit_reads.barcode_arrangement.str.contains('barcode02')], inplace = True)\n",
    "    ntblasthit_reads.drop(ntblasthit_reads.index[ntblasthit_reads.barcode_arrangement.str.contains('barcode03')], inplace = True)\n",
    "    ntblasthit_reads.drop(ntblasthit_reads.index[ntblasthit_reads.barcode_arrangement.str.contains('barcode04')], inplace = True)\n",
    "    ntblasthit_reads.drop(ntblasthit_reads.index[ntblasthit_reads.barcode_arrangement.str.contains('barcode05')], inplace = True)\n",
    "    ntblasthit_reads.drop(ntblasthit_reads.index[ntblasthit_reads.barcode_arrangement.str.contains('unclassified')], inplace = True)\n",
    "    old_barcodes = ['barcode07', 'barcode08', 'barcode09', 'barcode10', 'barcode11']\n",
    "    for old_barcode, new_barcode in zip(old_barcodes, new_barcodes):\n",
    "        ntblasthit_reads.loc\\\n",
    "        [ntblasthit_reads[ntblasthit_reads['barcode_arrangement'] == old_barcode].index,'barcode_arrangement']\\\n",
    "        = [new_barcode]\n",
    "    #ntblasthit_reads['barcode_arrangement'] = ntblasthit_reads.barcode_arrangement.replace(to_replace='barcode07', value='barcode01')\n",
    "\n",
    "    \n",
    "elif column_name == 'FAH05432':\n",
    "    ntblasthit_barcode01 = ntblasthit_reads[ntblasthit_reads.barcode_arrangement == 'barcode01']\n",
    "    ntblasthit_barcode02 = ntblasthit_reads[ntblasthit_reads.barcode_arrangement == 'barcode02']\n",
    "    ntblasthit_barcode03 = ntblasthit_reads[ntblasthit_reads.barcode_arrangement == 'barcode03']\n",
    "    ntblasthit_barcode04 = ntblasthit_reads[ntblasthit_reads.barcode_arrangement == 'barcode04']\n",
    "    ntblasthit_barcode05 = ntblasthit_reads[ntblasthit_reads.barcode_arrangement == 'barcode05']\n",
    "elif column_name == 'FAH05731':\n",
    "    ntblasthit_barcode01 = ntblasthit_reads[ntblasthit_reads.barcode_arrangement == 'barcode01']\n",
    "    ntblasthit_barcode02 = ntblasthit_reads[ntblasthit_reads.barcode_arrangement == 'barcode02']\n",
    "    ntblasthit_barcode03 = ntblasthit_reads[ntblasthit_reads.barcode_arrangement == 'barcode03']\n",
    "    ntblasthit_barcode04 = ntblasthit_reads[ntblasthit_reads.barcode_arrangement == 'barcode04']\n",
    "    ntblasthit_barcode05 = ntblasthit_reads[ntblasthit_reads.barcode_arrangement == 'barcode05']\n",
    "else:\n",
    "    pass\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# now just filter out all the reads that are not going to used for ploting\n",
    "ntblasthit_reads_barcode01 = ntblasthit_reads[ntblasthit_reads.barcode_arrangement.str.contains('barcode01')]\n",
    "ntblasthit_reads_barcode02 = ntblasthit_reads[ntblasthit_reads.barcode_arrangement.str.contains('barcode02')]\n",
    "ntblasthit_reads_barcode03 = ntblasthit_reads[ntblasthit_reads.barcode_arrangement.str.contains('barcode03')]\n",
    "ntblasthit_reads_barcode04 = ntblasthit_reads[ntblasthit_reads.barcode_arrangement.str.contains('barcode04')]\n",
    "ntblasthit_reads_barcode05 = ntblasthit_reads[ntblasthit_reads.barcode_arrangement.str.contains('barcode05')]\n",
    "\n",
    "# now concat them together. but remember that the index has to change to add the column for ncbi taxa\n",
    "ntblasthit_reads_filtered_barcodes = pd.concat([ntblasthit_reads_barcode01, \n",
    "                                                ntblasthit_reads_barcode02, \n",
    "                                                ntblasthit_reads_barcode03, \n",
    "                                                ntblasthit_reads_barcode04, \n",
    "                                                ntblasthit_reads_barcode05], ignore_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# as the concate before will result the staxid into a string\n",
    "# so now change the string back to float so it can be recognized by the NCBITaxa\n",
    "for taxid in ntblasthit_reads_filtered_barcodes['staxids_nt']:\n",
    "    if ';' in str(taxid):\n",
    "        taxid = taxid.split(';')[0]\n",
    "    else:\n",
    "        pass\n",
    "    float(taxid)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Function to generate taxonomy columns based on NCBITaxa results for NCBI hit dataframe, \n",
    "def search_rank_output_name_append_column(df, staxid_column, rank_search):\n",
    "    \"\"\"Input df, staxid_column from same df and rank_search (a desired taxonomic rank \n",
    "    from each staxid's lineage), outputs taxonomic name corresponding to rank_search or 'Unclassified' if\n",
    "    unavailable and appends to df row by row\"\"\"\n",
    "    rank_list = []\n",
    "    for read_index in range(0, len(staxid_column)):\n",
    "        taxid = ''\n",
    "        if ';' in str(staxid_column[read_index]):\n",
    "            taxid = staxid_column[read_index].split(';')[0]\n",
    "        else:\n",
    "            taxid = staxid_column[read_index]\n",
    "        \n",
    "        taxid_lineage = ''\n",
    "        taxid_lineage = ncbi.get_lineage(taxid)\n",
    "        \n",
    "        names = ''\n",
    "        names = ncbi.get_taxid_translator(taxid_lineage)\n",
    "        \n",
    "        ranks = ''\n",
    "        ranks = ncbi.get_rank(taxid_lineage) #Dict\n",
    "        \n",
    "        ranks2names = ''\n",
    "        ranks2names = {ranks[k]:names[k] for k in names.keys() & ranks}\n",
    "        \n",
    "        if rank_search in ranks2names.keys():\n",
    "            rank_list.append(ranks2names[rank_search])#if rank in dict, print name\n",
    "        else:\n",
    "            rank_list.append('Unclassified')\n",
    "    df[rank_search] = rank_list\n",
    "# NOTE: Appending is always slow, try and find a better way e.g df.apply to a column based on staxids column\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#ntblasthit_reads_filtered_barcodes_added_TaxaRank = ntblasthit_reads_filtered_barcodes.copy()\n",
    "rank_list = ['superkingdom', 'phylum', 'class', 'order', 'family', 'genus', 'species']\n",
    "for rank in rank_list:\n",
    "    search_rank_output_name_append_column(ntblasthit_reads_filtered_barcodes, \n",
    "                                          ntblasthit_reads_filtered_barcodes.staxids_nt, \n",
    "                                          rank)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>read_id</th>\n",
       "      <th>sequence_length_template</th>\n",
       "      <th>barcode_arrangement</th>\n",
       "      <th>staxids_nt</th>\n",
       "      <th>superkingdom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>807a8997-b021-4820-b33a-1b9095249dc7</td>\n",
       "      <td>2695</td>\n",
       "      <td>barcode01</td>\n",
       "      <td>200451</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Pseudomonadales</td>\n",
       "      <td>Pseudomonadaceae</td>\n",
       "      <td>Pseudomonas</td>\n",
       "      <td>Pseudomonas poae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>c53fa9c8-f4d5-4876-8526-e04fcea0ac5d</td>\n",
       "      <td>3286</td>\n",
       "      <td>barcode01</td>\n",
       "      <td>200451</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Pseudomonadales</td>\n",
       "      <td>Pseudomonadaceae</td>\n",
       "      <td>Pseudomonas</td>\n",
       "      <td>Pseudomonas poae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>19a62259-0925-41ef-94e4-6fe9ef776c81</td>\n",
       "      <td>2455</td>\n",
       "      <td>barcode01</td>\n",
       "      <td>192087</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Pseudomonadales</td>\n",
       "      <td>Pseudomonadaceae</td>\n",
       "      <td>Pseudomonas</td>\n",
       "      <td>Pseudomonas syringae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>7e4e83ba-26fa-4d0b-b714-724976a9c7a7</td>\n",
       "      <td>2347</td>\n",
       "      <td>barcode01</td>\n",
       "      <td>200451</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Pseudomonadales</td>\n",
       "      <td>Pseudomonadaceae</td>\n",
       "      <td>Pseudomonas</td>\n",
       "      <td>Pseudomonas poae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>c92e3d57-c554-4d73-9e17-a51bc14aa4d4</td>\n",
       "      <td>3210</td>\n",
       "      <td>barcode01</td>\n",
       "      <td>192087</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Pseudomonadales</td>\n",
       "      <td>Pseudomonadaceae</td>\n",
       "      <td>Pseudomonas</td>\n",
       "      <td>Pseudomonas syringae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>072bbc16-20a5-45b8-bcbf-e7239f48ed12</td>\n",
       "      <td>3164</td>\n",
       "      <td>barcode01</td>\n",
       "      <td>321</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Pseudomonadales</td>\n",
       "      <td>Pseudomonadaceae</td>\n",
       "      <td>Pseudomonas</td>\n",
       "      <td>Pseudomonas syringae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>7eb68975-bd44-40da-b0e8-2635fcedfbb0</td>\n",
       "      <td>2892</td>\n",
       "      <td>barcode01</td>\n",
       "      <td>552</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>Erwiniaceae</td>\n",
       "      <td>Erwinia</td>\n",
       "      <td>Erwinia amylovora</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>f876c268-fe76-4fb7-97e7-e3c6fa2d8c8b</td>\n",
       "      <td>3054</td>\n",
       "      <td>barcode01</td>\n",
       "      <td>1.28236e+06</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Pseudomonadales</td>\n",
       "      <td>Pseudomonadaceae</td>\n",
       "      <td>Pseudomonas</td>\n",
       "      <td>Pseudomonas poae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>2143605b-7298-43af-b034-400dd381acc7</td>\n",
       "      <td>2512</td>\n",
       "      <td>barcode01</td>\n",
       "      <td>48100</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Ascomycota</td>\n",
       "      <td>Dothideomycetes</td>\n",
       "      <td>Pleosporales</td>\n",
       "      <td>Pleosporaceae</td>\n",
       "      <td>Alternaria</td>\n",
       "      <td>Alternaria solani</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>00f1b4d2-cc23-4231-9b51-543873ddddd2</td>\n",
       "      <td>3047</td>\n",
       "      <td>barcode01</td>\n",
       "      <td>634500</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>Erwiniaceae</td>\n",
       "      <td>Erwinia</td>\n",
       "      <td>Erwinia billingiae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>74885d13-7285-4de6-ae85-89a8927c36ab</td>\n",
       "      <td>2751</td>\n",
       "      <td>barcode01</td>\n",
       "      <td>1.50567e+06</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Basidiomycota</td>\n",
       "      <td>Pucciniomycetes</td>\n",
       "      <td>Pucciniales</td>\n",
       "      <td>Pucciniaceae</td>\n",
       "      <td>Puccinia</td>\n",
       "      <td>Puccinia cf. psidii AE-2014</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>fe662e8d-bb9d-45d5-b699-7891406c5df4</td>\n",
       "      <td>1916</td>\n",
       "      <td>barcode01</td>\n",
       "      <td>192087</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Pseudomonadales</td>\n",
       "      <td>Pseudomonadaceae</td>\n",
       "      <td>Pseudomonas</td>\n",
       "      <td>Pseudomonas syringae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>d5533aed-db93-4957-928a-60cd6caa2d0f</td>\n",
       "      <td>2941</td>\n",
       "      <td>barcode01</td>\n",
       "      <td>199201</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Pseudomonadales</td>\n",
       "      <td>Pseudomonadaceae</td>\n",
       "      <td>Pseudomonas</td>\n",
       "      <td>Pseudomonas syringae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>2b27f02a-bd3c-4c0d-a4fb-7e1575c76fe0</td>\n",
       "      <td>2138</td>\n",
       "      <td>barcode01</td>\n",
       "      <td>889211</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>Erwiniaceae</td>\n",
       "      <td>Erwinia</td>\n",
       "      <td>Erwinia amylovora</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>d6a81cd2-3423-4242-9cb3-84f81be34892</td>\n",
       "      <td>2143</td>\n",
       "      <td>barcode01</td>\n",
       "      <td>192087</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Pseudomonadales</td>\n",
       "      <td>Pseudomonadaceae</td>\n",
       "      <td>Pseudomonas</td>\n",
       "      <td>Pseudomonas syringae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>2828d03d-7417-46c5-8382-792e21003d99</td>\n",
       "      <td>3452</td>\n",
       "      <td>barcode01</td>\n",
       "      <td>192087</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Pseudomonadales</td>\n",
       "      <td>Pseudomonadaceae</td>\n",
       "      <td>Pseudomonas</td>\n",
       "      <td>Pseudomonas syringae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>f7760040-5944-4aec-9d3f-b98389bda403</td>\n",
       "      <td>415</td>\n",
       "      <td>barcode01</td>\n",
       "      <td>587753</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Pseudomonadales</td>\n",
       "      <td>Pseudomonadaceae</td>\n",
       "      <td>Pseudomonas</td>\n",
       "      <td>Pseudomonas chlororaphis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>fa63e5c1-71b2-42e9-906d-879ad1ad4f3b</td>\n",
       "      <td>3081</td>\n",
       "      <td>barcode01</td>\n",
       "      <td>665914</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>Erwiniaceae</td>\n",
       "      <td>Pantoea</td>\n",
       "      <td>Pantoea gaviniae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>be038069-e219-48b3-8352-a70beaa767e0</td>\n",
       "      <td>2375</td>\n",
       "      <td>barcode01</td>\n",
       "      <td>192087</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Pseudomonadales</td>\n",
       "      <td>Pseudomonadaceae</td>\n",
       "      <td>Pseudomonas</td>\n",
       "      <td>Pseudomonas syringae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>f31b2c3c-87ba-4957-a72c-dd8126a70476</td>\n",
       "      <td>1778</td>\n",
       "      <td>barcode01</td>\n",
       "      <td>199201</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Pseudomonadales</td>\n",
       "      <td>Pseudomonadaceae</td>\n",
       "      <td>Pseudomonas</td>\n",
       "      <td>Pseudomonas syringae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>52767e8d-6540-439b-a930-bb5f8ece7bb8</td>\n",
       "      <td>1981</td>\n",
       "      <td>barcode01</td>\n",
       "      <td>1.32493e+06</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Pseudomonadales</td>\n",
       "      <td>Pseudomonadaceae</td>\n",
       "      <td>Pseudomonas</td>\n",
       "      <td>Pseudomonas syringae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>b89419aa-7906-4bdb-ad6d-06bd26e65944</td>\n",
       "      <td>2161</td>\n",
       "      <td>barcode01</td>\n",
       "      <td>35703</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "      <td>Citrobacter</td>\n",
       "      <td>Citrobacter amalonaticus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>6047063f-eb37-4ef4-a9b0-304ced98946e</td>\n",
       "      <td>3141</td>\n",
       "      <td>barcode01</td>\n",
       "      <td>418459</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Basidiomycota</td>\n",
       "      <td>Pucciniomycetes</td>\n",
       "      <td>Pucciniales</td>\n",
       "      <td>Pucciniaceae</td>\n",
       "      <td>Puccinia</td>\n",
       "      <td>Puccinia graminis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>af09a516-1d55-4685-8bf6-dba8d13eb50e</td>\n",
       "      <td>728</td>\n",
       "      <td>barcode01</td>\n",
       "      <td>192087</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Pseudomonadales</td>\n",
       "      <td>Pseudomonadaceae</td>\n",
       "      <td>Pseudomonas</td>\n",
       "      <td>Pseudomonas syringae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>b0ded9cf-3b68-4dbb-b12c-c7e8b18dd35e</td>\n",
       "      <td>3073</td>\n",
       "      <td>barcode01</td>\n",
       "      <td>1.32493e+06</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Pseudomonadales</td>\n",
       "      <td>Pseudomonadaceae</td>\n",
       "      <td>Pseudomonas</td>\n",
       "      <td>Pseudomonas syringae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>8e937eda-6ba6-4f1c-b5ea-fd9f3690ac2c</td>\n",
       "      <td>2018</td>\n",
       "      <td>barcode01</td>\n",
       "      <td>465817</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>Erwiniaceae</td>\n",
       "      <td>Erwinia</td>\n",
       "      <td>Erwinia tasmaniensis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>ec446ef2-0236-4dc3-a8db-29dcf2a4d5f8</td>\n",
       "      <td>3006</td>\n",
       "      <td>barcode01</td>\n",
       "      <td>199201</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Pseudomonadales</td>\n",
       "      <td>Pseudomonadaceae</td>\n",
       "      <td>Pseudomonas</td>\n",
       "      <td>Pseudomonas syringae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>502dab27-225d-4462-ab53-5572a4ff7d46</td>\n",
       "      <td>2049</td>\n",
       "      <td>barcode01</td>\n",
       "      <td>1.28236e+06</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Pseudomonadales</td>\n",
       "      <td>Pseudomonadaceae</td>\n",
       "      <td>Pseudomonas</td>\n",
       "      <td>Pseudomonas poae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>0d733956-7720-40c7-9ab0-19842a016005</td>\n",
       "      <td>2015</td>\n",
       "      <td>barcode01</td>\n",
       "      <td>192087</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Pseudomonadales</td>\n",
       "      <td>Pseudomonadaceae</td>\n",
       "      <td>Pseudomonas</td>\n",
       "      <td>Pseudomonas syringae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>c77016d6-1b7f-49f7-8c4b-6c7ca171b291</td>\n",
       "      <td>2332</td>\n",
       "      <td>barcode01</td>\n",
       "      <td>1.32493e+06</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Pseudomonadales</td>\n",
       "      <td>Pseudomonadaceae</td>\n",
       "      <td>Pseudomonas</td>\n",
       "      <td>Pseudomonas syringae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6030</th>\n",
       "      <td>036f16cc-2e7a-4414-94a2-55b80e502345</td>\n",
       "      <td>2046</td>\n",
       "      <td>barcode05</td>\n",
       "      <td>48100</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Ascomycota</td>\n",
       "      <td>Dothideomycetes</td>\n",
       "      <td>Pleosporales</td>\n",
       "      <td>Pleosporaceae</td>\n",
       "      <td>Alternaria</td>\n",
       "      <td>Alternaria solani</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6031</th>\n",
       "      <td>3febc7c5-f308-4f64-9b37-bc2357608d43</td>\n",
       "      <td>926</td>\n",
       "      <td>barcode05</td>\n",
       "      <td>317</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Pseudomonadales</td>\n",
       "      <td>Pseudomonadaceae</td>\n",
       "      <td>Pseudomonas</td>\n",
       "      <td>Pseudomonas syringae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6032</th>\n",
       "      <td>2de9f0c0-1338-4978-9f88-f34336e3c16d</td>\n",
       "      <td>3216</td>\n",
       "      <td>barcode05</td>\n",
       "      <td>341663</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Ascomycota</td>\n",
       "      <td>Eurotiomycetes</td>\n",
       "      <td>Eurotiales</td>\n",
       "      <td>Aspergillaceae</td>\n",
       "      <td>Aspergillus</td>\n",
       "      <td>Aspergillus terreus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6033</th>\n",
       "      <td>1c4cf73a-df25-4eb8-a96d-dfa117243cb0</td>\n",
       "      <td>2046</td>\n",
       "      <td>barcode05</td>\n",
       "      <td>398673</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Ascomycota</td>\n",
       "      <td>Sordariomycetes</td>\n",
       "      <td>Hypocreales</td>\n",
       "      <td>Hypocreaceae</td>\n",
       "      <td>Trichoderma</td>\n",
       "      <td>Trichoderma gamsii</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6034</th>\n",
       "      <td>f0f4e05a-7bf0-4eeb-9a0a-5b84c6590bc2</td>\n",
       "      <td>2568</td>\n",
       "      <td>barcode05</td>\n",
       "      <td>76761</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Pseudomonadales</td>\n",
       "      <td>Pseudomonadaceae</td>\n",
       "      <td>Pseudomonas</td>\n",
       "      <td>Pseudomonas veronii</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6035</th>\n",
       "      <td>b109ccf5-048d-45e6-a0c1-e568d4498b66</td>\n",
       "      <td>2575</td>\n",
       "      <td>barcode05</td>\n",
       "      <td>1.6036e+06</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Insecta</td>\n",
       "      <td>Hemiptera</td>\n",
       "      <td>Pentatomidae</td>\n",
       "      <td>Cressona</td>\n",
       "      <td>Cressona rufa</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6036</th>\n",
       "      <td>00cb05e4-9dcc-4562-97b6-1d34db28b2ce</td>\n",
       "      <td>699</td>\n",
       "      <td>barcode05</td>\n",
       "      <td>169297</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Liliopsida</td>\n",
       "      <td>Poales</td>\n",
       "      <td>Poaceae</td>\n",
       "      <td>Aegilops</td>\n",
       "      <td>Aegilops tauschii</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6037</th>\n",
       "      <td>0be8c7f8-9da2-4d6c-84bc-4852d29a23ff</td>\n",
       "      <td>3670</td>\n",
       "      <td>barcode05</td>\n",
       "      <td>48100</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Ascomycota</td>\n",
       "      <td>Dothideomycetes</td>\n",
       "      <td>Pleosporales</td>\n",
       "      <td>Pleosporaceae</td>\n",
       "      <td>Alternaria</td>\n",
       "      <td>Alternaria solani</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6038</th>\n",
       "      <td>4e1bc3af-c37e-44be-82e4-765000b17887</td>\n",
       "      <td>2376</td>\n",
       "      <td>barcode05</td>\n",
       "      <td>45151</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Ascomycota</td>\n",
       "      <td>Dothideomycetes</td>\n",
       "      <td>Pleosporales</td>\n",
       "      <td>Pleosporaceae</td>\n",
       "      <td>Pyrenophora</td>\n",
       "      <td>Pyrenophora tritici-repentis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6039</th>\n",
       "      <td>54b4c516-a392-4c67-bc3f-20d9cb08dfd1</td>\n",
       "      <td>3185</td>\n",
       "      <td>barcode05</td>\n",
       "      <td>1.32493e+06</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Pseudomonadales</td>\n",
       "      <td>Pseudomonadaceae</td>\n",
       "      <td>Pseudomonas</td>\n",
       "      <td>Pseudomonas syringae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6040</th>\n",
       "      <td>e15f756e-f926-4da7-bc99-2a1b94308bf9</td>\n",
       "      <td>5388</td>\n",
       "      <td>barcode05</td>\n",
       "      <td>112498</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Ascomycota</td>\n",
       "      <td>Dothideomycetes</td>\n",
       "      <td>Capnodiales</td>\n",
       "      <td>Mycosphaerellaceae</td>\n",
       "      <td>Ramularia</td>\n",
       "      <td>Ramularia collo-cygni</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6041</th>\n",
       "      <td>5dfeeb74-f577-4706-bc10-4173ea8c13d4</td>\n",
       "      <td>3439</td>\n",
       "      <td>barcode05</td>\n",
       "      <td>671987</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Ascomycota</td>\n",
       "      <td>Dothideomycetes</td>\n",
       "      <td>Pleosporales</td>\n",
       "      <td>Pleosporaceae</td>\n",
       "      <td>Exserohilum</td>\n",
       "      <td>Exserohilum turcicum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6042</th>\n",
       "      <td>248cff4b-734e-402d-95ba-aeda3dfd5a78</td>\n",
       "      <td>3337</td>\n",
       "      <td>barcode05</td>\n",
       "      <td>29760</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Vitales</td>\n",
       "      <td>Vitaceae</td>\n",
       "      <td>Vitis</td>\n",
       "      <td>Vitis vinifera</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6043</th>\n",
       "      <td>434bf795-fa15-4709-872e-fea2f0bf28fa</td>\n",
       "      <td>860</td>\n",
       "      <td>barcode05</td>\n",
       "      <td>4565</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Liliopsida</td>\n",
       "      <td>Poales</td>\n",
       "      <td>Poaceae</td>\n",
       "      <td>Triticum</td>\n",
       "      <td>Triticum aestivum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6044</th>\n",
       "      <td>0e196992-8736-4a23-a766-b6a48ac53f92</td>\n",
       "      <td>443</td>\n",
       "      <td>barcode05</td>\n",
       "      <td>7165</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Insecta</td>\n",
       "      <td>Diptera</td>\n",
       "      <td>Culicidae</td>\n",
       "      <td>Anopheles</td>\n",
       "      <td>Anopheles gambiae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6045</th>\n",
       "      <td>3422c128-320c-40a4-8f54-8df22e42117e</td>\n",
       "      <td>2227</td>\n",
       "      <td>barcode05</td>\n",
       "      <td>51953</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Liliopsida</td>\n",
       "      <td>Arecales</td>\n",
       "      <td>Arecaceae</td>\n",
       "      <td>Elaeis</td>\n",
       "      <td>Elaeis guineensis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6046</th>\n",
       "      <td>afe0dea5-5835-41ad-a3d5-0111885fadb4</td>\n",
       "      <td>4887</td>\n",
       "      <td>barcode05</td>\n",
       "      <td>562</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>Escherichia coli</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6047</th>\n",
       "      <td>74af4e06-3343-446c-8d5d-4b0904a192fe</td>\n",
       "      <td>2957</td>\n",
       "      <td>barcode05</td>\n",
       "      <td>39946</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Liliopsida</td>\n",
       "      <td>Poales</td>\n",
       "      <td>Poaceae</td>\n",
       "      <td>Oryza</td>\n",
       "      <td>Oryza sativa</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6048</th>\n",
       "      <td>b734a7f7-1c8f-4dcf-a175-702a74f07473</td>\n",
       "      <td>3285</td>\n",
       "      <td>barcode05</td>\n",
       "      <td>5599</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Ascomycota</td>\n",
       "      <td>Dothideomycetes</td>\n",
       "      <td>Pleosporales</td>\n",
       "      <td>Pleosporaceae</td>\n",
       "      <td>Alternaria</td>\n",
       "      <td>Alternaria alternata</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6049</th>\n",
       "      <td>7e83c505-957c-43a7-bd2f-8921b8bacede</td>\n",
       "      <td>2764</td>\n",
       "      <td>barcode05</td>\n",
       "      <td>930090</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Ascomycota</td>\n",
       "      <td>Dothideomycetes</td>\n",
       "      <td>Pleosporales</td>\n",
       "      <td>Pleosporaceae</td>\n",
       "      <td>Bipolaris</td>\n",
       "      <td>Bipolaris oryzae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6050</th>\n",
       "      <td>9607245c-061a-4f58-8d18-a93d2b3e8cdc</td>\n",
       "      <td>2592</td>\n",
       "      <td>barcode05</td>\n",
       "      <td>7955</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Actinopteri</td>\n",
       "      <td>Cypriniformes</td>\n",
       "      <td>Cyprinidae</td>\n",
       "      <td>Danio</td>\n",
       "      <td>Danio rerio</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6051</th>\n",
       "      <td>9a4c0335-3fbf-4b16-ac09-a346b6365978</td>\n",
       "      <td>1834</td>\n",
       "      <td>barcode05</td>\n",
       "      <td>4999</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Basidiomycota</td>\n",
       "      <td>Tremellomycetes</td>\n",
       "      <td>Tremellales</td>\n",
       "      <td>Cuniculitremaceae</td>\n",
       "      <td>Kockovaella</td>\n",
       "      <td>Kockovaella imperatae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6052</th>\n",
       "      <td>f6d6b97e-9204-41ef-9e43-2126f7dd8557</td>\n",
       "      <td>351</td>\n",
       "      <td>barcode05</td>\n",
       "      <td>7955</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Actinopteri</td>\n",
       "      <td>Cypriniformes</td>\n",
       "      <td>Cyprinidae</td>\n",
       "      <td>Danio</td>\n",
       "      <td>Danio rerio</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6053</th>\n",
       "      <td>a9dab824-a355-4c7d-bad8-9f46ad5db179</td>\n",
       "      <td>2552</td>\n",
       "      <td>barcode05</td>\n",
       "      <td>353152</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Apicomplexa</td>\n",
       "      <td>Conoidasida</td>\n",
       "      <td>Eucoccidiorida</td>\n",
       "      <td>Cryptosporidiidae</td>\n",
       "      <td>Cryptosporidium</td>\n",
       "      <td>Cryptosporidium parvum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6054</th>\n",
       "      <td>032c2734-9254-49df-b949-f21606858835</td>\n",
       "      <td>2109</td>\n",
       "      <td>barcode05</td>\n",
       "      <td>741155</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Pseudomonadales</td>\n",
       "      <td>Pseudomonadaceae</td>\n",
       "      <td>Pseudomonas</td>\n",
       "      <td>Pseudomonas alcaliphila</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6055</th>\n",
       "      <td>59353f77-87ee-40a0-930f-328719e2bfe4</td>\n",
       "      <td>2592</td>\n",
       "      <td>barcode05</td>\n",
       "      <td>48100</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Ascomycota</td>\n",
       "      <td>Dothideomycetes</td>\n",
       "      <td>Pleosporales</td>\n",
       "      <td>Pleosporaceae</td>\n",
       "      <td>Alternaria</td>\n",
       "      <td>Alternaria solani</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6056</th>\n",
       "      <td>86bf9dba-4728-463d-b767-231f0863ebea</td>\n",
       "      <td>2595</td>\n",
       "      <td>barcode05</td>\n",
       "      <td>5599</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Ascomycota</td>\n",
       "      <td>Dothideomycetes</td>\n",
       "      <td>Pleosporales</td>\n",
       "      <td>Pleosporaceae</td>\n",
       "      <td>Alternaria</td>\n",
       "      <td>Alternaria alternata</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6057</th>\n",
       "      <td>88013815-813d-4f85-833c-2d5d0f647a94</td>\n",
       "      <td>3832</td>\n",
       "      <td>barcode05</td>\n",
       "      <td>930090</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Ascomycota</td>\n",
       "      <td>Dothideomycetes</td>\n",
       "      <td>Pleosporales</td>\n",
       "      <td>Pleosporaceae</td>\n",
       "      <td>Bipolaris</td>\n",
       "      <td>Bipolaris oryzae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6058</th>\n",
       "      <td>62375b9c-54db-4097-95af-4bd20c723822</td>\n",
       "      <td>2544</td>\n",
       "      <td>barcode05</td>\n",
       "      <td>33069</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Pseudomonadales</td>\n",
       "      <td>Pseudomonadaceae</td>\n",
       "      <td>Pseudomonas</td>\n",
       "      <td>Pseudomonas viridiflava</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6059</th>\n",
       "      <td>7f21fa02-74cf-4d6e-896d-76ebc4773042</td>\n",
       "      <td>3987</td>\n",
       "      <td>barcode05</td>\n",
       "      <td>1785145</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Pseudomonadales</td>\n",
       "      <td>Pseudomonadaceae</td>\n",
       "      <td>Pseudomonas</td>\n",
       "      <td>Pseudomonas sp. MS586</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6060 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   read_id sequence_length_template  \\\n",
       "0     807a8997-b021-4820-b33a-1b9095249dc7                     2695   \n",
       "1     c53fa9c8-f4d5-4876-8526-e04fcea0ac5d                     3286   \n",
       "2     19a62259-0925-41ef-94e4-6fe9ef776c81                     2455   \n",
       "3     7e4e83ba-26fa-4d0b-b714-724976a9c7a7                     2347   \n",
       "4     c92e3d57-c554-4d73-9e17-a51bc14aa4d4                     3210   \n",
       "5     072bbc16-20a5-45b8-bcbf-e7239f48ed12                     3164   \n",
       "6     7eb68975-bd44-40da-b0e8-2635fcedfbb0                     2892   \n",
       "7     f876c268-fe76-4fb7-97e7-e3c6fa2d8c8b                     3054   \n",
       "8     2143605b-7298-43af-b034-400dd381acc7                     2512   \n",
       "9     00f1b4d2-cc23-4231-9b51-543873ddddd2                     3047   \n",
       "10    74885d13-7285-4de6-ae85-89a8927c36ab                     2751   \n",
       "11    fe662e8d-bb9d-45d5-b699-7891406c5df4                     1916   \n",
       "12    d5533aed-db93-4957-928a-60cd6caa2d0f                     2941   \n",
       "13    2b27f02a-bd3c-4c0d-a4fb-7e1575c76fe0                     2138   \n",
       "14    d6a81cd2-3423-4242-9cb3-84f81be34892                     2143   \n",
       "15    2828d03d-7417-46c5-8382-792e21003d99                     3452   \n",
       "16    f7760040-5944-4aec-9d3f-b98389bda403                      415   \n",
       "17    fa63e5c1-71b2-42e9-906d-879ad1ad4f3b                     3081   \n",
       "18    be038069-e219-48b3-8352-a70beaa767e0                     2375   \n",
       "19    f31b2c3c-87ba-4957-a72c-dd8126a70476                     1778   \n",
       "20    52767e8d-6540-439b-a930-bb5f8ece7bb8                     1981   \n",
       "21    b89419aa-7906-4bdb-ad6d-06bd26e65944                     2161   \n",
       "22    6047063f-eb37-4ef4-a9b0-304ced98946e                     3141   \n",
       "23    af09a516-1d55-4685-8bf6-dba8d13eb50e                      728   \n",
       "24    b0ded9cf-3b68-4dbb-b12c-c7e8b18dd35e                     3073   \n",
       "25    8e937eda-6ba6-4f1c-b5ea-fd9f3690ac2c                     2018   \n",
       "26    ec446ef2-0236-4dc3-a8db-29dcf2a4d5f8                     3006   \n",
       "27    502dab27-225d-4462-ab53-5572a4ff7d46                     2049   \n",
       "28    0d733956-7720-40c7-9ab0-19842a016005                     2015   \n",
       "29    c77016d6-1b7f-49f7-8c4b-6c7ca171b291                     2332   \n",
       "...                                    ...                      ...   \n",
       "6030  036f16cc-2e7a-4414-94a2-55b80e502345                     2046   \n",
       "6031  3febc7c5-f308-4f64-9b37-bc2357608d43                      926   \n",
       "6032  2de9f0c0-1338-4978-9f88-f34336e3c16d                     3216   \n",
       "6033  1c4cf73a-df25-4eb8-a96d-dfa117243cb0                     2046   \n",
       "6034  f0f4e05a-7bf0-4eeb-9a0a-5b84c6590bc2                     2568   \n",
       "6035  b109ccf5-048d-45e6-a0c1-e568d4498b66                     2575   \n",
       "6036  00cb05e4-9dcc-4562-97b6-1d34db28b2ce                      699   \n",
       "6037  0be8c7f8-9da2-4d6c-84bc-4852d29a23ff                     3670   \n",
       "6038  4e1bc3af-c37e-44be-82e4-765000b17887                     2376   \n",
       "6039  54b4c516-a392-4c67-bc3f-20d9cb08dfd1                     3185   \n",
       "6040  e15f756e-f926-4da7-bc99-2a1b94308bf9                     5388   \n",
       "6041  5dfeeb74-f577-4706-bc10-4173ea8c13d4                     3439   \n",
       "6042  248cff4b-734e-402d-95ba-aeda3dfd5a78                     3337   \n",
       "6043  434bf795-fa15-4709-872e-fea2f0bf28fa                      860   \n",
       "6044  0e196992-8736-4a23-a766-b6a48ac53f92                      443   \n",
       "6045  3422c128-320c-40a4-8f54-8df22e42117e                     2227   \n",
       "6046  afe0dea5-5835-41ad-a3d5-0111885fadb4                     4887   \n",
       "6047  74af4e06-3343-446c-8d5d-4b0904a192fe                     2957   \n",
       "6048  b734a7f7-1c8f-4dcf-a175-702a74f07473                     3285   \n",
       "6049  7e83c505-957c-43a7-bd2f-8921b8bacede                     2764   \n",
       "6050  9607245c-061a-4f58-8d18-a93d2b3e8cdc                     2592   \n",
       "6051  9a4c0335-3fbf-4b16-ac09-a346b6365978                     1834   \n",
       "6052  f6d6b97e-9204-41ef-9e43-2126f7dd8557                      351   \n",
       "6053  a9dab824-a355-4c7d-bad8-9f46ad5db179                     2552   \n",
       "6054  032c2734-9254-49df-b949-f21606858835                     2109   \n",
       "6055  59353f77-87ee-40a0-930f-328719e2bfe4                     2592   \n",
       "6056  86bf9dba-4728-463d-b767-231f0863ebea                     2595   \n",
       "6057  88013815-813d-4f85-833c-2d5d0f647a94                     3832   \n",
       "6058  62375b9c-54db-4097-95af-4bd20c723822                     2544   \n",
       "6059  7f21fa02-74cf-4d6e-896d-76ebc4773042                     3987   \n",
       "\n",
       "     barcode_arrangement   staxids_nt superkingdom          phylum  \\\n",
       "0              barcode01       200451     Bacteria  Proteobacteria   \n",
       "1              barcode01       200451     Bacteria  Proteobacteria   \n",
       "2              barcode01       192087     Bacteria  Proteobacteria   \n",
       "3              barcode01       200451     Bacteria  Proteobacteria   \n",
       "4              barcode01       192087     Bacteria  Proteobacteria   \n",
       "5              barcode01          321     Bacteria  Proteobacteria   \n",
       "6              barcode01          552     Bacteria  Proteobacteria   \n",
       "7              barcode01  1.28236e+06     Bacteria  Proteobacteria   \n",
       "8              barcode01        48100    Eukaryota      Ascomycota   \n",
       "9              barcode01       634500     Bacteria  Proteobacteria   \n",
       "10             barcode01  1.50567e+06    Eukaryota   Basidiomycota   \n",
       "11             barcode01       192087     Bacteria  Proteobacteria   \n",
       "12             barcode01       199201     Bacteria  Proteobacteria   \n",
       "13             barcode01       889211     Bacteria  Proteobacteria   \n",
       "14             barcode01       192087     Bacteria  Proteobacteria   \n",
       "15             barcode01       192087     Bacteria  Proteobacteria   \n",
       "16             barcode01       587753     Bacteria  Proteobacteria   \n",
       "17             barcode01       665914     Bacteria  Proteobacteria   \n",
       "18             barcode01       192087     Bacteria  Proteobacteria   \n",
       "19             barcode01       199201     Bacteria  Proteobacteria   \n",
       "20             barcode01  1.32493e+06     Bacteria  Proteobacteria   \n",
       "21             barcode01        35703     Bacteria  Proteobacteria   \n",
       "22             barcode01       418459    Eukaryota   Basidiomycota   \n",
       "23             barcode01       192087     Bacteria  Proteobacteria   \n",
       "24             barcode01  1.32493e+06     Bacteria  Proteobacteria   \n",
       "25             barcode01       465817     Bacteria  Proteobacteria   \n",
       "26             barcode01       199201     Bacteria  Proteobacteria   \n",
       "27             barcode01  1.28236e+06     Bacteria  Proteobacteria   \n",
       "28             barcode01       192087     Bacteria  Proteobacteria   \n",
       "29             barcode01  1.32493e+06     Bacteria  Proteobacteria   \n",
       "...                  ...          ...          ...             ...   \n",
       "6030           barcode05        48100    Eukaryota      Ascomycota   \n",
       "6031           barcode05          317     Bacteria  Proteobacteria   \n",
       "6032           barcode05       341663    Eukaryota      Ascomycota   \n",
       "6033           barcode05       398673    Eukaryota      Ascomycota   \n",
       "6034           barcode05        76761     Bacteria  Proteobacteria   \n",
       "6035           barcode05   1.6036e+06    Eukaryota      Arthropoda   \n",
       "6036           barcode05       169297    Eukaryota    Streptophyta   \n",
       "6037           barcode05        48100    Eukaryota      Ascomycota   \n",
       "6038           barcode05        45151    Eukaryota      Ascomycota   \n",
       "6039           barcode05  1.32493e+06     Bacteria  Proteobacteria   \n",
       "6040           barcode05       112498    Eukaryota      Ascomycota   \n",
       "6041           barcode05       671987    Eukaryota      Ascomycota   \n",
       "6042           barcode05        29760    Eukaryota    Streptophyta   \n",
       "6043           barcode05         4565    Eukaryota    Streptophyta   \n",
       "6044           barcode05         7165    Eukaryota      Arthropoda   \n",
       "6045           barcode05        51953    Eukaryota    Streptophyta   \n",
       "6046           barcode05          562     Bacteria  Proteobacteria   \n",
       "6047           barcode05        39946    Eukaryota    Streptophyta   \n",
       "6048           barcode05         5599    Eukaryota      Ascomycota   \n",
       "6049           barcode05       930090    Eukaryota      Ascomycota   \n",
       "6050           barcode05         7955    Eukaryota        Chordata   \n",
       "6051           barcode05         4999    Eukaryota   Basidiomycota   \n",
       "6052           barcode05         7955    Eukaryota        Chordata   \n",
       "6053           barcode05       353152    Eukaryota     Apicomplexa   \n",
       "6054           barcode05       741155     Bacteria  Proteobacteria   \n",
       "6055           barcode05        48100    Eukaryota      Ascomycota   \n",
       "6056           barcode05         5599    Eukaryota      Ascomycota   \n",
       "6057           barcode05       930090    Eukaryota      Ascomycota   \n",
       "6058           barcode05        33069     Bacteria  Proteobacteria   \n",
       "6059           barcode05      1785145     Bacteria  Proteobacteria   \n",
       "\n",
       "                    class             order              family  \\\n",
       "0     Gammaproteobacteria   Pseudomonadales    Pseudomonadaceae   \n",
       "1     Gammaproteobacteria   Pseudomonadales    Pseudomonadaceae   \n",
       "2     Gammaproteobacteria   Pseudomonadales    Pseudomonadaceae   \n",
       "3     Gammaproteobacteria   Pseudomonadales    Pseudomonadaceae   \n",
       "4     Gammaproteobacteria   Pseudomonadales    Pseudomonadaceae   \n",
       "5     Gammaproteobacteria   Pseudomonadales    Pseudomonadaceae   \n",
       "6     Gammaproteobacteria  Enterobacterales         Erwiniaceae   \n",
       "7     Gammaproteobacteria   Pseudomonadales    Pseudomonadaceae   \n",
       "8         Dothideomycetes      Pleosporales       Pleosporaceae   \n",
       "9     Gammaproteobacteria  Enterobacterales         Erwiniaceae   \n",
       "10        Pucciniomycetes       Pucciniales        Pucciniaceae   \n",
       "11    Gammaproteobacteria   Pseudomonadales    Pseudomonadaceae   \n",
       "12    Gammaproteobacteria   Pseudomonadales    Pseudomonadaceae   \n",
       "13    Gammaproteobacteria  Enterobacterales         Erwiniaceae   \n",
       "14    Gammaproteobacteria   Pseudomonadales    Pseudomonadaceae   \n",
       "15    Gammaproteobacteria   Pseudomonadales    Pseudomonadaceae   \n",
       "16    Gammaproteobacteria   Pseudomonadales    Pseudomonadaceae   \n",
       "17    Gammaproteobacteria  Enterobacterales         Erwiniaceae   \n",
       "18    Gammaproteobacteria   Pseudomonadales    Pseudomonadaceae   \n",
       "19    Gammaproteobacteria   Pseudomonadales    Pseudomonadaceae   \n",
       "20    Gammaproteobacteria   Pseudomonadales    Pseudomonadaceae   \n",
       "21    Gammaproteobacteria  Enterobacterales  Enterobacteriaceae   \n",
       "22        Pucciniomycetes       Pucciniales        Pucciniaceae   \n",
       "23    Gammaproteobacteria   Pseudomonadales    Pseudomonadaceae   \n",
       "24    Gammaproteobacteria   Pseudomonadales    Pseudomonadaceae   \n",
       "25    Gammaproteobacteria  Enterobacterales         Erwiniaceae   \n",
       "26    Gammaproteobacteria   Pseudomonadales    Pseudomonadaceae   \n",
       "27    Gammaproteobacteria   Pseudomonadales    Pseudomonadaceae   \n",
       "28    Gammaproteobacteria   Pseudomonadales    Pseudomonadaceae   \n",
       "29    Gammaproteobacteria   Pseudomonadales    Pseudomonadaceae   \n",
       "...                   ...               ...                 ...   \n",
       "6030      Dothideomycetes      Pleosporales       Pleosporaceae   \n",
       "6031  Gammaproteobacteria   Pseudomonadales    Pseudomonadaceae   \n",
       "6032       Eurotiomycetes        Eurotiales      Aspergillaceae   \n",
       "6033      Sordariomycetes       Hypocreales        Hypocreaceae   \n",
       "6034  Gammaproteobacteria   Pseudomonadales    Pseudomonadaceae   \n",
       "6035              Insecta         Hemiptera        Pentatomidae   \n",
       "6036           Liliopsida            Poales             Poaceae   \n",
       "6037      Dothideomycetes      Pleosporales       Pleosporaceae   \n",
       "6038      Dothideomycetes      Pleosporales       Pleosporaceae   \n",
       "6039  Gammaproteobacteria   Pseudomonadales    Pseudomonadaceae   \n",
       "6040      Dothideomycetes       Capnodiales  Mycosphaerellaceae   \n",
       "6041      Dothideomycetes      Pleosporales       Pleosporaceae   \n",
       "6042         Unclassified           Vitales            Vitaceae   \n",
       "6043           Liliopsida            Poales             Poaceae   \n",
       "6044              Insecta           Diptera           Culicidae   \n",
       "6045           Liliopsida          Arecales           Arecaceae   \n",
       "6046  Gammaproteobacteria  Enterobacterales  Enterobacteriaceae   \n",
       "6047           Liliopsida            Poales             Poaceae   \n",
       "6048      Dothideomycetes      Pleosporales       Pleosporaceae   \n",
       "6049      Dothideomycetes      Pleosporales       Pleosporaceae   \n",
       "6050          Actinopteri     Cypriniformes          Cyprinidae   \n",
       "6051      Tremellomycetes       Tremellales   Cuniculitremaceae   \n",
       "6052          Actinopteri     Cypriniformes          Cyprinidae   \n",
       "6053          Conoidasida    Eucoccidiorida   Cryptosporidiidae   \n",
       "6054  Gammaproteobacteria   Pseudomonadales    Pseudomonadaceae   \n",
       "6055      Dothideomycetes      Pleosporales       Pleosporaceae   \n",
       "6056      Dothideomycetes      Pleosporales       Pleosporaceae   \n",
       "6057      Dothideomycetes      Pleosporales       Pleosporaceae   \n",
       "6058  Gammaproteobacteria   Pseudomonadales    Pseudomonadaceae   \n",
       "6059  Gammaproteobacteria   Pseudomonadales    Pseudomonadaceae   \n",
       "\n",
       "                genus                       species  \n",
       "0         Pseudomonas              Pseudomonas poae  \n",
       "1         Pseudomonas              Pseudomonas poae  \n",
       "2         Pseudomonas          Pseudomonas syringae  \n",
       "3         Pseudomonas              Pseudomonas poae  \n",
       "4         Pseudomonas          Pseudomonas syringae  \n",
       "5         Pseudomonas          Pseudomonas syringae  \n",
       "6             Erwinia             Erwinia amylovora  \n",
       "7         Pseudomonas              Pseudomonas poae  \n",
       "8          Alternaria             Alternaria solani  \n",
       "9             Erwinia            Erwinia billingiae  \n",
       "10           Puccinia   Puccinia cf. psidii AE-2014  \n",
       "11        Pseudomonas          Pseudomonas syringae  \n",
       "12        Pseudomonas          Pseudomonas syringae  \n",
       "13            Erwinia             Erwinia amylovora  \n",
       "14        Pseudomonas          Pseudomonas syringae  \n",
       "15        Pseudomonas          Pseudomonas syringae  \n",
       "16        Pseudomonas      Pseudomonas chlororaphis  \n",
       "17            Pantoea              Pantoea gaviniae  \n",
       "18        Pseudomonas          Pseudomonas syringae  \n",
       "19        Pseudomonas          Pseudomonas syringae  \n",
       "20        Pseudomonas          Pseudomonas syringae  \n",
       "21        Citrobacter      Citrobacter amalonaticus  \n",
       "22           Puccinia             Puccinia graminis  \n",
       "23        Pseudomonas          Pseudomonas syringae  \n",
       "24        Pseudomonas          Pseudomonas syringae  \n",
       "25            Erwinia          Erwinia tasmaniensis  \n",
       "26        Pseudomonas          Pseudomonas syringae  \n",
       "27        Pseudomonas              Pseudomonas poae  \n",
       "28        Pseudomonas          Pseudomonas syringae  \n",
       "29        Pseudomonas          Pseudomonas syringae  \n",
       "...               ...                           ...  \n",
       "6030       Alternaria             Alternaria solani  \n",
       "6031      Pseudomonas          Pseudomonas syringae  \n",
       "6032      Aspergillus           Aspergillus terreus  \n",
       "6033      Trichoderma            Trichoderma gamsii  \n",
       "6034      Pseudomonas           Pseudomonas veronii  \n",
       "6035         Cressona                 Cressona rufa  \n",
       "6036         Aegilops             Aegilops tauschii  \n",
       "6037       Alternaria             Alternaria solani  \n",
       "6038      Pyrenophora  Pyrenophora tritici-repentis  \n",
       "6039      Pseudomonas          Pseudomonas syringae  \n",
       "6040        Ramularia         Ramularia collo-cygni  \n",
       "6041      Exserohilum          Exserohilum turcicum  \n",
       "6042            Vitis                Vitis vinifera  \n",
       "6043         Triticum             Triticum aestivum  \n",
       "6044        Anopheles             Anopheles gambiae  \n",
       "6045           Elaeis             Elaeis guineensis  \n",
       "6046      Escherichia              Escherichia coli  \n",
       "6047            Oryza                  Oryza sativa  \n",
       "6048       Alternaria          Alternaria alternata  \n",
       "6049        Bipolaris              Bipolaris oryzae  \n",
       "6050            Danio                   Danio rerio  \n",
       "6051      Kockovaella         Kockovaella imperatae  \n",
       "6052            Danio                   Danio rerio  \n",
       "6053  Cryptosporidium        Cryptosporidium parvum  \n",
       "6054      Pseudomonas       Pseudomonas alcaliphila  \n",
       "6055       Alternaria             Alternaria solani  \n",
       "6056       Alternaria          Alternaria alternata  \n",
       "6057        Bipolaris              Bipolaris oryzae  \n",
       "6058      Pseudomonas       Pseudomonas viridiflava  \n",
       "6059      Pseudomonas         Pseudomonas sp. MS586  \n",
       "\n",
       "[6060 rows x 11 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ntblasthit_reads_filtered_barcodes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# add the flowcell column just in case\n",
    "ntblasthit_reads_filtered_barcodes = ntblasthit_reads_filtered_barcodes.assign(Flowcell = column_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "ntblasthit_reads_filtered_barcodes.to_csv(r'/home/yiheng/analysis/WGS/%s_taxa.tab' % column_name, header=column_name, index=None, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Defines a function to make manual adjustments to the NCBI taxonomic sorting\n",
    "def manually_adjust_ncbi_taxonomy(rank_select, entry_select, ranks_to_edit, new_entry):\n",
    "    \"\"\"Select dataframe entries to edit by a specific rank (rank_select) \n",
    "    and entry related to that rank (entry_select), \n",
    "    provide a list of ranks to edit for those selected entries (ranks_to_edit),\n",
    "    replace these ranks with a new entry (new_entry)\"\"\"\n",
    "    if entry_select in list(ntblasthit_reads_filtered_barcodes[rank_select].unique()):\n",
    "        selection = (ntblasthit_reads_filtered_barcodes[rank_select] == entry_select)\n",
    "        selected_tax_ranks = ranks_to_edit\n",
    "        for ranks in selected_tax_ranks:\n",
    "            ntblasthit_reads_filtered_barcodes[ranks][selection] = new_entry"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "manually_adjust_ncbi_taxonomy('species', \n",
    "                            'Cloning vector lambda TXF97', \n",
    "                            ['superkingdom', 'phylum', 'class', 'order', 'family', 'genus'], \n",
    "                            'Cloning vector lambda TXF97')\n",
    "\n",
    "manually_adjust_ncbi_taxonomy('superkingdom', \n",
    "                            'Viruses', \n",
    "                            ['phylum', 'class'], \n",
    "                            'dsDNA viruses, no RNA stage')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#count pivot table of dataframe with taxonomic columns: \n",
    "#rows - rank names, columns - barcode, sorted by desired class\n",
    "def generate_ncbi_taxonomy_pivot(tax_df, rank, bcs, num):\n",
    "    \"\"\"From tax_df, generate a pivot table listing num rank counts, sorted by bcs\"\"\"\n",
    "    pivot_table = tax_df.pivot_table(values='order', \n",
    "                                            index=rank, \n",
    "                                            columns='barcode_arrangement', \n",
    "                                            aggfunc='count', \n",
    "                                            fill_value=0, \n",
    "                                            margins=True)\n",
    "    pivot_table.columns.name = None\n",
    "    return pivot_table.sort_values(bcs, axis=0, ascending=False).head(n=num)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>barcode01</th>\n",
       "      <th>barcode02</th>\n",
       "      <th>barcode03</th>\n",
       "      <th>barcode04</th>\n",
       "      <th>barcode05</th>\n",
       "      <th>All</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>family</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>All</th>\n",
       "      <td>1942</td>\n",
       "      <td>1855</td>\n",
       "      <td>37</td>\n",
       "      <td>1748</td>\n",
       "      <td>271</td>\n",
       "      <td>5853</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pseudomonadaceae</th>\n",
       "      <td>1470</td>\n",
       "      <td>814</td>\n",
       "      <td>0</td>\n",
       "      <td>1104</td>\n",
       "      <td>58</td>\n",
       "      <td>3446</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pleosporaceae</th>\n",
       "      <td>44</td>\n",
       "      <td>263</td>\n",
       "      <td>1</td>\n",
       "      <td>160</td>\n",
       "      <td>77</td>\n",
       "      <td>545</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sphingomonadaceae</th>\n",
       "      <td>1</td>\n",
       "      <td>161</td>\n",
       "      <td>0</td>\n",
       "      <td>83</td>\n",
       "      <td>1</td>\n",
       "      <td>246</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mycosphaerellaceae</th>\n",
       "      <td>1</td>\n",
       "      <td>103</td>\n",
       "      <td>0</td>\n",
       "      <td>43</td>\n",
       "      <td>2</td>\n",
       "      <td>149</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Hymenobacteraceae</th>\n",
       "      <td>1</td>\n",
       "      <td>82</td>\n",
       "      <td>0</td>\n",
       "      <td>39</td>\n",
       "      <td>0</td>\n",
       "      <td>122</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Enterobacteriaceae</th>\n",
       "      <td>51</td>\n",
       "      <td>56</td>\n",
       "      <td>10</td>\n",
       "      <td>35</td>\n",
       "      <td>5</td>\n",
       "      <td>157</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Microbacteriaceae</th>\n",
       "      <td>3</td>\n",
       "      <td>37</td>\n",
       "      <td>0</td>\n",
       "      <td>31</td>\n",
       "      <td>0</td>\n",
       "      <td>71</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Poaceae</th>\n",
       "      <td>14</td>\n",
       "      <td>8</td>\n",
       "      <td>13</td>\n",
       "      <td>29</td>\n",
       "      <td>41</td>\n",
       "      <td>105</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Leptosphaeriaceae</th>\n",
       "      <td>7</td>\n",
       "      <td>34</td>\n",
       "      <td>1</td>\n",
       "      <td>18</td>\n",
       "      <td>17</td>\n",
       "      <td>77</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Oxalobacteraceae</th>\n",
       "      <td>16</td>\n",
       "      <td>18</td>\n",
       "      <td>0</td>\n",
       "      <td>15</td>\n",
       "      <td>0</td>\n",
       "      <td>49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Erwiniaceae</th>\n",
       "      <td>165</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>185</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Unclassified</th>\n",
       "      <td>6</td>\n",
       "      <td>20</td>\n",
       "      <td>0</td>\n",
       "      <td>10</td>\n",
       "      <td>1</td>\n",
       "      <td>37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cryptococcaceae</th>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>8</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Burkholderiaceae</th>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pucciniaceae</th>\n",
       "      <td>56</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>62</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Nocardiaceae</th>\n",
       "      <td>2</td>\n",
       "      <td>33</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sporidiobolaceae</th>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Paenibacillaceae</th>\n",
       "      <td>1</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sphingobacteriaceae</th>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     barcode01  barcode02  barcode03  barcode04  barcode05  \\\n",
       "family                                                                       \n",
       "All                       1942       1855         37       1748        271   \n",
       "Pseudomonadaceae          1470        814          0       1104         58   \n",
       "Pleosporaceae               44        263          1        160         77   \n",
       "Sphingomonadaceae            1        161          0         83          1   \n",
       "Mycosphaerellaceae           1        103          0         43          2   \n",
       "Hymenobacteraceae            1         82          0         39          0   \n",
       "Enterobacteriaceae          51         56         10         35          5   \n",
       "Microbacteriaceae            3         37          0         31          0   \n",
       "Poaceae                     14          8         13         29         41   \n",
       "Leptosphaeriaceae            7         34          1         18         17   \n",
       "Oxalobacteraceae            16         18          0         15          0   \n",
       "Erwiniaceae                165          7          0         13          0   \n",
       "Unclassified                 6         20          0         10          1   \n",
       "Cryptococcaceae              2          6          0          7          8   \n",
       "Burkholderiaceae             2          5          0          7          1   \n",
       "Pucciniaceae                56          0          0          6          0   \n",
       "Nocardiaceae                 2         33          0          6          0   \n",
       "Sporidiobolaceae             2          1          0          5          3   \n",
       "Paenibacillaceae             1          7          0          5          0   \n",
       "Sphingobacteriaceae          0          2          0          5          0   \n",
       "\n",
       "                      All  \n",
       "family                     \n",
       "All                  5853  \n",
       "Pseudomonadaceae     3446  \n",
       "Pleosporaceae         545  \n",
       "Sphingomonadaceae     246  \n",
       "Mycosphaerellaceae    149  \n",
       "Hymenobacteraceae     122  \n",
       "Enterobacteriaceae    157  \n",
       "Microbacteriaceae      71  \n",
       "Poaceae               105  \n",
       "Leptosphaeriaceae      77  \n",
       "Oxalobacteraceae       49  \n",
       "Erwiniaceae           185  \n",
       "Unclassified           37  \n",
       "Cryptococcaceae        23  \n",
       "Burkholderiaceae       15  \n",
       "Pucciniaceae           62  \n",
       "Nocardiaceae           41  \n",
       "Sporidiobolaceae       11  \n",
       "Paenibacillaceae       13  \n",
       "Sphingobacteriaceae     7  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "generate_ncbi_taxonomy_pivot(ntblasthit_reads_filtered_barcodes, 'family','barcode04',20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# % pivot table of dataframe with taxonomic columns: \n",
    "#rows - rank names, columns - barcode, sorted by desired class\n",
    "def generate_ncbi_taxonomy_pivot_pc(tax_df, rank, bcs, num):\n",
    "    \"\"\"From tax_df, generate a pivot table listing num rank %, sorted ascending by bcs\"\"\"\n",
    "    pivot_table_pc = tax_df.pivot_table(values='order', \n",
    "                                            index=rank, \n",
    "                                            columns='barcode_arrangement', \n",
    "                                            aggfunc='count', \n",
    "                                            fill_value=0, \n",
    "                                            margins=True)\n",
    "    pivot_table_pc.columns.name = None\n",
    "    \n",
    "    for brcs in pivot_table_pc.columns:\n",
    "        pivot_table_pc[str(brcs) + '%'] = pivot_table_pc[brcs].apply(lambda x: (x / pivot_table_pc[brcs].loc['All'])*100)\n",
    "    for brcs in pivot_table_pc.columns:\n",
    "        if '%' not in brcs:\n",
    "            pivot_table_pc.drop(brcs, axis=1, inplace=True)\n",
    "    pivot_table_pc.drop('All', axis=0, inplace=True)\n",
    "\n",
    "    pivot_table_pc2 = pivot_table_pc.sort_values(str(bcs) + '%', axis=0, ascending=False)\n",
    "    return pivot_table_pc2.applymap(lambda x: '%.3f' % x).head(n=num)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>barcode01%</th>\n",
       "      <th>barcode02%</th>\n",
       "      <th>barcode03%</th>\n",
       "      <th>barcode04%</th>\n",
       "      <th>barcode05%</th>\n",
       "      <th>All%</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>genus</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Pseudomonas</th>\n",
       "      <td>75.695</td>\n",
       "      <td>43.881</td>\n",
       "      <td>0.000</td>\n",
       "      <td>63.158</td>\n",
       "      <td>21.402</td>\n",
       "      <td>58.876</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Alternaria</th>\n",
       "      <td>1.442</td>\n",
       "      <td>11.321</td>\n",
       "      <td>2.703</td>\n",
       "      <td>7.037</td>\n",
       "      <td>20.295</td>\n",
       "      <td>7.125</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sphingomonas</th>\n",
       "      <td>0.051</td>\n",
       "      <td>7.925</td>\n",
       "      <td>0.000</td>\n",
       "      <td>4.291</td>\n",
       "      <td>0.369</td>\n",
       "      <td>3.827</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Zymoseptoria</th>\n",
       "      <td>0.051</td>\n",
       "      <td>5.337</td>\n",
       "      <td>0.000</td>\n",
       "      <td>2.346</td>\n",
       "      <td>0.000</td>\n",
       "      <td>2.409</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Hymenobacter</th>\n",
       "      <td>0.051</td>\n",
       "      <td>4.420</td>\n",
       "      <td>0.000</td>\n",
       "      <td>2.231</td>\n",
       "      <td>0.000</td>\n",
       "      <td>2.084</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Escherichia</th>\n",
       "      <td>0.566</td>\n",
       "      <td>2.426</td>\n",
       "      <td>24.324</td>\n",
       "      <td>1.716</td>\n",
       "      <td>1.476</td>\n",
       "      <td>1.691</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bipolaris</th>\n",
       "      <td>0.412</td>\n",
       "      <td>1.456</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1.201</td>\n",
       "      <td>4.059</td>\n",
       "      <td>1.145</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Curtobacterium</th>\n",
       "      <td>0.051</td>\n",
       "      <td>0.755</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1.144</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.598</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Leptosphaeria</th>\n",
       "      <td>0.360</td>\n",
       "      <td>1.833</td>\n",
       "      <td>2.703</td>\n",
       "      <td>1.030</td>\n",
       "      <td>6.273</td>\n",
       "      <td>1.316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Triticum</th>\n",
       "      <td>0.309</td>\n",
       "      <td>0.108</td>\n",
       "      <td>29.730</td>\n",
       "      <td>0.973</td>\n",
       "      <td>14.022</td>\n",
       "      <td>1.264</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Unclassified</th>\n",
       "      <td>0.463</td>\n",
       "      <td>0.485</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.858</td>\n",
       "      <td>0.369</td>\n",
       "      <td>0.581</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pantoea</th>\n",
       "      <td>1.493</td>\n",
       "      <td>0.216</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.744</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.786</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Setosphaeria</th>\n",
       "      <td>0.206</td>\n",
       "      <td>0.431</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.629</td>\n",
       "      <td>1.107</td>\n",
       "      <td>0.444</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Massilia</th>\n",
       "      <td>0.360</td>\n",
       "      <td>0.431</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.458</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.393</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Frondihabitans</th>\n",
       "      <td>0.000</td>\n",
       "      <td>0.270</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.400</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.205</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Aegilops</th>\n",
       "      <td>0.154</td>\n",
       "      <td>0.108</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.400</td>\n",
       "      <td>0.738</td>\n",
       "      <td>0.239</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Puccinia</th>\n",
       "      <td>2.781</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.343</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1.025</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhodococcus</th>\n",
       "      <td>0.103</td>\n",
       "      <td>1.725</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.343</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.683</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Janthinobacterium</th>\n",
       "      <td>0.309</td>\n",
       "      <td>0.377</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.343</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.325</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhodotorula</th>\n",
       "      <td>0.103</td>\n",
       "      <td>0.054</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.286</td>\n",
       "      <td>1.107</td>\n",
       "      <td>0.188</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  barcode01% barcode02% barcode03% barcode04% barcode05%  \\\n",
       "genus                                                                      \n",
       "Pseudomonas           75.695     43.881      0.000     63.158     21.402   \n",
       "Alternaria             1.442     11.321      2.703      7.037     20.295   \n",
       "Sphingomonas           0.051      7.925      0.000      4.291      0.369   \n",
       "Zymoseptoria           0.051      5.337      0.000      2.346      0.000   \n",
       "Hymenobacter           0.051      4.420      0.000      2.231      0.000   \n",
       "Escherichia            0.566      2.426     24.324      1.716      1.476   \n",
       "Bipolaris              0.412      1.456      0.000      1.201      4.059   \n",
       "Curtobacterium         0.051      0.755      0.000      1.144      0.000   \n",
       "Leptosphaeria          0.360      1.833      2.703      1.030      6.273   \n",
       "Triticum               0.309      0.108     29.730      0.973     14.022   \n",
       "Unclassified           0.463      0.485      0.000      0.858      0.369   \n",
       "Pantoea                1.493      0.216      0.000      0.744      0.000   \n",
       "Setosphaeria           0.206      0.431      0.000      0.629      1.107   \n",
       "Massilia               0.360      0.431      0.000      0.458      0.000   \n",
       "Frondihabitans         0.000      0.270      0.000      0.400      0.000   \n",
       "Aegilops               0.154      0.108      0.000      0.400      0.738   \n",
       "Puccinia               2.781      0.000      0.000      0.343      0.000   \n",
       "Rhodococcus            0.103      1.725      0.000      0.343      0.000   \n",
       "Janthinobacterium      0.309      0.377      0.000      0.343      0.000   \n",
       "Rhodotorula            0.103      0.054      0.000      0.286      1.107   \n",
       "\n",
       "                     All%  \n",
       "genus                      \n",
       "Pseudomonas        58.876  \n",
       "Alternaria          7.125  \n",
       "Sphingomonas        3.827  \n",
       "Zymoseptoria        2.409  \n",
       "Hymenobacter        2.084  \n",
       "Escherichia         1.691  \n",
       "Bipolaris           1.145  \n",
       "Curtobacterium      0.598  \n",
       "Leptosphaeria       1.316  \n",
       "Triticum            1.264  \n",
       "Unclassified        0.581  \n",
       "Pantoea             0.786  \n",
       "Setosphaeria        0.444  \n",
       "Massilia            0.393  \n",
       "Frondihabitans      0.205  \n",
       "Aegilops            0.239  \n",
       "Puccinia            1.025  \n",
       "Rhodococcus         0.683  \n",
       "Janthinobacterium   0.325  \n",
       "Rhodotorula         0.188  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "generate_ncbi_taxonomy_pivot_pc(ntblasthit_reads_filtered_barcodes, 'genus', 'barcode04', 20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Assesses number of barcodes, generates necessary number of plots (even) and relevant indexing system\n",
    "ncbiblast_barcodes = list(ntblasthit_reads_filtered_barcodes['barcode_arrangement'].unique())\n",
    "no_of_subplots = len(ncbiblast_barcodes)+1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Always 2 columns, bc-count/2 rows \n",
    "no_of_subplots_pair = [int(no_of_subplots/2), 2]\n",
    "\n",
    "#Produce pairs of indices correlating to the coordinates of the subplots\n",
    "subplot_coordinates = list(product(range(no_of_subplots_pair[0]), range(no_of_subplots_pair[1])))\n",
    "subplot_coordinates_list = [list(l) for l in subplot_coordinates]\n",
    "subplot_coordinates_list_rows = [i[0] for i in subplot_coordinates_list]\n",
    "subplot_coordinates_list_columns = [i[1] for i in subplot_coordinates_list]\n",
    "\n",
    "fig, ax = plt.subplots(no_of_subplots_pair[1], no_of_subplots_pair[0], figsize=(20, 6))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "name_list = ['Stripe rust', 'Septoria tritici blotch', 'Healthy wheat', 'Stripe rust and sptoria tritici blotch', 'Yellow spot']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def get_color(x):\n",
    "    x = x%10\n",
    "    return ('#%02X%02X%02X' % (int(plt.cm.tab10.colors[x][0]*255),int(plt.cm.tab10.colors[x][1]*255),int(plt.cm.tab10.colors[x][2]*255)))\n",
    "\n",
    "def applyTaxPlotStyle(ax_ind1, ax_ind2, bcs, rank, name):\n",
    "    labels = ''\n",
    "    labels = tuple(generate_ncbi_taxonomy_pivot_pc(ntblasthit_reads_filtered_barcodes, rank, bcs, 10)[bcs + '%'].index)\n",
    "    y_pos = ''\n",
    "    y_pos = np.arange(len(labels))\n",
    "    sizes = ''\n",
    "    sizes = tuple([float(i) for i in generate_ncbi_taxonomy_pivot_pc(ntblasthit_reads_filtered_barcodes, rank, bcs, 10)[bcs + '%']])\n",
    "    \n",
    "    ax[ax_ind1, ax_ind2].barh(y_pos, sizes, align='center', alpha=0.4, color=get_color(2*ax_ind2 + ax_ind1))\n",
    "    ax[ax_ind1, ax_ind2].set_yticks(y_pos)\n",
    "    ax[ax_ind1, ax_ind2].set_yticklabels(labels)\n",
    "    ax[ax_ind1, ax_ind2].tick_params(top=\"off\", right=\"off\", left=\"off\")\n",
    "    ax[ax_ind1, ax_ind2].set_xlabel('% Total Reads Hitting NCBI')\n",
    "    ax[ax_ind1, ax_ind2].set_xticks(range(0, 110, 10))\n",
    "    ax[ax_ind1, ax_ind2].set_title(name)\n",
    "    ax[ax_ind1, ax_ind2].invert_yaxis()\n",
    "    ax[ax_ind1, ax_ind2].grid(b=None, which='major', axis='x', linestyle='dashed', linewidth=0.2)\n",
    "    ax[ax_ind1, ax_ind2].annotate('E = 0.00001', \n",
    "             xy=(1.06, 0.3), xycoords='axes fraction', fontsize=10, xytext=(-20, -30), textcoords='offset points', \n",
    "             ha='right', va='bottom')\n",
    "\n",
    "    \n",
    "def generateTaxSubplots(search_rank):\n",
    "    for ax_ind1, ax_ind2, bcs, name in zip(subplot_coordinates_list_columns, subplot_coordinates_list_rows, ncbiblast_barcodes, name_list):\n",
    "        applyTaxPlotStyle(ax_ind1, ax_ind2, bcs, search_rank, name)    \n",
    "\n",
    "    if len(list(ntblasthit_reads_filtered_barcodes['barcode_arrangement'].unique())) != 0:\n",
    "        plt.delaxes(ax[subplot_coordinates_list_columns[-1], subplot_coordinates_list_rows[-1]])\n",
    "    \n",
    "    plt.suptitle('Reads Hitting NCBI Database - % By ' + str(search_rank).title() + ' By Sample', \n",
    "             fontsize='x-large',\n",
    "             y=1.02,\n",
    "             fontweight=\"bold\")\n",
    "    \n",
    "    plt.subplots_adjust(left=None, bottom=None, right=None, top=None, wspace=1.4, hspace=0.4)\n",
    "    plt.tight_layout()\n",
    "    \n",
    "    plt.savefig('/home/yiheng/analysis/WGS/figures' + 'figure3_%s_%s.png' % (column_name, search_rank), bbox_inches='tight', dpi=1000)\n",
    "    plt.show()\n",
    "            \n",
    "            \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(no_of_subplots_pair[1], no_of_subplots_pair[0], figsize=(20, 6))\n",
    "generateTaxSubplots('genus')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "df1=pd.DataFrame([[1, 2],[10, 1], [3, 4], [7, 8]], columns=['A', 'B'])\n",
    "df2=pd.DataFrame([[1, 4],[3, 5]], columns=['A', 'B'])\n",
    "# now replace the location use the index of each df selected from value by .loc function\n",
    "# first select the value from df2 column A\n",
    "for values in df2['A'].unique():\n",
    "    # then locate the position in df1 and replace it by the VALUE of the position in df2\n",
    "    df1.loc[df1[df1['A'] == (values)].index, 'B'] = df2.loc[df2[df2['A'] == (values)].index, 'B'].values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "df1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
