{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This is a notebook to pull down all the information for figure 3 (ncbi hits). The design is to compair between each sample (barcode) instead of each flowcell as they sequenced the same thing. Therefore I just pulled all the ncbi hits for each flowcell for each barcode and merged them together according to the barcode."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import os\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib\n",
    "from itertools import product\n",
    "import argparse\n",
    "from ete3 import NCBITaxa \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "ncbi = NCBITaxa()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# use argparse to do this\n",
    "#BASEDIR = args.BASEDIR\n",
    "BASEDIR = '/home/yiheng/data/20180524_FAH58271'\n",
    "\n",
    "# here we define the folder name of the dataframe it created by capturing the folder from the BASDIR\n",
    "folder_name = os.path.basename(BASEDIR)\n",
    "#column_name = folder_name.split('_')[-1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# first check if the analysis folder is there\n",
    "folder_list = 'analysis  basecalled_data  scripts  tracking  workspace'.split(' ')\n",
    "for x in range(0,folder_list.count('')):\n",
    "    folder_list.remove('')\n",
    "#fix this test\n",
    "if not set(os.listdir(os.path.abspath(BASEDIR))) == set (folder_list):\n",
    "    print(\"Something wrong with basefolder. check it please.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/yiheng/anaconda3/lib/python3.6/site-packages/IPython/core/interactiveshell.py:2717: DtypeWarning: Columns (1,2,3,5,16) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  interactivity=interactivity, compiler=compiler, result=result)\n"
     ]
    }
   ],
   "source": [
    "# get the dataframe there\n",
    "dataframe = os.path.join(BASEDIR, 'analysis', 'summary_df_%s.tab' % folder_name)\n",
    "sum_df = pd.read_csv(dataframe, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>read_id</th>\n",
       "      <th>passes_filtering</th>\n",
       "      <th>sequence_length_template</th>\n",
       "      <th>mean_qscore_template</th>\n",
       "      <th>barcode_arrangement</th>\n",
       "      <th>barcode_score</th>\n",
       "      <th>kit</th>\n",
       "      <th>variant</th>\n",
       "      <th>pc_survived</th>\n",
       "      <th>qseqid_nt</th>\n",
       "      <th>sseqid_nt</th>\n",
       "      <th>evalue_nt</th>\n",
       "      <th>length_nt</th>\n",
       "      <th>pident_nt</th>\n",
       "      <th>nident_nt</th>\n",
       "      <th>sacc_nt</th>\n",
       "      <th>staxids_nt</th>\n",
       "      <th>scomnames_nt</th>\n",
       "      <th>read_length_pc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3de506ef-9b26-4aee-b11e-bf8df5a65299</td>\n",
       "      <td>True</td>\n",
       "      <td>503</td>\n",
       "      <td>10.481</td>\n",
       "      <td>barcode07</td>\n",
       "      <td>84.0513</td>\n",
       "      <td>NB</td>\n",
       "      <td>var1</td>\n",
       "      <td>True</td>\n",
       "      <td>3de506ef-9b26-4aee-b11e-bf8df5a65299</td>\n",
       "      <td>gi|16605664|emb|AL358232.19|</td>\n",
       "      <td>3.000000e-152</td>\n",
       "      <td>430.0</td>\n",
       "      <td>90.70</td>\n",
       "      <td>390.0</td>\n",
       "      <td>AL358232</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>434.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>43a2ceb3-ea58-45c4-a9c0-9e845a3ee3f5</td>\n",
       "      <td>True</td>\n",
       "      <td>425</td>\n",
       "      <td>9.72</td>\n",
       "      <td>barcode07</td>\n",
       "      <td>90.8718</td>\n",
       "      <td>NB</td>\n",
       "      <td>var1</td>\n",
       "      <td>True</td>\n",
       "      <td>43a2ceb3-ea58-45c4-a9c0-9e845a3ee3f5</td>\n",
       "      <td>gi|18497210|gb|AC097491.2|</td>\n",
       "      <td>3.000000e-96</td>\n",
       "      <td>352.0</td>\n",
       "      <td>86.65</td>\n",
       "      <td>305.0</td>\n",
       "      <td>AC097491</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>339.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>45768783-cf86-4203-a864-57c1be00e671</td>\n",
       "      <td>True</td>\n",
       "      <td>726</td>\n",
       "      <td>8.658</td>\n",
       "      <td>barcode07</td>\n",
       "      <td>90.8718</td>\n",
       "      <td>NB</td>\n",
       "      <td>var1</td>\n",
       "      <td>True</td>\n",
       "      <td>45768783-cf86-4203-a864-57c1be00e671</td>\n",
       "      <td>gi|1871197|gb|U91318.1|HUU91318</td>\n",
       "      <td>7.000000e-165</td>\n",
       "      <td>429.0</td>\n",
       "      <td>92.07</td>\n",
       "      <td>395.0</td>\n",
       "      <td>U91318</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>654.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>576ecd5e-1a24-42af-97fd-f0d5fc3a4dde</td>\n",
       "      <td>True</td>\n",
       "      <td>831</td>\n",
       "      <td>9.943</td>\n",
       "      <td>barcode07</td>\n",
       "      <td>91.2308</td>\n",
       "      <td>NB</td>\n",
       "      <td>var1</td>\n",
       "      <td>True</td>\n",
       "      <td>576ecd5e-1a24-42af-97fd-f0d5fc3a4dde</td>\n",
       "      <td>gi|20377037|gb|AC109631.4|</td>\n",
       "      <td>6.000000e-166</td>\n",
       "      <td>460.0</td>\n",
       "      <td>91.09</td>\n",
       "      <td>419.0</td>\n",
       "      <td>AC109631</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>762.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>5c0dd0f2-884a-4d7d-857c-62b50825adc2</td>\n",
       "      <td>True</td>\n",
       "      <td>464</td>\n",
       "      <td>8.324</td>\n",
       "      <td>barcode07</td>\n",
       "      <td>79.4359</td>\n",
       "      <td>NB</td>\n",
       "      <td>var1</td>\n",
       "      <td>True</td>\n",
       "      <td>5c0dd0f2-884a-4d7d-857c-62b50825adc2</td>\n",
       "      <td>gi|13235051|emb|AL160237.4|</td>\n",
       "      <td>7.000000e-103</td>\n",
       "      <td>234.0</td>\n",
       "      <td>96.58</td>\n",
       "      <td>226.0</td>\n",
       "      <td>AL160237</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>393.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>602c44b6-f466-4dc7-8a3a-d763a30da22e</td>\n",
       "      <td>True</td>\n",
       "      <td>829</td>\n",
       "      <td>8.201</td>\n",
       "      <td>barcode07</td>\n",
       "      <td>65.3846</td>\n",
       "      <td>NB</td>\n",
       "      <td>var1</td>\n",
       "      <td>True</td>\n",
       "      <td>602c44b6-f466-4dc7-8a3a-d763a30da22e</td>\n",
       "      <td>gi|15706113|gb|AC022176.6|</td>\n",
       "      <td>3.000000e-154</td>\n",
       "      <td>402.0</td>\n",
       "      <td>92.29</td>\n",
       "      <td>371.0</td>\n",
       "      <td>AC022176</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>829.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>a8f7ec39-022a-4632-b4f4-6529cbf39074</td>\n",
       "      <td>True</td>\n",
       "      <td>761</td>\n",
       "      <td>9.036</td>\n",
       "      <td>barcode07</td>\n",
       "      <td>98.3077</td>\n",
       "      <td>NB</td>\n",
       "      <td>var1</td>\n",
       "      <td>True</td>\n",
       "      <td>a8f7ec39-022a-4632-b4f4-6529cbf39074</td>\n",
       "      <td>gi|1002623499|ref|NG_011975.2|</td>\n",
       "      <td>2.000000e-146</td>\n",
       "      <td>388.0</td>\n",
       "      <td>92.27</td>\n",
       "      <td>358.0</td>\n",
       "      <td>NG_011975</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>685.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ba58174b-4528-4f2a-a619-d67e66d37d67</td>\n",
       "      <td>True</td>\n",
       "      <td>990</td>\n",
       "      <td>9.436</td>\n",
       "      <td>barcode07</td>\n",
       "      <td>76.8718</td>\n",
       "      <td>NB</td>\n",
       "      <td>var2</td>\n",
       "      <td>True</td>\n",
       "      <td>ba58174b-4528-4f2a-a619-d67e66d37d67</td>\n",
       "      <td>gi|28173291|gb|AC114815.5|</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>913.0</td>\n",
       "      <td>86.09</td>\n",
       "      <td>786.0</td>\n",
       "      <td>AC114815</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>869.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>bac39b1c-06c2-4b4d-b66d-7f1bf58832df</td>\n",
       "      <td>True</td>\n",
       "      <td>835</td>\n",
       "      <td>8.581</td>\n",
       "      <td>barcode07</td>\n",
       "      <td>76.1026</td>\n",
       "      <td>NB</td>\n",
       "      <td>var1</td>\n",
       "      <td>True</td>\n",
       "      <td>bac39b1c-06c2-4b4d-b66d-7f1bf58832df</td>\n",
       "      <td>gi|117414195|gb|AC090616.13|</td>\n",
       "      <td>6.000000e-166</td>\n",
       "      <td>445.0</td>\n",
       "      <td>91.91</td>\n",
       "      <td>409.0</td>\n",
       "      <td>AC090616</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>764.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>c7ba1d1d-80dd-4b90-bf3a-6d866b79b072</td>\n",
       "      <td>True</td>\n",
       "      <td>660</td>\n",
       "      <td>9.043</td>\n",
       "      <td>barcode07</td>\n",
       "      <td>91.2308</td>\n",
       "      <td>NB</td>\n",
       "      <td>var1</td>\n",
       "      <td>True</td>\n",
       "      <td>c7ba1d1d-80dd-4b90-bf3a-6d866b79b072</td>\n",
       "      <td>gi|2052037|emb|Z94801.1|</td>\n",
       "      <td>6.000000e-155</td>\n",
       "      <td>558.0</td>\n",
       "      <td>86.38</td>\n",
       "      <td>482.0</td>\n",
       "      <td>Z94801</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>592.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>f1c062a6-7f6e-4637-8137-9108f3f2b1e2</td>\n",
       "      <td>True</td>\n",
       "      <td>704</td>\n",
       "      <td>8.643</td>\n",
       "      <td>barcode07</td>\n",
       "      <td>98.3077</td>\n",
       "      <td>NB</td>\n",
       "      <td>var1</td>\n",
       "      <td>True</td>\n",
       "      <td>f1c062a6-7f6e-4637-8137-9108f3f2b1e2</td>\n",
       "      <td>gi|14031071|emb|AL136160.18|</td>\n",
       "      <td>7.000000e-100</td>\n",
       "      <td>369.0</td>\n",
       "      <td>86.45</td>\n",
       "      <td>319.0</td>\n",
       "      <td>AL136160</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>615.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>f36087d3-636a-4036-b313-c0f2980ba534</td>\n",
       "      <td>True</td>\n",
       "      <td>524</td>\n",
       "      <td>9.846</td>\n",
       "      <td>barcode07</td>\n",
       "      <td>78.8205</td>\n",
       "      <td>NB</td>\n",
       "      <td>var1</td>\n",
       "      <td>True</td>\n",
       "      <td>f36087d3-636a-4036-b313-c0f2980ba534</td>\n",
       "      <td>gi|1034651703|ref|XM_011536146.2|</td>\n",
       "      <td>5.000000e-150</td>\n",
       "      <td>464.0</td>\n",
       "      <td>89.01</td>\n",
       "      <td>413.0</td>\n",
       "      <td>XM_011536146</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>448.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>00ddb787-1b87-4e47-8d86-52d826aca670</td>\n",
       "      <td>True</td>\n",
       "      <td>863</td>\n",
       "      <td>9.655</td>\n",
       "      <td>barcode07</td>\n",
       "      <td>98.6154</td>\n",
       "      <td>NB</td>\n",
       "      <td>var2</td>\n",
       "      <td>True</td>\n",
       "      <td>00ddb787-1b87-4e47-8d86-52d826aca670</td>\n",
       "      <td>gi|24371354|gb|AC126174.9|</td>\n",
       "      <td>2.000000e-145</td>\n",
       "      <td>518.0</td>\n",
       "      <td>86.49</td>\n",
       "      <td>448.0</td>\n",
       "      <td>AC126174</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>725.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>0480b709-7f84-4453-8777-1663eac5e09d</td>\n",
       "      <td>False</td>\n",
       "      <td>731</td>\n",
       "      <td>6.24</td>\n",
       "      <td>barcode07</td>\n",
       "      <td>67.9487</td>\n",
       "      <td>NB</td>\n",
       "      <td>var1</td>\n",
       "      <td>True</td>\n",
       "      <td>0480b709-7f84-4453-8777-1663eac5e09d</td>\n",
       "      <td>gi|20806260|gb|AC092896.9|</td>\n",
       "      <td>4.000000e-48</td>\n",
       "      <td>368.0</td>\n",
       "      <td>77.99</td>\n",
       "      <td>287.0</td>\n",
       "      <td>AC092896</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>662.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>0a888311-4b8f-45dd-877c-90fa55a7d249</td>\n",
       "      <td>True</td>\n",
       "      <td>555</td>\n",
       "      <td>7.279</td>\n",
       "      <td>barcode07</td>\n",
       "      <td>75.0769</td>\n",
       "      <td>NB</td>\n",
       "      <td>var1</td>\n",
       "      <td>True</td>\n",
       "      <td>0a888311-4b8f-45dd-877c-90fa55a7d249</td>\n",
       "      <td>gi|21240688|gb|AC063977.6|</td>\n",
       "      <td>9.000000e-73</td>\n",
       "      <td>283.0</td>\n",
       "      <td>86.22</td>\n",
       "      <td>244.0</td>\n",
       "      <td>AC063977</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>485.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>0d97acf8-bc81-42de-9804-66c471b2a199</td>\n",
       "      <td>True</td>\n",
       "      <td>649</td>\n",
       "      <td>8.726</td>\n",
       "      <td>barcode07</td>\n",
       "      <td>90.7179</td>\n",
       "      <td>NB</td>\n",
       "      <td>var1</td>\n",
       "      <td>True</td>\n",
       "      <td>0d97acf8-bc81-42de-9804-66c471b2a199</td>\n",
       "      <td>gi|19807715|gb|AC063965.8|</td>\n",
       "      <td>5.000000e-121</td>\n",
       "      <td>341.0</td>\n",
       "      <td>91.20</td>\n",
       "      <td>311.0</td>\n",
       "      <td>AC063965</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>577.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>2071f5ef-b1a7-4d71-b559-1c45fe011fae</td>\n",
       "      <td>True</td>\n",
       "      <td>691</td>\n",
       "      <td>8.385</td>\n",
       "      <td>barcode07</td>\n",
       "      <td>83.2821</td>\n",
       "      <td>NB</td>\n",
       "      <td>var1</td>\n",
       "      <td>True</td>\n",
       "      <td>2071f5ef-b1a7-4d71-b559-1c45fe011fae</td>\n",
       "      <td>gi|13872265|emb|AL158822.19|</td>\n",
       "      <td>8.000000e-134</td>\n",
       "      <td>368.0</td>\n",
       "      <td>91.58</td>\n",
       "      <td>337.0</td>\n",
       "      <td>AL158822</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>606.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>70c6d2c5-5d19-403e-a790-c1c8c8fe9369</td>\n",
       "      <td>True</td>\n",
       "      <td>929</td>\n",
       "      <td>10.422</td>\n",
       "      <td>barcode07</td>\n",
       "      <td>91.2308</td>\n",
       "      <td>NB</td>\n",
       "      <td>var1</td>\n",
       "      <td>True</td>\n",
       "      <td>70c6d2c5-5d19-403e-a790-c1c8c8fe9369</td>\n",
       "      <td>gi|26050975|gb|AC112694.12|</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>709.0</td>\n",
       "      <td>92.52</td>\n",
       "      <td>656.0</td>\n",
       "      <td>AC112694</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>859.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>7f227f67-411f-4fdf-943a-28fb8902fcd5</td>\n",
       "      <td>True</td>\n",
       "      <td>589</td>\n",
       "      <td>8.195</td>\n",
       "      <td>barcode07</td>\n",
       "      <td>94.2051</td>\n",
       "      <td>NB</td>\n",
       "      <td>var1</td>\n",
       "      <td>True</td>\n",
       "      <td>7f227f67-411f-4fdf-943a-28fb8902fcd5</td>\n",
       "      <td>gi|7381629|gb|AC022078.12|AC022078</td>\n",
       "      <td>6.000000e-100</td>\n",
       "      <td>293.0</td>\n",
       "      <td>90.44</td>\n",
       "      <td>265.0</td>\n",
       "      <td>AC022078</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>516.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>8b7e2a43-32ae-419c-9326-2cb14eaf1535</td>\n",
       "      <td>True</td>\n",
       "      <td>569</td>\n",
       "      <td>9.69</td>\n",
       "      <td>unclassified</td>\n",
       "      <td>36.4</td>\n",
       "      <td>NB</td>\n",
       "      <td>var1</td>\n",
       "      <td>True</td>\n",
       "      <td>8b7e2a43-32ae-419c-9326-2cb14eaf1535</td>\n",
       "      <td>gi|380353348|emb|HE681722.1|</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>511.0</td>\n",
       "      <td>90.02</td>\n",
       "      <td>460.0</td>\n",
       "      <td>HE681722</td>\n",
       "      <td>1136231</td>\n",
       "      <td>Candida orthopsilosis Co 90-125</td>\n",
       "      <td>569.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>bdacc264-db52-4d72-8d6b-3a256c26f576</td>\n",
       "      <td>True</td>\n",
       "      <td>785</td>\n",
       "      <td>9.98</td>\n",
       "      <td>barcode07</td>\n",
       "      <td>90.2051</td>\n",
       "      <td>NB</td>\n",
       "      <td>var1</td>\n",
       "      <td>True</td>\n",
       "      <td>bdacc264-db52-4d72-8d6b-3a256c26f576</td>\n",
       "      <td>gi|337621|gb|M77219.1|HUMRSCME</td>\n",
       "      <td>7.000000e-16</td>\n",
       "      <td>88.0</td>\n",
       "      <td>87.50</td>\n",
       "      <td>77.0</td>\n",
       "      <td>M77219</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>714.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>c425bc37-e969-41cf-9904-3b70790331dc</td>\n",
       "      <td>True</td>\n",
       "      <td>512</td>\n",
       "      <td>9.742</td>\n",
       "      <td>barcode07</td>\n",
       "      <td>91.2308</td>\n",
       "      <td>NB</td>\n",
       "      <td>var1</td>\n",
       "      <td>True</td>\n",
       "      <td>c425bc37-e969-41cf-9904-3b70790331dc</td>\n",
       "      <td>gi|15131484|emb|AL391476.20|</td>\n",
       "      <td>2.000000e-74</td>\n",
       "      <td>220.0</td>\n",
       "      <td>91.36</td>\n",
       "      <td>201.0</td>\n",
       "      <td>AL391476</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>391.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>cb3af2d3-7fe3-4bac-8565-0ae34344af18</td>\n",
       "      <td>True</td>\n",
       "      <td>982</td>\n",
       "      <td>8.334</td>\n",
       "      <td>barcode06</td>\n",
       "      <td>94.4615</td>\n",
       "      <td>NB</td>\n",
       "      <td>var1</td>\n",
       "      <td>True</td>\n",
       "      <td>cb3af2d3-7fe3-4bac-8565-0ae34344af18</td>\n",
       "      <td>gi|354543245|emb|HE605203.1|</td>\n",
       "      <td>6.000000e-167</td>\n",
       "      <td>501.0</td>\n",
       "      <td>89.22</td>\n",
       "      <td>447.0</td>\n",
       "      <td>HE605203</td>\n",
       "      <td>5480</td>\n",
       "      <td>Candida parapsilosis</td>\n",
       "      <td>904.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>d1f8a29b-fcc9-4ff8-812d-9b4cedf53116</td>\n",
       "      <td>True</td>\n",
       "      <td>631</td>\n",
       "      <td>9.865</td>\n",
       "      <td>barcode05</td>\n",
       "      <td>86.9231</td>\n",
       "      <td>NB</td>\n",
       "      <td>var1</td>\n",
       "      <td>True</td>\n",
       "      <td>d1f8a29b-fcc9-4ff8-812d-9b4cedf53116</td>\n",
       "      <td>gi|1069598653|ref|XM_018323264.1|</td>\n",
       "      <td>6.000000e-105</td>\n",
       "      <td>383.0</td>\n",
       "      <td>86.42</td>\n",
       "      <td>331.0</td>\n",
       "      <td>XM_018323264</td>\n",
       "      <td>33203</td>\n",
       "      <td>Purpureocillium lilacinum</td>\n",
       "      <td>560.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>25168a41-18d3-4cf5-b2b9-102182e97027</td>\n",
       "      <td>True</td>\n",
       "      <td>944</td>\n",
       "      <td>10.339</td>\n",
       "      <td>barcode07</td>\n",
       "      <td>98.3077</td>\n",
       "      <td>NB</td>\n",
       "      <td>var1</td>\n",
       "      <td>True</td>\n",
       "      <td>25168a41-18d3-4cf5-b2b9-102182e97027</td>\n",
       "      <td>gi|16517072|emb|AL590662.5|</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>574.0</td>\n",
       "      <td>91.99</td>\n",
       "      <td>528.0</td>\n",
       "      <td>AL590662</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>870.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>2a5e0b04-70e8-41ba-9a47-a93f77d0914d</td>\n",
       "      <td>True</td>\n",
       "      <td>718</td>\n",
       "      <td>7.906</td>\n",
       "      <td>barcode07</td>\n",
       "      <td>90.7179</td>\n",
       "      <td>NB</td>\n",
       "      <td>var1</td>\n",
       "      <td>True</td>\n",
       "      <td>2a5e0b04-70e8-41ba-9a47-a93f77d0914d</td>\n",
       "      <td>gi|2181872|emb|Z84497.1|</td>\n",
       "      <td>5.000000e-96</td>\n",
       "      <td>338.0</td>\n",
       "      <td>86.98</td>\n",
       "      <td>294.0</td>\n",
       "      <td>Z84497</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>624.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>2f53ad37-e3cf-49cd-804e-7064d603da3d</td>\n",
       "      <td>True</td>\n",
       "      <td>1564</td>\n",
       "      <td>9.063</td>\n",
       "      <td>barcode07</td>\n",
       "      <td>90.7179</td>\n",
       "      <td>NB</td>\n",
       "      <td>var1</td>\n",
       "      <td>True</td>\n",
       "      <td>2f53ad37-e3cf-49cd-804e-7064d603da3d</td>\n",
       "      <td>gi|8217573|emb|AL139350.17|</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>988.0</td>\n",
       "      <td>88.87</td>\n",
       "      <td>878.0</td>\n",
       "      <td>AL139350</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>1484.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>45803d09-ef84-4636-8546-0d0cbada3bc6</td>\n",
       "      <td>True</td>\n",
       "      <td>663</td>\n",
       "      <td>9.031</td>\n",
       "      <td>barcode07</td>\n",
       "      <td>87.3846</td>\n",
       "      <td>NB</td>\n",
       "      <td>var1</td>\n",
       "      <td>True</td>\n",
       "      <td>45803d09-ef84-4636-8546-0d0cbada3bc6</td>\n",
       "      <td>gi|13751498|emb|AL451103.7|</td>\n",
       "      <td>8.000000e-119</td>\n",
       "      <td>350.0</td>\n",
       "      <td>90.29</td>\n",
       "      <td>316.0</td>\n",
       "      <td>AL451103</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>592.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>5255afda-4298-4094-a546-209dc170b9b2</td>\n",
       "      <td>True</td>\n",
       "      <td>1142</td>\n",
       "      <td>9.975</td>\n",
       "      <td>barcode07</td>\n",
       "      <td>65.4872</td>\n",
       "      <td>NB</td>\n",
       "      <td>var1</td>\n",
       "      <td>True</td>\n",
       "      <td>5255afda-4298-4094-a546-209dc170b9b2</td>\n",
       "      <td>gi|20270170|gb|AC099782.2|</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>865.0</td>\n",
       "      <td>87.98</td>\n",
       "      <td>761.0</td>\n",
       "      <td>AC099782</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>1077.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>5dc9fed9-1c13-4e46-a86c-b316992f9a05</td>\n",
       "      <td>True</td>\n",
       "      <td>872</td>\n",
       "      <td>9.116</td>\n",
       "      <td>barcode07</td>\n",
       "      <td>72.6667</td>\n",
       "      <td>NB</td>\n",
       "      <td>var1</td>\n",
       "      <td>True</td>\n",
       "      <td>5dc9fed9-1c13-4e46-a86c-b316992f9a05</td>\n",
       "      <td>gi|19807899|gb|AC110769.2|</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>520.0</td>\n",
       "      <td>91.92</td>\n",
       "      <td>478.0</td>\n",
       "      <td>AC110769</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>802.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4752682</th>\n",
       "      <td>4d151404-0ec1-495b-ab62-3b7f79d9eef8</td>\n",
       "      <td>True</td>\n",
       "      <td>1470</td>\n",
       "      <td>10.624</td>\n",
       "      <td>barcode06</td>\n",
       "      <td>98.359</td>\n",
       "      <td>NB</td>\n",
       "      <td>var2</td>\n",
       "      <td>True</td>\n",
       "      <td>4d151404-0ec1-495b-ab62-3b7f79d9eef8</td>\n",
       "      <td>gi|70991147|ref|XM_745330.1|</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>710.0</td>\n",
       "      <td>88.59</td>\n",
       "      <td>629.0</td>\n",
       "      <td>XM_745330</td>\n",
       "      <td>330879</td>\n",
       "      <td>Aspergillus fumigatus Af293</td>\n",
       "      <td>1319.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4752683</th>\n",
       "      <td>6da6eb99-3922-446f-89cd-a9044f7a088d</td>\n",
       "      <td>True</td>\n",
       "      <td>2246</td>\n",
       "      <td>8.759</td>\n",
       "      <td>barcode06</td>\n",
       "      <td>64.4615</td>\n",
       "      <td>NB</td>\n",
       "      <td>var1</td>\n",
       "      <td>True</td>\n",
       "      <td>6da6eb99-3922-446f-89cd-a9044f7a088d</td>\n",
       "      <td>gi|48526539|gb|AY560609.1|</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1418.0</td>\n",
       "      <td>82.30</td>\n",
       "      <td>1167.0</td>\n",
       "      <td>AY560609</td>\n",
       "      <td>40410</td>\n",
       "      <td>Cryptococcus neoformans var. neoformans</td>\n",
       "      <td>2229.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4752686</th>\n",
       "      <td>92be42d8-83a6-4004-98c5-63f036a17ceb</td>\n",
       "      <td>True</td>\n",
       "      <td>6373</td>\n",
       "      <td>9.123</td>\n",
       "      <td>barcode05</td>\n",
       "      <td>71.2308</td>\n",
       "      <td>NB</td>\n",
       "      <td>var1</td>\n",
       "      <td>True</td>\n",
       "      <td>92be42d8-83a6-4004-98c5-63f036a17ceb</td>\n",
       "      <td>gi|532597203|gb|KF015766.1|</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>643.0</td>\n",
       "      <td>86.31</td>\n",
       "      <td>555.0</td>\n",
       "      <td>KF015766</td>\n",
       "      <td>36911</td>\n",
       "      <td>Clavispora lusitaniae</td>\n",
       "      <td>6295.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4752688</th>\n",
       "      <td>13eedd39-7d1a-429d-9f4d-be8c57af7e59</td>\n",
       "      <td>True</td>\n",
       "      <td>1526</td>\n",
       "      <td>10.257</td>\n",
       "      <td>barcode05</td>\n",
       "      <td>94.5641</td>\n",
       "      <td>NB</td>\n",
       "      <td>var1</td>\n",
       "      <td>True</td>\n",
       "      <td>13eedd39-7d1a-429d-9f4d-be8c57af7e59</td>\n",
       "      <td>gi|1069579669|ref|XM_018318480.1|</td>\n",
       "      <td>2.000000e-128</td>\n",
       "      <td>353.0</td>\n",
       "      <td>91.78</td>\n",
       "      <td>324.0</td>\n",
       "      <td>XM_018318480</td>\n",
       "      <td>33203</td>\n",
       "      <td>Purpureocillium lilacinum</td>\n",
       "      <td>1390.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4752691</th>\n",
       "      <td>379d3855-07f1-4407-9374-a27e93e790fa</td>\n",
       "      <td>True</td>\n",
       "      <td>1209</td>\n",
       "      <td>8.351</td>\n",
       "      <td>barcode07</td>\n",
       "      <td>86.8205</td>\n",
       "      <td>NB</td>\n",
       "      <td>var2</td>\n",
       "      <td>True</td>\n",
       "      <td>379d3855-07f1-4407-9374-a27e93e790fa</td>\n",
       "      <td>gi|17149768|gb|AC023950.6|</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1083.0</td>\n",
       "      <td>85.13</td>\n",
       "      <td>922.0</td>\n",
       "      <td>AC023950</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>1077.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4752692</th>\n",
       "      <td>4e556279-987d-4861-aacd-f25777056fbf</td>\n",
       "      <td>True</td>\n",
       "      <td>5092</td>\n",
       "      <td>8.933</td>\n",
       "      <td>barcode06</td>\n",
       "      <td>90.1538</td>\n",
       "      <td>NB</td>\n",
       "      <td>var2</td>\n",
       "      <td>True</td>\n",
       "      <td>4e556279-987d-4861-aacd-f25777056fbf</td>\n",
       "      <td>gi|196049150|emb|CR380956.2|</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>5219.0</td>\n",
       "      <td>83.81</td>\n",
       "      <td>4374.0</td>\n",
       "      <td>CR380956</td>\n",
       "      <td>5478</td>\n",
       "      <td>Candida glabrata</td>\n",
       "      <td>4964.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4752693</th>\n",
       "      <td>4fff9153-b942-4b45-a53d-111b21e7e04d</td>\n",
       "      <td>True</td>\n",
       "      <td>743</td>\n",
       "      <td>9.12</td>\n",
       "      <td>barcode07</td>\n",
       "      <td>91.2308</td>\n",
       "      <td>NB</td>\n",
       "      <td>var1</td>\n",
       "      <td>True</td>\n",
       "      <td>4fff9153-b942-4b45-a53d-111b21e7e04d</td>\n",
       "      <td>gi|16415916|emb|AL627245.1|</td>\n",
       "      <td>7.000000e-140</td>\n",
       "      <td>388.0</td>\n",
       "      <td>91.24</td>\n",
       "      <td>354.0</td>\n",
       "      <td>AL627245</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>672.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4752694</th>\n",
       "      <td>5ce685aa-a67f-46ad-87bf-b87d7e73417c</td>\n",
       "      <td>True</td>\n",
       "      <td>678</td>\n",
       "      <td>8.33</td>\n",
       "      <td>barcode06</td>\n",
       "      <td>80.4103</td>\n",
       "      <td>NB</td>\n",
       "      <td>var1</td>\n",
       "      <td>True</td>\n",
       "      <td>5ce685aa-a67f-46ad-87bf-b87d7e73417c</td>\n",
       "      <td>gi|169772144|ref|XM_001820490.1|</td>\n",
       "      <td>8.000000e-114</td>\n",
       "      <td>574.0</td>\n",
       "      <td>81.71</td>\n",
       "      <td>469.0</td>\n",
       "      <td>XM_001820490</td>\n",
       "      <td>510516</td>\n",
       "      <td>Aspergillus oryzae RIB40</td>\n",
       "      <td>548.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4752695</th>\n",
       "      <td>793c1235-69c1-4aea-80aa-bd24acb664a7</td>\n",
       "      <td>True</td>\n",
       "      <td>525</td>\n",
       "      <td>8.402</td>\n",
       "      <td>barcode06</td>\n",
       "      <td>94.4615</td>\n",
       "      <td>NB</td>\n",
       "      <td>var1</td>\n",
       "      <td>True</td>\n",
       "      <td>793c1235-69c1-4aea-80aa-bd24acb664a7</td>\n",
       "      <td>gi|70994853|ref|XM_747111.1|</td>\n",
       "      <td>7.000000e-84</td>\n",
       "      <td>266.0</td>\n",
       "      <td>89.47</td>\n",
       "      <td>238.0</td>\n",
       "      <td>XM_747111</td>\n",
       "      <td>330879</td>\n",
       "      <td>Aspergillus fumigatus Af293</td>\n",
       "      <td>455.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4752698</th>\n",
       "      <td>9da0e9bc-6912-46e8-8918-24aa8c1b5257</td>\n",
       "      <td>True</td>\n",
       "      <td>1187</td>\n",
       "      <td>8.779</td>\n",
       "      <td>barcode05</td>\n",
       "      <td>91.1795</td>\n",
       "      <td>NB</td>\n",
       "      <td>var2</td>\n",
       "      <td>True</td>\n",
       "      <td>9da0e9bc-6912-46e8-8918-24aa8c1b5257</td>\n",
       "      <td>gi|1069628914|ref|XM_018328678.1|</td>\n",
       "      <td>7.000000e-132</td>\n",
       "      <td>585.0</td>\n",
       "      <td>83.59</td>\n",
       "      <td>489.0</td>\n",
       "      <td>XM_018328678</td>\n",
       "      <td>33203</td>\n",
       "      <td>Purpureocillium lilacinum</td>\n",
       "      <td>1089.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4752699</th>\n",
       "      <td>b7a2a56f-c7ce-4258-93b0-99485433cbf6</td>\n",
       "      <td>True</td>\n",
       "      <td>753</td>\n",
       "      <td>9.501</td>\n",
       "      <td>barcode05</td>\n",
       "      <td>67.1795</td>\n",
       "      <td>NB</td>\n",
       "      <td>var1</td>\n",
       "      <td>True</td>\n",
       "      <td>b7a2a56f-c7ce-4258-93b0-99485433cbf6</td>\n",
       "      <td>gi|1069577477|ref|XM_018317384.1|</td>\n",
       "      <td>6.000000e-56</td>\n",
       "      <td>222.0</td>\n",
       "      <td>86.94</td>\n",
       "      <td>193.0</td>\n",
       "      <td>XM_018317384</td>\n",
       "      <td>33203</td>\n",
       "      <td>Purpureocillium lilacinum</td>\n",
       "      <td>679.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4752700</th>\n",
       "      <td>c381f362-c75a-4591-8943-f2396b469c8e</td>\n",
       "      <td>True</td>\n",
       "      <td>2802</td>\n",
       "      <td>8.542</td>\n",
       "      <td>barcode06</td>\n",
       "      <td>62.9744</td>\n",
       "      <td>NB</td>\n",
       "      <td>var1</td>\n",
       "      <td>True</td>\n",
       "      <td>c381f362-c75a-4591-8943-f2396b469c8e</td>\n",
       "      <td>gi|146415395|ref|XM_001483618.1|</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1704.0</td>\n",
       "      <td>83.45</td>\n",
       "      <td>1422.0</td>\n",
       "      <td>XM_001483618</td>\n",
       "      <td>294746</td>\n",
       "      <td>Meyerozyma guilliermondii ATCC 6260</td>\n",
       "      <td>2676.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4752703</th>\n",
       "      <td>0fc5af02-a55d-418a-993d-c118c1fa7fca</td>\n",
       "      <td>True</td>\n",
       "      <td>2559</td>\n",
       "      <td>9.215</td>\n",
       "      <td>barcode05</td>\n",
       "      <td>76.6154</td>\n",
       "      <td>NB</td>\n",
       "      <td>var1</td>\n",
       "      <td>True</td>\n",
       "      <td>0fc5af02-a55d-418a-993d-c118c1fa7fca</td>\n",
       "      <td>gi|354543245|emb|HE605203.1|</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>2522.0</td>\n",
       "      <td>85.09</td>\n",
       "      <td>2146.0</td>\n",
       "      <td>HE605203</td>\n",
       "      <td>5480</td>\n",
       "      <td>Candida parapsilosis</td>\n",
       "      <td>2488.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4752706</th>\n",
       "      <td>35a99b2c-8666-44e6-83e9-74f030f7841d</td>\n",
       "      <td>True</td>\n",
       "      <td>481</td>\n",
       "      <td>9.217</td>\n",
       "      <td>barcode07</td>\n",
       "      <td>64.1538</td>\n",
       "      <td>NB</td>\n",
       "      <td>var1</td>\n",
       "      <td>True</td>\n",
       "      <td>35a99b2c-8666-44e6-83e9-74f030f7841d</td>\n",
       "      <td>gi|18491332|emb|AL512306.16|</td>\n",
       "      <td>2.000000e-108</td>\n",
       "      <td>379.0</td>\n",
       "      <td>87.34</td>\n",
       "      <td>331.0</td>\n",
       "      <td>AL512306</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>410.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4752707</th>\n",
       "      <td>51ccadc9-570b-41eb-b7bd-822aaabc1aaa</td>\n",
       "      <td>True</td>\n",
       "      <td>2173</td>\n",
       "      <td>9.807</td>\n",
       "      <td>barcode06</td>\n",
       "      <td>66.5641</td>\n",
       "      <td>NB</td>\n",
       "      <td>var2</td>\n",
       "      <td>True</td>\n",
       "      <td>51ccadc9-570b-41eb-b7bd-822aaabc1aaa</td>\n",
       "      <td>gi|146419909|ref|XM_001485864.1|</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1264.0</td>\n",
       "      <td>82.44</td>\n",
       "      <td>1042.0</td>\n",
       "      <td>XM_001485864</td>\n",
       "      <td>294746</td>\n",
       "      <td>Meyerozyma guilliermondii ATCC 6260</td>\n",
       "      <td>2059.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4752709</th>\n",
       "      <td>af50a15a-8549-4d03-8692-39dfa4875d71</td>\n",
       "      <td>True</td>\n",
       "      <td>892</td>\n",
       "      <td>10.044</td>\n",
       "      <td>barcode06</td>\n",
       "      <td>87.0769</td>\n",
       "      <td>NB</td>\n",
       "      <td>var2</td>\n",
       "      <td>True</td>\n",
       "      <td>af50a15a-8549-4d03-8692-39dfa4875d71</td>\n",
       "      <td>gi|49649227|emb|CR382131.1|</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>807.0</td>\n",
       "      <td>88.35</td>\n",
       "      <td>713.0</td>\n",
       "      <td>CR382131</td>\n",
       "      <td>284591</td>\n",
       "      <td>Yarrowia lipolytica CLIB122</td>\n",
       "      <td>763.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4752713</th>\n",
       "      <td>ddd1b72d-7de5-4eca-bdfc-259df0901e92</td>\n",
       "      <td>True</td>\n",
       "      <td>865</td>\n",
       "      <td>9.901</td>\n",
       "      <td>barcode05</td>\n",
       "      <td>90.7692</td>\n",
       "      <td>NB</td>\n",
       "      <td>var2</td>\n",
       "      <td>True</td>\n",
       "      <td>ddd1b72d-7de5-4eca-bdfc-259df0901e92</td>\n",
       "      <td>gi|1027073461|ref|XM_016784031.1|</td>\n",
       "      <td>1.000000e-23</td>\n",
       "      <td>104.0</td>\n",
       "      <td>89.42</td>\n",
       "      <td>93.0</td>\n",
       "      <td>XM_016784031</td>\n",
       "      <td>563466</td>\n",
       "      <td>Pseudallescheria apiosperma</td>\n",
       "      <td>742.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4752714</th>\n",
       "      <td>f32da813-4cb2-4ef9-8c2c-7eb69b8b5e2d</td>\n",
       "      <td>True</td>\n",
       "      <td>901</td>\n",
       "      <td>9.097</td>\n",
       "      <td>barcode05</td>\n",
       "      <td>94.5641</td>\n",
       "      <td>NB</td>\n",
       "      <td>var1</td>\n",
       "      <td>True</td>\n",
       "      <td>f32da813-4cb2-4ef9-8c2c-7eb69b8b5e2d</td>\n",
       "      <td>gi|1069583067|ref|XM_018320179.1|</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>848.0</td>\n",
       "      <td>85.61</td>\n",
       "      <td>726.0</td>\n",
       "      <td>XM_018320179</td>\n",
       "      <td>33203</td>\n",
       "      <td>Purpureocillium lilacinum</td>\n",
       "      <td>831.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4752717</th>\n",
       "      <td>20ceff7f-99ea-45f1-8209-23f8e4d638c1</td>\n",
       "      <td>True</td>\n",
       "      <td>622</td>\n",
       "      <td>9.575</td>\n",
       "      <td>barcode05</td>\n",
       "      <td>73.9487</td>\n",
       "      <td>NB</td>\n",
       "      <td>var1</td>\n",
       "      <td>True</td>\n",
       "      <td>20ceff7f-99ea-45f1-8209-23f8e4d638c1</td>\n",
       "      <td>gi|1027070327|ref|XM_016790036.1|</td>\n",
       "      <td>2.000000e-109</td>\n",
       "      <td>549.0</td>\n",
       "      <td>81.79</td>\n",
       "      <td>449.0</td>\n",
       "      <td>XM_016790036</td>\n",
       "      <td>563466</td>\n",
       "      <td>Pseudallescheria apiosperma</td>\n",
       "      <td>551.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4752718</th>\n",
       "      <td>35d131b7-aecd-4228-a8d5-e9d4d07a37b1</td>\n",
       "      <td>True</td>\n",
       "      <td>1507</td>\n",
       "      <td>9.811</td>\n",
       "      <td>barcode05</td>\n",
       "      <td>94.5641</td>\n",
       "      <td>NB</td>\n",
       "      <td>var1</td>\n",
       "      <td>True</td>\n",
       "      <td>35d131b7-aecd-4228-a8d5-e9d4d07a37b1</td>\n",
       "      <td>gi|354545630|emb|HE605206.1|</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1449.0</td>\n",
       "      <td>86.61</td>\n",
       "      <td>1255.0</td>\n",
       "      <td>HE605206</td>\n",
       "      <td>5480</td>\n",
       "      <td>Candida parapsilosis</td>\n",
       "      <td>1384.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4752719</th>\n",
       "      <td>5f469677-ebf6-40d5-9c00-8ad14cb99701</td>\n",
       "      <td>True</td>\n",
       "      <td>742</td>\n",
       "      <td>7.028</td>\n",
       "      <td>unclassified</td>\n",
       "      <td>55.5385</td>\n",
       "      <td>NB</td>\n",
       "      <td>var1</td>\n",
       "      <td>True</td>\n",
       "      <td>5f469677-ebf6-40d5-9c00-8ad14cb99701</td>\n",
       "      <td>gi|21070696|gb|AC114486.2|</td>\n",
       "      <td>2.000000e-66</td>\n",
       "      <td>247.0</td>\n",
       "      <td>87.45</td>\n",
       "      <td>216.0</td>\n",
       "      <td>AC114486</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>665.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4752720</th>\n",
       "      <td>7ccf4ade-575f-4c69-bb2f-fd00a5b1ec74</td>\n",
       "      <td>True</td>\n",
       "      <td>1011</td>\n",
       "      <td>8.391</td>\n",
       "      <td>unclassified</td>\n",
       "      <td>48.6667</td>\n",
       "      <td>NB</td>\n",
       "      <td>var2</td>\n",
       "      <td>True</td>\n",
       "      <td>7ccf4ade-575f-4c69-bb2f-fd00a5b1ec74</td>\n",
       "      <td>gi|29500866|emb|AJ511533.1|</td>\n",
       "      <td>4.000000e-29</td>\n",
       "      <td>254.0</td>\n",
       "      <td>78.35</td>\n",
       "      <td>199.0</td>\n",
       "      <td>AJ511533</td>\n",
       "      <td>5478</td>\n",
       "      <td>Candida glabrata</td>\n",
       "      <td>976.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4752722</th>\n",
       "      <td>97d49e0a-928c-440f-b2dc-79300301ab88</td>\n",
       "      <td>True</td>\n",
       "      <td>12853</td>\n",
       "      <td>7.044</td>\n",
       "      <td>barcode06</td>\n",
       "      <td>75.2821</td>\n",
       "      <td>NB</td>\n",
       "      <td>var1</td>\n",
       "      <td>True</td>\n",
       "      <td>97d49e0a-928c-440f-b2dc-79300301ab88</td>\n",
       "      <td>gi|354542869|emb|HE605202.1|</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>4369.0</td>\n",
       "      <td>75.37</td>\n",
       "      <td>3293.0</td>\n",
       "      <td>HE605202</td>\n",
       "      <td>5480</td>\n",
       "      <td>Candida parapsilosis</td>\n",
       "      <td>12758.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4752723</th>\n",
       "      <td>a0dc6dbc-8315-434a-881b-9c340649c956</td>\n",
       "      <td>True</td>\n",
       "      <td>1521</td>\n",
       "      <td>10.861</td>\n",
       "      <td>barcode06</td>\n",
       "      <td>87.3846</td>\n",
       "      <td>NB</td>\n",
       "      <td>var1</td>\n",
       "      <td>True</td>\n",
       "      <td>a0dc6dbc-8315-434a-881b-9c340649c956</td>\n",
       "      <td>gi|1105516061|emb|LT635759.1|</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1407.0</td>\n",
       "      <td>90.62</td>\n",
       "      <td>1275.0</td>\n",
       "      <td>LT635759</td>\n",
       "      <td>45354</td>\n",
       "      <td>Candida intermedia</td>\n",
       "      <td>1377.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4752724</th>\n",
       "      <td>b17ad68f-ae07-438a-b649-fd24871de434</td>\n",
       "      <td>True</td>\n",
       "      <td>373</td>\n",
       "      <td>9.284</td>\n",
       "      <td>barcode06</td>\n",
       "      <td>86.7692</td>\n",
       "      <td>NB</td>\n",
       "      <td>var1</td>\n",
       "      <td>True</td>\n",
       "      <td>b17ad68f-ae07-438a-b649-fd24871de434</td>\n",
       "      <td>gi|83775803|dbj|AP007175.1|</td>\n",
       "      <td>1.000000e-59</td>\n",
       "      <td>266.0</td>\n",
       "      <td>84.96</td>\n",
       "      <td>226.0</td>\n",
       "      <td>AP007175</td>\n",
       "      <td>510516</td>\n",
       "      <td>Aspergillus oryzae RIB40</td>\n",
       "      <td>297.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4752726</th>\n",
       "      <td>ba7064f5-134e-4394-8d15-f5224d96aaca</td>\n",
       "      <td>True</td>\n",
       "      <td>690</td>\n",
       "      <td>9.169</td>\n",
       "      <td>barcode05</td>\n",
       "      <td>72.4615</td>\n",
       "      <td>NB</td>\n",
       "      <td>var1</td>\n",
       "      <td>True</td>\n",
       "      <td>ba7064f5-134e-4394-8d15-f5224d96aaca</td>\n",
       "      <td>gi|1027071385|ref|XM_016790565.1|</td>\n",
       "      <td>5.000000e-66</td>\n",
       "      <td>289.0</td>\n",
       "      <td>83.74</td>\n",
       "      <td>242.0</td>\n",
       "      <td>XM_016790565</td>\n",
       "      <td>563466</td>\n",
       "      <td>Pseudallescheria apiosperma</td>\n",
       "      <td>566.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4752731</th>\n",
       "      <td>6afba2ec-645e-486d-ba8c-5a524adbd012</td>\n",
       "      <td>True</td>\n",
       "      <td>9036</td>\n",
       "      <td>9.374</td>\n",
       "      <td>barcode06</td>\n",
       "      <td>98.5641</td>\n",
       "      <td>NB</td>\n",
       "      <td>var1</td>\n",
       "      <td>True</td>\n",
       "      <td>6afba2ec-645e-486d-ba8c-5a524adbd012</td>\n",
       "      <td>gi|119481510|ref|XM_001260783.1|</td>\n",
       "      <td>6.000000e-96</td>\n",
       "      <td>470.0</td>\n",
       "      <td>82.34</td>\n",
       "      <td>387.0</td>\n",
       "      <td>XM_001260783</td>\n",
       "      <td>331117</td>\n",
       "      <td>Neosartorya fischeri NRRL 181</td>\n",
       "      <td>8957.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4752732</th>\n",
       "      <td>784e5220-76e8-4884-8908-ef8e3cc2470f</td>\n",
       "      <td>True</td>\n",
       "      <td>1220</td>\n",
       "      <td>9.144</td>\n",
       "      <td>unclassified</td>\n",
       "      <td>48.7692</td>\n",
       "      <td>NB</td>\n",
       "      <td>var1</td>\n",
       "      <td>True</td>\n",
       "      <td>784e5220-76e8-4884-8908-ef8e3cc2470f</td>\n",
       "      <td>gi|1069580681|ref|XM_018318986.1|</td>\n",
       "      <td>3.000000e-86</td>\n",
       "      <td>259.0</td>\n",
       "      <td>90.73</td>\n",
       "      <td>235.0</td>\n",
       "      <td>XM_018318986</td>\n",
       "      <td>33203</td>\n",
       "      <td>Purpureocillium lilacinum</td>\n",
       "      <td>1106.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4752733</th>\n",
       "      <td>a5a03b02-cc19-405b-8adc-93043e5fd248</td>\n",
       "      <td>True</td>\n",
       "      <td>19088</td>\n",
       "      <td>10.507</td>\n",
       "      <td>barcode06</td>\n",
       "      <td>72.8205</td>\n",
       "      <td>NB</td>\n",
       "      <td>var2</td>\n",
       "      <td>True</td>\n",
       "      <td>a5a03b02-cc19-405b-8adc-93043e5fd248</td>\n",
       "      <td>gi|574144059|dbj|AP012219.1|</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>16081.0</td>\n",
       "      <td>88.96</td>\n",
       "      <td>14306.0</td>\n",
       "      <td>AP012219</td>\n",
       "      <td>1.00334e+06</td>\n",
       "      <td>Kluyveromyces marxianus DMKU3-1042</td>\n",
       "      <td>18966.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4752734</th>\n",
       "      <td>0c1933d2-470c-4a5f-bd06-82836839dc78</td>\n",
       "      <td>True</td>\n",
       "      <td>30176</td>\n",
       "      <td>10.54</td>\n",
       "      <td>barcode06</td>\n",
       "      <td>73.9487</td>\n",
       "      <td>NB</td>\n",
       "      <td>var1</td>\n",
       "      <td>True</td>\n",
       "      <td>0c1933d2-470c-4a5f-bd06-82836839dc78</td>\n",
       "      <td>gi|1078718512|gb|CP017624.1|</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>14685.0</td>\n",
       "      <td>88.25</td>\n",
       "      <td>12959.0</td>\n",
       "      <td>CP017624</td>\n",
       "      <td>237561</td>\n",
       "      <td>Candida albicans SC5314</td>\n",
       "      <td>30051.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2817017 rows  19 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      read_id passes_filtering  \\\n",
       "0        3de506ef-9b26-4aee-b11e-bf8df5a65299             True   \n",
       "1        43a2ceb3-ea58-45c4-a9c0-9e845a3ee3f5             True   \n",
       "2        45768783-cf86-4203-a864-57c1be00e671             True   \n",
       "4        576ecd5e-1a24-42af-97fd-f0d5fc3a4dde             True   \n",
       "5        5c0dd0f2-884a-4d7d-857c-62b50825adc2             True   \n",
       "6        602c44b6-f466-4dc7-8a3a-d763a30da22e             True   \n",
       "7        a8f7ec39-022a-4632-b4f4-6529cbf39074             True   \n",
       "8        ba58174b-4528-4f2a-a619-d67e66d37d67             True   \n",
       "9        bac39b1c-06c2-4b4d-b66d-7f1bf58832df             True   \n",
       "10       c7ba1d1d-80dd-4b90-bf3a-6d866b79b072             True   \n",
       "12       f1c062a6-7f6e-4637-8137-9108f3f2b1e2             True   \n",
       "13       f36087d3-636a-4036-b313-c0f2980ba534             True   \n",
       "14       00ddb787-1b87-4e47-8d86-52d826aca670             True   \n",
       "15       0480b709-7f84-4453-8777-1663eac5e09d            False   \n",
       "16       0a888311-4b8f-45dd-877c-90fa55a7d249             True   \n",
       "17       0d97acf8-bc81-42de-9804-66c471b2a199             True   \n",
       "19       2071f5ef-b1a7-4d71-b559-1c45fe011fae             True   \n",
       "21       70c6d2c5-5d19-403e-a790-c1c8c8fe9369             True   \n",
       "23       7f227f67-411f-4fdf-943a-28fb8902fcd5             True   \n",
       "25       8b7e2a43-32ae-419c-9326-2cb14eaf1535             True   \n",
       "26       bdacc264-db52-4d72-8d6b-3a256c26f576             True   \n",
       "27       c425bc37-e969-41cf-9904-3b70790331dc             True   \n",
       "28       cb3af2d3-7fe3-4bac-8565-0ae34344af18             True   \n",
       "29       d1f8a29b-fcc9-4ff8-812d-9b4cedf53116             True   \n",
       "31       25168a41-18d3-4cf5-b2b9-102182e97027             True   \n",
       "32       2a5e0b04-70e8-41ba-9a47-a93f77d0914d             True   \n",
       "33       2f53ad37-e3cf-49cd-804e-7064d603da3d             True   \n",
       "36       45803d09-ef84-4636-8546-0d0cbada3bc6             True   \n",
       "37       5255afda-4298-4094-a546-209dc170b9b2             True   \n",
       "38       5dc9fed9-1c13-4e46-a86c-b316992f9a05             True   \n",
       "...                                       ...              ...   \n",
       "4752682  4d151404-0ec1-495b-ab62-3b7f79d9eef8             True   \n",
       "4752683  6da6eb99-3922-446f-89cd-a9044f7a088d             True   \n",
       "4752686  92be42d8-83a6-4004-98c5-63f036a17ceb             True   \n",
       "4752688  13eedd39-7d1a-429d-9f4d-be8c57af7e59             True   \n",
       "4752691  379d3855-07f1-4407-9374-a27e93e790fa             True   \n",
       "4752692  4e556279-987d-4861-aacd-f25777056fbf             True   \n",
       "4752693  4fff9153-b942-4b45-a53d-111b21e7e04d             True   \n",
       "4752694  5ce685aa-a67f-46ad-87bf-b87d7e73417c             True   \n",
       "4752695  793c1235-69c1-4aea-80aa-bd24acb664a7             True   \n",
       "4752698  9da0e9bc-6912-46e8-8918-24aa8c1b5257             True   \n",
       "4752699  b7a2a56f-c7ce-4258-93b0-99485433cbf6             True   \n",
       "4752700  c381f362-c75a-4591-8943-f2396b469c8e             True   \n",
       "4752703  0fc5af02-a55d-418a-993d-c118c1fa7fca             True   \n",
       "4752706  35a99b2c-8666-44e6-83e9-74f030f7841d             True   \n",
       "4752707  51ccadc9-570b-41eb-b7bd-822aaabc1aaa             True   \n",
       "4752709  af50a15a-8549-4d03-8692-39dfa4875d71             True   \n",
       "4752713  ddd1b72d-7de5-4eca-bdfc-259df0901e92             True   \n",
       "4752714  f32da813-4cb2-4ef9-8c2c-7eb69b8b5e2d             True   \n",
       "4752717  20ceff7f-99ea-45f1-8209-23f8e4d638c1             True   \n",
       "4752718  35d131b7-aecd-4228-a8d5-e9d4d07a37b1             True   \n",
       "4752719  5f469677-ebf6-40d5-9c00-8ad14cb99701             True   \n",
       "4752720  7ccf4ade-575f-4c69-bb2f-fd00a5b1ec74             True   \n",
       "4752722  97d49e0a-928c-440f-b2dc-79300301ab88             True   \n",
       "4752723  a0dc6dbc-8315-434a-881b-9c340649c956             True   \n",
       "4752724  b17ad68f-ae07-438a-b649-fd24871de434             True   \n",
       "4752726  ba7064f5-134e-4394-8d15-f5224d96aaca             True   \n",
       "4752731  6afba2ec-645e-486d-ba8c-5a524adbd012             True   \n",
       "4752732  784e5220-76e8-4884-8908-ef8e3cc2470f             True   \n",
       "4752733  a5a03b02-cc19-405b-8adc-93043e5fd248             True   \n",
       "4752734  0c1933d2-470c-4a5f-bd06-82836839dc78             True   \n",
       "\n",
       "        sequence_length_template mean_qscore_template barcode_arrangement  \\\n",
       "0                            503               10.481           barcode07   \n",
       "1                            425                 9.72           barcode07   \n",
       "2                            726                8.658           barcode07   \n",
       "4                            831                9.943           barcode07   \n",
       "5                            464                8.324           barcode07   \n",
       "6                            829                8.201           barcode07   \n",
       "7                            761                9.036           barcode07   \n",
       "8                            990                9.436           barcode07   \n",
       "9                            835                8.581           barcode07   \n",
       "10                           660                9.043           barcode07   \n",
       "12                           704                8.643           barcode07   \n",
       "13                           524                9.846           barcode07   \n",
       "14                           863                9.655           barcode07   \n",
       "15                           731                 6.24           barcode07   \n",
       "16                           555                7.279           barcode07   \n",
       "17                           649                8.726           barcode07   \n",
       "19                           691                8.385           barcode07   \n",
       "21                           929               10.422           barcode07   \n",
       "23                           589                8.195           barcode07   \n",
       "25                           569                 9.69        unclassified   \n",
       "26                           785                 9.98           barcode07   \n",
       "27                           512                9.742           barcode07   \n",
       "28                           982                8.334           barcode06   \n",
       "29                           631                9.865           barcode05   \n",
       "31                           944               10.339           barcode07   \n",
       "32                           718                7.906           barcode07   \n",
       "33                          1564                9.063           barcode07   \n",
       "36                           663                9.031           barcode07   \n",
       "37                          1142                9.975           barcode07   \n",
       "38                           872                9.116           barcode07   \n",
       "...                          ...                  ...                 ...   \n",
       "4752682                     1470               10.624           barcode06   \n",
       "4752683                     2246                8.759           barcode06   \n",
       "4752686                     6373                9.123           barcode05   \n",
       "4752688                     1526               10.257           barcode05   \n",
       "4752691                     1209                8.351           barcode07   \n",
       "4752692                     5092                8.933           barcode06   \n",
       "4752693                      743                 9.12           barcode07   \n",
       "4752694                      678                 8.33           barcode06   \n",
       "4752695                      525                8.402           barcode06   \n",
       "4752698                     1187                8.779           barcode05   \n",
       "4752699                      753                9.501           barcode05   \n",
       "4752700                     2802                8.542           barcode06   \n",
       "4752703                     2559                9.215           barcode05   \n",
       "4752706                      481                9.217           barcode07   \n",
       "4752707                     2173                9.807           barcode06   \n",
       "4752709                      892               10.044           barcode06   \n",
       "4752713                      865                9.901           barcode05   \n",
       "4752714                      901                9.097           barcode05   \n",
       "4752717                      622                9.575           barcode05   \n",
       "4752718                     1507                9.811           barcode05   \n",
       "4752719                      742                7.028        unclassified   \n",
       "4752720                     1011                8.391        unclassified   \n",
       "4752722                    12853                7.044           barcode06   \n",
       "4752723                     1521               10.861           barcode06   \n",
       "4752724                      373                9.284           barcode06   \n",
       "4752726                      690                9.169           barcode05   \n",
       "4752731                     9036                9.374           barcode06   \n",
       "4752732                     1220                9.144        unclassified   \n",
       "4752733                    19088               10.507           barcode06   \n",
       "4752734                    30176                10.54           barcode06   \n",
       "\n",
       "        barcode_score kit variant  pc_survived  \\\n",
       "0             84.0513  NB    var1         True   \n",
       "1             90.8718  NB    var1         True   \n",
       "2             90.8718  NB    var1         True   \n",
       "4             91.2308  NB    var1         True   \n",
       "5             79.4359  NB    var1         True   \n",
       "6             65.3846  NB    var1         True   \n",
       "7             98.3077  NB    var1         True   \n",
       "8             76.8718  NB    var2         True   \n",
       "9             76.1026  NB    var1         True   \n",
       "10            91.2308  NB    var1         True   \n",
       "12            98.3077  NB    var1         True   \n",
       "13            78.8205  NB    var1         True   \n",
       "14            98.6154  NB    var2         True   \n",
       "15            67.9487  NB    var1         True   \n",
       "16            75.0769  NB    var1         True   \n",
       "17            90.7179  NB    var1         True   \n",
       "19            83.2821  NB    var1         True   \n",
       "21            91.2308  NB    var1         True   \n",
       "23            94.2051  NB    var1         True   \n",
       "25               36.4  NB    var1         True   \n",
       "26            90.2051  NB    var1         True   \n",
       "27            91.2308  NB    var1         True   \n",
       "28            94.4615  NB    var1         True   \n",
       "29            86.9231  NB    var1         True   \n",
       "31            98.3077  NB    var1         True   \n",
       "32            90.7179  NB    var1         True   \n",
       "33            90.7179  NB    var1         True   \n",
       "36            87.3846  NB    var1         True   \n",
       "37            65.4872  NB    var1         True   \n",
       "38            72.6667  NB    var1         True   \n",
       "...               ...  ..     ...          ...   \n",
       "4752682        98.359  NB    var2         True   \n",
       "4752683       64.4615  NB    var1         True   \n",
       "4752686       71.2308  NB    var1         True   \n",
       "4752688       94.5641  NB    var1         True   \n",
       "4752691       86.8205  NB    var2         True   \n",
       "4752692       90.1538  NB    var2         True   \n",
       "4752693       91.2308  NB    var1         True   \n",
       "4752694       80.4103  NB    var1         True   \n",
       "4752695       94.4615  NB    var1         True   \n",
       "4752698       91.1795  NB    var2         True   \n",
       "4752699       67.1795  NB    var1         True   \n",
       "4752700       62.9744  NB    var1         True   \n",
       "4752703       76.6154  NB    var1         True   \n",
       "4752706       64.1538  NB    var1         True   \n",
       "4752707       66.5641  NB    var2         True   \n",
       "4752709       87.0769  NB    var2         True   \n",
       "4752713       90.7692  NB    var2         True   \n",
       "4752714       94.5641  NB    var1         True   \n",
       "4752717       73.9487  NB    var1         True   \n",
       "4752718       94.5641  NB    var1         True   \n",
       "4752719       55.5385  NB    var1         True   \n",
       "4752720       48.6667  NB    var2         True   \n",
       "4752722       75.2821  NB    var1         True   \n",
       "4752723       87.3846  NB    var1         True   \n",
       "4752724       86.7692  NB    var1         True   \n",
       "4752726       72.4615  NB    var1         True   \n",
       "4752731       98.5641  NB    var1         True   \n",
       "4752732       48.7692  NB    var1         True   \n",
       "4752733       72.8205  NB    var2         True   \n",
       "4752734       73.9487  NB    var1         True   \n",
       "\n",
       "                                    qseqid_nt  \\\n",
       "0        3de506ef-9b26-4aee-b11e-bf8df5a65299   \n",
       "1        43a2ceb3-ea58-45c4-a9c0-9e845a3ee3f5   \n",
       "2        45768783-cf86-4203-a864-57c1be00e671   \n",
       "4        576ecd5e-1a24-42af-97fd-f0d5fc3a4dde   \n",
       "5        5c0dd0f2-884a-4d7d-857c-62b50825adc2   \n",
       "6        602c44b6-f466-4dc7-8a3a-d763a30da22e   \n",
       "7        a8f7ec39-022a-4632-b4f4-6529cbf39074   \n",
       "8        ba58174b-4528-4f2a-a619-d67e66d37d67   \n",
       "9        bac39b1c-06c2-4b4d-b66d-7f1bf58832df   \n",
       "10       c7ba1d1d-80dd-4b90-bf3a-6d866b79b072   \n",
       "12       f1c062a6-7f6e-4637-8137-9108f3f2b1e2   \n",
       "13       f36087d3-636a-4036-b313-c0f2980ba534   \n",
       "14       00ddb787-1b87-4e47-8d86-52d826aca670   \n",
       "15       0480b709-7f84-4453-8777-1663eac5e09d   \n",
       "16       0a888311-4b8f-45dd-877c-90fa55a7d249   \n",
       "17       0d97acf8-bc81-42de-9804-66c471b2a199   \n",
       "19       2071f5ef-b1a7-4d71-b559-1c45fe011fae   \n",
       "21       70c6d2c5-5d19-403e-a790-c1c8c8fe9369   \n",
       "23       7f227f67-411f-4fdf-943a-28fb8902fcd5   \n",
       "25       8b7e2a43-32ae-419c-9326-2cb14eaf1535   \n",
       "26       bdacc264-db52-4d72-8d6b-3a256c26f576   \n",
       "27       c425bc37-e969-41cf-9904-3b70790331dc   \n",
       "28       cb3af2d3-7fe3-4bac-8565-0ae34344af18   \n",
       "29       d1f8a29b-fcc9-4ff8-812d-9b4cedf53116   \n",
       "31       25168a41-18d3-4cf5-b2b9-102182e97027   \n",
       "32       2a5e0b04-70e8-41ba-9a47-a93f77d0914d   \n",
       "33       2f53ad37-e3cf-49cd-804e-7064d603da3d   \n",
       "36       45803d09-ef84-4636-8546-0d0cbada3bc6   \n",
       "37       5255afda-4298-4094-a546-209dc170b9b2   \n",
       "38       5dc9fed9-1c13-4e46-a86c-b316992f9a05   \n",
       "...                                       ...   \n",
       "4752682  4d151404-0ec1-495b-ab62-3b7f79d9eef8   \n",
       "4752683  6da6eb99-3922-446f-89cd-a9044f7a088d   \n",
       "4752686  92be42d8-83a6-4004-98c5-63f036a17ceb   \n",
       "4752688  13eedd39-7d1a-429d-9f4d-be8c57af7e59   \n",
       "4752691  379d3855-07f1-4407-9374-a27e93e790fa   \n",
       "4752692  4e556279-987d-4861-aacd-f25777056fbf   \n",
       "4752693  4fff9153-b942-4b45-a53d-111b21e7e04d   \n",
       "4752694  5ce685aa-a67f-46ad-87bf-b87d7e73417c   \n",
       "4752695  793c1235-69c1-4aea-80aa-bd24acb664a7   \n",
       "4752698  9da0e9bc-6912-46e8-8918-24aa8c1b5257   \n",
       "4752699  b7a2a56f-c7ce-4258-93b0-99485433cbf6   \n",
       "4752700  c381f362-c75a-4591-8943-f2396b469c8e   \n",
       "4752703  0fc5af02-a55d-418a-993d-c118c1fa7fca   \n",
       "4752706  35a99b2c-8666-44e6-83e9-74f030f7841d   \n",
       "4752707  51ccadc9-570b-41eb-b7bd-822aaabc1aaa   \n",
       "4752709  af50a15a-8549-4d03-8692-39dfa4875d71   \n",
       "4752713  ddd1b72d-7de5-4eca-bdfc-259df0901e92   \n",
       "4752714  f32da813-4cb2-4ef9-8c2c-7eb69b8b5e2d   \n",
       "4752717  20ceff7f-99ea-45f1-8209-23f8e4d638c1   \n",
       "4752718  35d131b7-aecd-4228-a8d5-e9d4d07a37b1   \n",
       "4752719  5f469677-ebf6-40d5-9c00-8ad14cb99701   \n",
       "4752720  7ccf4ade-575f-4c69-bb2f-fd00a5b1ec74   \n",
       "4752722  97d49e0a-928c-440f-b2dc-79300301ab88   \n",
       "4752723  a0dc6dbc-8315-434a-881b-9c340649c956   \n",
       "4752724  b17ad68f-ae07-438a-b649-fd24871de434   \n",
       "4752726  ba7064f5-134e-4394-8d15-f5224d96aaca   \n",
       "4752731  6afba2ec-645e-486d-ba8c-5a524adbd012   \n",
       "4752732  784e5220-76e8-4884-8908-ef8e3cc2470f   \n",
       "4752733  a5a03b02-cc19-405b-8adc-93043e5fd248   \n",
       "4752734  0c1933d2-470c-4a5f-bd06-82836839dc78   \n",
       "\n",
       "                                  sseqid_nt      evalue_nt  length_nt  \\\n",
       "0              gi|16605664|emb|AL358232.19|  3.000000e-152      430.0   \n",
       "1                gi|18497210|gb|AC097491.2|   3.000000e-96      352.0   \n",
       "2           gi|1871197|gb|U91318.1|HUU91318  7.000000e-165      429.0   \n",
       "4                gi|20377037|gb|AC109631.4|  6.000000e-166      460.0   \n",
       "5               gi|13235051|emb|AL160237.4|  7.000000e-103      234.0   \n",
       "6                gi|15706113|gb|AC022176.6|  3.000000e-154      402.0   \n",
       "7            gi|1002623499|ref|NG_011975.2|  2.000000e-146      388.0   \n",
       "8                gi|28173291|gb|AC114815.5|   0.000000e+00      913.0   \n",
       "9              gi|117414195|gb|AC090616.13|  6.000000e-166      445.0   \n",
       "10                 gi|2052037|emb|Z94801.1|  6.000000e-155      558.0   \n",
       "12             gi|14031071|emb|AL136160.18|  7.000000e-100      369.0   \n",
       "13        gi|1034651703|ref|XM_011536146.2|  5.000000e-150      464.0   \n",
       "14               gi|24371354|gb|AC126174.9|  2.000000e-145      518.0   \n",
       "15               gi|20806260|gb|AC092896.9|   4.000000e-48      368.0   \n",
       "16               gi|21240688|gb|AC063977.6|   9.000000e-73      283.0   \n",
       "17               gi|19807715|gb|AC063965.8|  5.000000e-121      341.0   \n",
       "19             gi|13872265|emb|AL158822.19|  8.000000e-134      368.0   \n",
       "21              gi|26050975|gb|AC112694.12|   0.000000e+00      709.0   \n",
       "23       gi|7381629|gb|AC022078.12|AC022078  6.000000e-100      293.0   \n",
       "25             gi|380353348|emb|HE681722.1|   0.000000e+00      511.0   \n",
       "26           gi|337621|gb|M77219.1|HUMRSCME   7.000000e-16       88.0   \n",
       "27             gi|15131484|emb|AL391476.20|   2.000000e-74      220.0   \n",
       "28             gi|354543245|emb|HE605203.1|  6.000000e-167      501.0   \n",
       "29        gi|1069598653|ref|XM_018323264.1|  6.000000e-105      383.0   \n",
       "31              gi|16517072|emb|AL590662.5|   0.000000e+00      574.0   \n",
       "32                 gi|2181872|emb|Z84497.1|   5.000000e-96      338.0   \n",
       "33              gi|8217573|emb|AL139350.17|   0.000000e+00      988.0   \n",
       "36              gi|13751498|emb|AL451103.7|  8.000000e-119      350.0   \n",
       "37               gi|20270170|gb|AC099782.2|   0.000000e+00      865.0   \n",
       "38               gi|19807899|gb|AC110769.2|   0.000000e+00      520.0   \n",
       "...                                     ...            ...        ...   \n",
       "4752682        gi|70991147|ref|XM_745330.1|   0.000000e+00      710.0   \n",
       "4752683          gi|48526539|gb|AY560609.1|   0.000000e+00     1418.0   \n",
       "4752686         gi|532597203|gb|KF015766.1|   0.000000e+00      643.0   \n",
       "4752688   gi|1069579669|ref|XM_018318480.1|  2.000000e-128      353.0   \n",
       "4752691          gi|17149768|gb|AC023950.6|   0.000000e+00     1083.0   \n",
       "4752692        gi|196049150|emb|CR380956.2|   0.000000e+00     5219.0   \n",
       "4752693         gi|16415916|emb|AL627245.1|  7.000000e-140      388.0   \n",
       "4752694    gi|169772144|ref|XM_001820490.1|  8.000000e-114      574.0   \n",
       "4752695        gi|70994853|ref|XM_747111.1|   7.000000e-84      266.0   \n",
       "4752698   gi|1069628914|ref|XM_018328678.1|  7.000000e-132      585.0   \n",
       "4752699   gi|1069577477|ref|XM_018317384.1|   6.000000e-56      222.0   \n",
       "4752700    gi|146415395|ref|XM_001483618.1|   0.000000e+00     1704.0   \n",
       "4752703        gi|354543245|emb|HE605203.1|   0.000000e+00     2522.0   \n",
       "4752706        gi|18491332|emb|AL512306.16|  2.000000e-108      379.0   \n",
       "4752707    gi|146419909|ref|XM_001485864.1|   0.000000e+00     1264.0   \n",
       "4752709         gi|49649227|emb|CR382131.1|   0.000000e+00      807.0   \n",
       "4752713   gi|1027073461|ref|XM_016784031.1|   1.000000e-23      104.0   \n",
       "4752714   gi|1069583067|ref|XM_018320179.1|   0.000000e+00      848.0   \n",
       "4752717   gi|1027070327|ref|XM_016790036.1|  2.000000e-109      549.0   \n",
       "4752718        gi|354545630|emb|HE605206.1|   0.000000e+00     1449.0   \n",
       "4752719          gi|21070696|gb|AC114486.2|   2.000000e-66      247.0   \n",
       "4752720         gi|29500866|emb|AJ511533.1|   4.000000e-29      254.0   \n",
       "4752722        gi|354542869|emb|HE605202.1|   0.000000e+00     4369.0   \n",
       "4752723       gi|1105516061|emb|LT635759.1|   0.000000e+00     1407.0   \n",
       "4752724         gi|83775803|dbj|AP007175.1|   1.000000e-59      266.0   \n",
       "4752726   gi|1027071385|ref|XM_016790565.1|   5.000000e-66      289.0   \n",
       "4752731    gi|119481510|ref|XM_001260783.1|   6.000000e-96      470.0   \n",
       "4752732   gi|1069580681|ref|XM_018318986.1|   3.000000e-86      259.0   \n",
       "4752733        gi|574144059|dbj|AP012219.1|   0.000000e+00    16081.0   \n",
       "4752734        gi|1078718512|gb|CP017624.1|   0.000000e+00    14685.0   \n",
       "\n",
       "         pident_nt  nident_nt       sacc_nt   staxids_nt  \\\n",
       "0            90.70      390.0      AL358232         9606   \n",
       "1            86.65      305.0      AC097491         9606   \n",
       "2            92.07      395.0        U91318         9606   \n",
       "4            91.09      419.0      AC109631         9606   \n",
       "5            96.58      226.0      AL160237         9606   \n",
       "6            92.29      371.0      AC022176         9606   \n",
       "7            92.27      358.0     NG_011975         9606   \n",
       "8            86.09      786.0      AC114815         9606   \n",
       "9            91.91      409.0      AC090616         9606   \n",
       "10           86.38      482.0        Z94801         9606   \n",
       "12           86.45      319.0      AL136160         9606   \n",
       "13           89.01      413.0  XM_011536146         9606   \n",
       "14           86.49      448.0      AC126174         9606   \n",
       "15           77.99      287.0      AC092896         9606   \n",
       "16           86.22      244.0      AC063977         9606   \n",
       "17           91.20      311.0      AC063965         9606   \n",
       "19           91.58      337.0      AL158822         9606   \n",
       "21           92.52      656.0      AC112694         9606   \n",
       "23           90.44      265.0      AC022078         9606   \n",
       "25           90.02      460.0      HE681722      1136231   \n",
       "26           87.50       77.0        M77219         9606   \n",
       "27           91.36      201.0      AL391476         9606   \n",
       "28           89.22      447.0      HE605203         5480   \n",
       "29           86.42      331.0  XM_018323264        33203   \n",
       "31           91.99      528.0      AL590662         9606   \n",
       "32           86.98      294.0        Z84497         9606   \n",
       "33           88.87      878.0      AL139350         9606   \n",
       "36           90.29      316.0      AL451103         9606   \n",
       "37           87.98      761.0      AC099782         9606   \n",
       "38           91.92      478.0      AC110769         9606   \n",
       "...            ...        ...           ...          ...   \n",
       "4752682      88.59      629.0     XM_745330       330879   \n",
       "4752683      82.30     1167.0      AY560609        40410   \n",
       "4752686      86.31      555.0      KF015766        36911   \n",
       "4752688      91.78      324.0  XM_018318480        33203   \n",
       "4752691      85.13      922.0      AC023950         9606   \n",
       "4752692      83.81     4374.0      CR380956         5478   \n",
       "4752693      91.24      354.0      AL627245         9606   \n",
       "4752694      81.71      469.0  XM_001820490       510516   \n",
       "4752695      89.47      238.0     XM_747111       330879   \n",
       "4752698      83.59      489.0  XM_018328678        33203   \n",
       "4752699      86.94      193.0  XM_018317384        33203   \n",
       "4752700      83.45     1422.0  XM_001483618       294746   \n",
       "4752703      85.09     2146.0      HE605203         5480   \n",
       "4752706      87.34      331.0      AL512306         9606   \n",
       "4752707      82.44     1042.0  XM_001485864       294746   \n",
       "4752709      88.35      713.0      CR382131       284591   \n",
       "4752713      89.42       93.0  XM_016784031       563466   \n",
       "4752714      85.61      726.0  XM_018320179        33203   \n",
       "4752717      81.79      449.0  XM_016790036       563466   \n",
       "4752718      86.61     1255.0      HE605206         5480   \n",
       "4752719      87.45      216.0      AC114486         9606   \n",
       "4752720      78.35      199.0      AJ511533         5478   \n",
       "4752722      75.37     3293.0      HE605202         5480   \n",
       "4752723      90.62     1275.0      LT635759        45354   \n",
       "4752724      84.96      226.0      AP007175       510516   \n",
       "4752726      83.74      242.0  XM_016790565       563466   \n",
       "4752731      82.34      387.0  XM_001260783       331117   \n",
       "4752732      90.73      235.0  XM_018318986        33203   \n",
       "4752733      88.96    14306.0      AP012219  1.00334e+06   \n",
       "4752734      88.25    12959.0      CP017624       237561   \n",
       "\n",
       "                                    scomnames_nt  read_length_pc  \n",
       "0                                   Homo sapiens           434.0  \n",
       "1                                   Homo sapiens           339.0  \n",
       "2                                   Homo sapiens           654.0  \n",
       "4                                   Homo sapiens           762.0  \n",
       "5                                   Homo sapiens           393.0  \n",
       "6                                   Homo sapiens           829.0  \n",
       "7                                   Homo sapiens           685.0  \n",
       "8                                   Homo sapiens           869.0  \n",
       "9                                   Homo sapiens           764.0  \n",
       "10                                  Homo sapiens           592.0  \n",
       "12                                  Homo sapiens           615.0  \n",
       "13                                  Homo sapiens           448.0  \n",
       "14                                  Homo sapiens           725.0  \n",
       "15                                  Homo sapiens           662.0  \n",
       "16                                  Homo sapiens           485.0  \n",
       "17                                  Homo sapiens           577.0  \n",
       "19                                  Homo sapiens           606.0  \n",
       "21                                  Homo sapiens           859.0  \n",
       "23                                  Homo sapiens           516.0  \n",
       "25               Candida orthopsilosis Co 90-125           569.0  \n",
       "26                                  Homo sapiens           714.0  \n",
       "27                                  Homo sapiens           391.0  \n",
       "28                          Candida parapsilosis           904.0  \n",
       "29                     Purpureocillium lilacinum           560.0  \n",
       "31                                  Homo sapiens           870.0  \n",
       "32                                  Homo sapiens           624.0  \n",
       "33                                  Homo sapiens          1484.0  \n",
       "36                                  Homo sapiens           592.0  \n",
       "37                                  Homo sapiens          1077.0  \n",
       "38                                  Homo sapiens           802.0  \n",
       "...                                          ...             ...  \n",
       "4752682              Aspergillus fumigatus Af293          1319.0  \n",
       "4752683  Cryptococcus neoformans var. neoformans          2229.0  \n",
       "4752686                    Clavispora lusitaniae          6295.0  \n",
       "4752688                Purpureocillium lilacinum          1390.0  \n",
       "4752691                             Homo sapiens          1077.0  \n",
       "4752692                         Candida glabrata          4964.0  \n",
       "4752693                             Homo sapiens           672.0  \n",
       "4752694                 Aspergillus oryzae RIB40           548.0  \n",
       "4752695              Aspergillus fumigatus Af293           455.0  \n",
       "4752698                Purpureocillium lilacinum          1089.0  \n",
       "4752699                Purpureocillium lilacinum           679.0  \n",
       "4752700      Meyerozyma guilliermondii ATCC 6260          2676.0  \n",
       "4752703                     Candida parapsilosis          2488.0  \n",
       "4752706                             Homo sapiens           410.0  \n",
       "4752707      Meyerozyma guilliermondii ATCC 6260          2059.0  \n",
       "4752709              Yarrowia lipolytica CLIB122           763.0  \n",
       "4752713              Pseudallescheria apiosperma           742.0  \n",
       "4752714                Purpureocillium lilacinum           831.0  \n",
       "4752717              Pseudallescheria apiosperma           551.0  \n",
       "4752718                     Candida parapsilosis          1384.0  \n",
       "4752719                             Homo sapiens           665.0  \n",
       "4752720                         Candida glabrata           976.0  \n",
       "4752722                     Candida parapsilosis         12758.0  \n",
       "4752723                       Candida intermedia          1377.0  \n",
       "4752724                 Aspergillus oryzae RIB40           297.0  \n",
       "4752726              Pseudallescheria apiosperma           566.0  \n",
       "4752731            Neosartorya fischeri NRRL 181          8957.0  \n",
       "4752732                Purpureocillium lilacinum          1106.0  \n",
       "4752733       Kluyveromyces marxianus DMKU3-1042         18966.0  \n",
       "4752734                  Candida albicans SC5314         30051.0  \n",
       "\n",
       "[2817017 rows x 19 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sum_df[sum_df.pident_nt > 0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# fills the nan with T/F for easier handling\n",
    "#sum_df.sseqid_rg.fillna(False, inplace=True)\n",
    "sum_df.sseqid_nt.fillna(False, inplace=True)\n",
    "# filter out the rg blast hit\n",
    "#ntblasthit_reads = sum_df[(sum_df.sseqid_rg == False) & (sum_df.sseqid_nt != False) & (sum_df.passes_filtering == True) & (sum_df.pc_survived == True) & (sum_df.nl_survived == True)]\n",
    "total_reads = sum_df[(sum_df.sseqid_nt != False) & (sum_df.passes_filtering == True) & (sum_df.pc_survived == True)]\n",
    "#ntblasthit_reads = total_reads[total_reads.sseqid_nt != False]\n",
    "pd.set_option('display.max_columns', None)\n",
    "\n",
    "#Removing unncessary columns from joint_NCBI_df_taxonomy\n",
    "#I hard coded to select as it depends on what information you need\n",
    "for columns in ['passes_filtering', 'sequence_length_template',\n",
    "       'mean_qscore_template', 'barcode_score', 'kit',\n",
    "       'variant', 'pc_survived', 'qseqid_nt', 'sseqid_nt', 'evalue_nt',\n",
    "       'length_nt', 'pident_nt', 'nident_nt', 'sacc_nt', 'read_length_pc']:\n",
    "    del total_reads[columns] \n",
    "# keep the common name in to filter the human out"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "###########This block for barcodes order\n",
    "# now everything left is what we want to plot out\n",
    "# arrange different barcodes\n",
    "# REMEMBER: This need to be manual check:\n",
    "\n",
    "total_reads_barcode05 = total_reads[total_reads.barcode_arrangement == 'barcode05']\n",
    "total_reads_barcode06 = total_reads[total_reads.barcode_arrangement == 'barcode06']\n",
    "total_reads_barcode07 = total_reads[total_reads.barcode_arrangement == 'barcode07']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# now just filter out all the reads that are not going to used for ploting\n",
    "barcode05_total = total_reads[total_reads.barcode_arrangement.str.contains('barcode05')]\n",
    "barcode06_total = total_reads[total_reads.barcode_arrangement.str.contains('barcode06')]\n",
    "barcode07_total = total_reads[total_reads.barcode_arrangement.str.contains('barcode07')]\n",
    "\n",
    "# now concat them together. but remember that the index has to change to add the column for ncbi taxa\n",
    "total_reads_filtered_barcodes = pd.concat([barcode05_total, \n",
    "                                           barcode06_total, \n",
    "                                           barcode07_total], ignore_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "total_reads_filtered_barcodes.staxids_nt.fillna(False, inplace=True)\n",
    "ntblasthit_reads_filtered_barcodes = total_reads_filtered_barcodes[(total_reads_filtered_barcodes.staxids_nt != False)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "ntblasthit_reads_barcode07_nohuman = ntblasthit_reads_filtered_barcodes[(ntblasthit_reads_filtered_barcodes.scomnames_nt != 'Homo sapiens') & (ntblasthit_reads_filtered_barcodes.barcode_arrangement == 'barcode07')]\n",
    "ntblasthit_reads_barcode07_nohuman = ntblasthit_reads_barcode07_nohuman.reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# as the concate before will result the staxid into a string\n",
    "# so now change the string back to float so it can be recognized by the NCBITaxa\n",
    "for taxid in ntblasthit_reads_barcode07_nohuman['staxids_nt']:\n",
    "    if ';' in str(taxid):\n",
    "        taxid = taxid.split(';')[0]\n",
    "\n",
    "    else:\n",
    "        pass\n",
    "    float(taxid)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Function to generate taxonomy columns based on NCBITaxa results for NCBI hit dataframe, \n",
    "def search_rank_output_name_append_column(df, staxid_column, rank_search):\n",
    "    \"\"\"Input df, staxid_column from same df and rank_search (a desired taxonomic rank \n",
    "    from each staxid's lineage), outputs taxonomic name corresponding to rank_search or 'Unclassified' if\n",
    "    unavailable and appends to df row by row\"\"\"\n",
    "    rank_list = []\n",
    "    for read_index in range(0, len(staxid_column)):\n",
    "        taxid = ''\n",
    "        if ';' in str(staxid_column[read_index]):\n",
    "            taxid = staxid_column[read_index].split(';')[0]\n",
    "################# BE CAREFUL HERE !!! #######################################\n",
    "#            df.loc[staxid_column =='71426;73839', 'staxids_nt'] = 73839\n",
    "        else:\n",
    "            taxid = staxid_column[read_index]\n",
    "        \n",
    "        \n",
    "        taxid_lineage = ''\n",
    "        taxid_lineage = ncbi.get_lineage(taxid)\n",
    "        \n",
    "        names = ''\n",
    "        names = ncbi.get_taxid_translator(taxid_lineage)\n",
    "        \n",
    "        ranks = ''\n",
    "        ranks = ncbi.get_rank(taxid_lineage) #Dict\n",
    "        \n",
    "        ranks2names = ''\n",
    "        ranks2names = {ranks[k]:names[k] for k in names.keys() & ranks}\n",
    "        \n",
    "        if rank_search in ranks2names.keys():\n",
    "            rank_list.append(ranks2names[rank_search])#if rank in dict, print name\n",
    "        else:\n",
    "            rank_list.append('Unclassified')\n",
    "    df[rank_search] = rank_list\n",
    "# NOTE: Appending is always slow, try and find a better way e.g df.apply to a column based on staxids column\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/yiheng/anaconda3/lib/python3.6/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:240: UserWarning: taxid 291364 was translated into 2231603\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/home/yiheng/anaconda3/lib/python3.6/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:240: UserWarning: taxid 1635066 was translated into 1508258\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/home/yiheng/anaconda3/lib/python3.6/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:240: UserWarning: taxid 72029 was translated into 1836592\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/home/yiheng/anaconda3/lib/python3.6/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:240: UserWarning: taxid 251300 was translated into 332405\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/home/yiheng/anaconda3/lib/python3.6/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:240: UserWarning: taxid 1916993 was translated into 303\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/home/yiheng/anaconda3/lib/python3.6/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:240: UserWarning: taxid 595593 was translated into 656366\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/home/yiheng/anaconda3/lib/python3.6/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:240: UserWarning: taxid 71426 was translated into 1915362\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1, 131567, 2759, 33154, 33208, 6072, 33213, 33511, 7711, 89593, 7742, 7776, 117570, 117571, 8287, 1338369, 32523, 32524, 40674, 32525, 9347, 1437010, 314146, 9443, 376913, 314293, 9526, 314295, 9604, 207598, 9605, 9606]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/yiheng/anaconda3/lib/python3.6/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:240: UserWarning: taxid 278968 was translated into 2025498\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/home/yiheng/anaconda3/lib/python3.6/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:240: UserWarning: taxid 38483 was translated into 1436886\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/home/yiheng/anaconda3/lib/python3.6/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:240: UserWarning: taxid 112484 was translated into 1431917\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n"
     ]
    }
   ],
   "source": [
    "ncbi.get_lineage(291364)\n",
    "ncbi.get_lineage(1635066)\n",
    "ncbi.get_lineage(72029)\n",
    "ncbi.get_lineage(251300)\n",
    "ncbi.get_lineage(332405)\n",
    "ncbi.get_lineage(1916993)\n",
    "ncbi.get_lineage(595593)\n",
    "ncbi.get_lineage(71426)\n",
    "ncbi.get_lineage(291364)\n",
    "ncbi.get_lineage(278968)\n",
    "ncbi.get_lineage(38483)\n",
    "ncbi.get_lineage(112484)\n",
    "print(ncbi.get_lineage(9606))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{}"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ncbi.get_rank([1969666])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{6: 'Azorhizobium', 7: 'Azorhizobium caulinodans'}"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ncbi.get_taxid_translator('40674')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "# see above, some taxids are changed (translated) sometimes, which is probably due to the version of the local nt database.\n",
    "# therefore, here I mannualy adjusted all the taxids that has been changed to the new numbers.\n",
    "\n",
    "ntblasthit_reads_barcode07_nohuman.loc[ntblasthit_reads_barcode07_nohuman.staxids_nt=='291364', 'staxids_nt'] = 2231603\n",
    "ntblasthit_reads_barcode07_nohuman.loc[ntblasthit_reads_barcode07_nohuman.staxids_nt=='1635066', 'staxids_nt'] = 1508258\n",
    "ntblasthit_reads_barcode07_nohuman.loc[ntblasthit_reads_barcode07_nohuman.staxids_nt=='72029', 'staxids_nt'] = 1836592\n",
    "ntblasthit_reads_barcode07_nohuman.loc[ntblasthit_reads_barcode07_nohuman.staxids_nt=='251300', 'staxids_nt'] = 332405\n",
    "ntblasthit_reads_barcode07_nohuman.loc[ntblasthit_reads_barcode07_nohuman.staxids_nt=='652050', 'staxids_nt'] = 2204306\n",
    "ntblasthit_reads_barcode07_nohuman.loc[ntblasthit_reads_barcode07_nohuman.staxids_nt=='1916993', 'staxids_nt'] = 303\n",
    "ntblasthit_reads_barcode07_nohuman.loc[ntblasthit_reads_barcode07_nohuman.staxids_nt=='595593', 'staxids_nt'] = 656366\n",
    "ntblasthit_reads_barcode07_nohuman.loc[ntblasthit_reads_barcode07_nohuman.staxids_nt=='71426;73839', 'staxids_nt'] = 73839\n",
    "ntblasthit_reads_barcode07_nohuman.loc[ntblasthit_reads_barcode07_nohuman.staxids_nt=='278968', 'staxids_nt'] = 2025498\n",
    "ntblasthit_reads_barcode07_nohuman.loc[ntblasthit_reads_barcode07_nohuman.staxids_nt=='38483', 'staxids_nt'] = 1436886\n",
    "ntblasthit_reads_barcode07_nohuman.loc[ntblasthit_reads_barcode07_nohuman.staxids_nt=='112484', 'staxids_nt'] = 1431917\n",
    "# the taxdb was changed and this taxid was removed and untracked\n",
    "ntblasthit_reads_barcode07_nohuman = ntblasthit_reads_barcode07_nohuman[ntblasthit_reads_barcode07_nohuman.staxids_nt!='1969666']\n",
    "ntblasthit_reads_barcode07_nohuman = ntblasthit_reads_barcode07_nohuman.reset_index(drop=True)\n",
    "#ntblasthit_reads_filtered_barcodes.loc[ntblasthit_reads_filtered_barcodes.staxids_nt=='278968', 'staxids_nt'] = 2025498\n",
    "#ntblasthit_reads_filtered_barcodes.loc[ntblasthit_reads_filtered_barcodes.staxids_nt=='71426', 'staxids_nt'] = 1915362\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Int64Index([], dtype='int64')"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ntblasthit_reads_barcode07_nohuman.index[ntblasthit_reads_barcode07_nohuman.staxids_nt.str.contains('811799') == True]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>read_id</th>\n",
       "      <th>barcode_arrangement</th>\n",
       "      <th>staxids_nt</th>\n",
       "      <th>scomnames_nt</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [read_id, barcode_arrangement, staxids_nt, scomnames_nt]\n",
       "Index: []"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ntblasthit_reads_barcode07_nohuman[ntblasthit_reads_barcode07_nohuman.staxids_nt.str.contains('1431917') == True]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#ntblasthit_reads_filtered_barcodes_added_TaxaRank = ntblasthit_reads_filtered_barcodes.copy()\n",
    "# pretty slow\n",
    "rank_list = ['superkingdom', 'phylum', 'class', 'order', 'family', 'genus', 'species']\n",
    "for rank in rank_list:\n",
    "    search_rank_output_name_append_column(ntblasthit_reads_barcode07_nohuman, \n",
    "                                          ntblasthit_reads_barcode07_nohuman.staxids_nt, \n",
    "                                          rank)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# add the flowcell column just in case\n",
    "#ntblasthit_reads_filtered_barcodes = ntblasthit_reads_filtered_barcodes.assign(Flowcell = column_name)\n",
    "#total_reads_filtered_barcodes = total_reads_filtered_barcodes.assign(Flowcell = column_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#total_reads_filtered_barcodes.to_csv(r'/home/yiheng/analysis/WGS/%s_totaltaxa.tab' % column_name, header=column_name, index=None, sep='\\t')\n",
    "ntblasthit_reads_barcode07_nomammal = ntblasthit_reads_barcode07_nohuman[ntblasthit_reads_barcode07_nohuman['class'] != 'Mammalia']\n",
    "ntblasthit_reads_barcode07_nomammal = ntblasthit_reads_barcode07_nomammal.reset_index(drop = True)\n",
    "ntblasthit_reads_barcode07_nomammal.to_csv(r'/home/yiheng/test/analysis/%s_barcode07_nomammal_nttaxa.tab' % folder_name, header='clinical_2', index=None, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Defines a function to make manual adjustments to the NCBI taxonomic sorting\n",
    "def manually_adjust_ncbi_taxonomy(rank_select, entry_select, ranks_to_edit, new_entry):\n",
    "    \"\"\"Select dataframe entries to edit by a specific rank (rank_select) \n",
    "    and entry related to that rank (entry_select), \n",
    "    provide a list of ranks to edit for those selected entries (ranks_to_edit),\n",
    "    replace these ranks with a new entry (new_entry)\"\"\"\n",
    "    if entry_select in list(ntblasthit_reads_filtered_barcodes[rank_select].unique()):\n",
    "        selection = (ntblasthit_reads_filtered_barcodes[rank_select] == entry_select)\n",
    "        selected_tax_ranks = ranks_to_edit\n",
    "        for ranks in selected_tax_ranks:\n",
    "            ntblasthit_reads_filtered_barcodes[ranks][selection] = new_entry"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "#manually_adjust_ncbi_taxonomy('species', \n",
    "#                            'Cloning vector lambda TXF97', \n",
    "#                            ['superkingdom', 'phylum', 'class', 'order', 'family', 'genus'], \n",
    "#                            'Cloning vector lambda TXF97')\n",
    "\n",
    "#manually_adjust_ncbi_taxonomy('superkingdom', \n",
    "#                            'Viruses', \n",
    "#                            ['phylum', 'class'], \n",
    "#                            'dsDNA viruses, no RNA stage')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#count pivot table of dataframe with taxonomic columns: \n",
    "#rows - rank names, columns - barcode, sorted by desired class\n",
    "def generate_ncbi_taxonomy_pivot(tax_df, rank, bcs, num):\n",
    "    \"\"\"From tax_df, generate a pivot table listing num rank counts, sorted by bcs\"\"\"\n",
    "    pivot_table = tax_df.pivot_table(values='read_id', \n",
    "                                            index=rank, \n",
    "                                            columns='barcode_arrangement', \n",
    "                                            aggfunc='count', \n",
    "                                            fill_value=0)\n",
    "    pivot_table.columns.name = None\n",
    "    pivot_table = pivot_table.sort_values(bcs, axis=0, ascending=False).head(n=num)\n",
    "    return pivot_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'genus'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-34-9dd7bc054ec9>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mgenerate_ncbi_taxonomy_pivot\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mntblasthit_reads_filtered_barcodes\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'genus'\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m'barcode07'\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;36m10\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m<ipython-input-33-1f97ae0d7ce4>\u001b[0m in \u001b[0;36mgenerate_ncbi_taxonomy_pivot\u001b[0;34m(tax_df, rank, bcs, num)\u001b[0m\n\u001b[1;32m      7\u001b[0m                                             \u001b[0mcolumns\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'barcode_arrangement'\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      8\u001b[0m                                             \u001b[0maggfunc\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'count'\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 9\u001b[0;31m                                             fill_value=0)\n\u001b[0m\u001b[1;32m     10\u001b[0m     \u001b[0mpivot_table\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mname\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     11\u001b[0m     \u001b[0mpivot_table\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpivot_table\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msort_values\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mbcs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mascending\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mhead\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mn\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mnum\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/home/yiheng/anaconda3/lib/python3.6/site-packages/pandas/core/frame.py\u001b[0m in \u001b[0;36mpivot_table\u001b[0;34m(self, values, index, columns, aggfunc, fill_value, margins, dropna, margins_name)\u001b[0m\n\u001b[1;32m   4466\u001b[0m                            \u001b[0maggfunc\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0maggfunc\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfill_value\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mfill_value\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   4467\u001b[0m                            \u001b[0mmargins\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mmargins\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdropna\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mdropna\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 4468\u001b[0;31m                            margins_name=margins_name)\n\u001b[0m\u001b[1;32m   4469\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   4470\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mstack\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlevel\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m-\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdropna\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/home/yiheng/anaconda3/lib/python3.6/site-packages/pandas/core/reshape/pivot.py\u001b[0m in \u001b[0;36mpivot_table\u001b[0;34m(data, values, index, columns, aggfunc, fill_value, margins, dropna, margins_name)\u001b[0m\n\u001b[1;32m     79\u001b[0m         \u001b[0mvalues\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mlist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mvalues\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     80\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 81\u001b[0;31m     \u001b[0mgrouped\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdata\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgroupby\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkeys\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     82\u001b[0m     \u001b[0magged\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mgrouped\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0magg\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0maggfunc\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     83\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/home/yiheng/anaconda3/lib/python3.6/site-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36mgroupby\u001b[0;34m(self, by, axis, level, as_index, sort, group_keys, squeeze, **kwargs)\u001b[0m\n\u001b[1;32m   5160\u001b[0m         return groupby(self, by=by, axis=axis, level=level, as_index=as_index,\n\u001b[1;32m   5161\u001b[0m                        \u001b[0msort\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0msort\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mgroup_keys\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mgroup_keys\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msqueeze\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0msqueeze\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 5162\u001b[0;31m                        **kwargs)\n\u001b[0m\u001b[1;32m   5163\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   5164\u001b[0m     def asfreq(self, freq, method=None, how=None, normalize=False,\n",
      "\u001b[0;32m/home/yiheng/anaconda3/lib/python3.6/site-packages/pandas/core/groupby.py\u001b[0m in \u001b[0;36mgroupby\u001b[0;34m(obj, by, **kwds)\u001b[0m\n\u001b[1;32m   1846\u001b[0m         \u001b[0;32mraise\u001b[0m \u001b[0mTypeError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'invalid type: %s'\u001b[0m \u001b[0;34m%\u001b[0m \u001b[0mtype\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mobj\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1847\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1848\u001b[0;31m     \u001b[0;32mreturn\u001b[0m \u001b[0mklass\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mobj\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mby\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwds\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1849\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1850\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/home/yiheng/anaconda3/lib/python3.6/site-packages/pandas/core/groupby.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, obj, keys, axis, level, grouper, exclusions, selection, as_index, sort, group_keys, squeeze, **kwargs)\u001b[0m\n\u001b[1;32m    514\u001b[0m                                                     \u001b[0mlevel\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mlevel\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    515\u001b[0m                                                     \u001b[0msort\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0msort\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 516\u001b[0;31m                                                     mutated=self.mutated)\n\u001b[0m\u001b[1;32m    517\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    518\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mobj\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mobj\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/home/yiheng/anaconda3/lib/python3.6/site-packages/pandas/core/groupby.py\u001b[0m in \u001b[0;36m_get_grouper\u001b[0;34m(obj, key, axis, level, sort, mutated, validate)\u001b[0m\n\u001b[1;32m   2932\u001b[0m                 \u001b[0min_axis\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlevel\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mgpr\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mFalse\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mgpr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2933\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2934\u001b[0;31m                 \u001b[0;32mraise\u001b[0m \u001b[0mKeyError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mgpr\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   2935\u001b[0m         \u001b[0;32melif\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mgpr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mGrouper\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0mgpr\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mkey\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2936\u001b[0m             \u001b[0;31m# Add key to exclusions\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'genus'"
     ]
    }
   ],
   "source": [
    "generate_ncbi_taxonomy_pivot(ntblasthit_reads_filtered_barcodes, 'genus','barcode07', 10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "total_no_of_each_barcode_df = pd.DataFrame({'barcode01':len(total_reads_filtered_barcodes[total_reads_filtered_barcodes.barcode_arrangement == 'barcode01']),\n",
    "                                            'barcode02':len(total_reads_filtered_barcodes[total_reads_filtered_barcodes.barcode_arrangement == 'barcode02']),\n",
    "                                            'barcode03':len(total_reads_filtered_barcodes[total_reads_filtered_barcodes.barcode_arrangement == 'barcode03']),\n",
    "                                            'barcode04':len(total_reads_filtered_barcodes[total_reads_filtered_barcodes.barcode_arrangement == 'barcode04']),\n",
    "                                            'barcode05':len(total_reads_filtered_barcodes[total_reads_filtered_barcodes.barcode_arrangement == 'barcode05'])}, index = [0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>barcode01</th>\n",
       "      <th>barcode02</th>\n",
       "      <th>barcode03</th>\n",
       "      <th>barcode04</th>\n",
       "      <th>barcode05</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>42133</td>\n",
       "      <td>70120</td>\n",
       "      <td>30148</td>\n",
       "      <td>2443</td>\n",
       "      <td>4597</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   barcode01  barcode02  barcode03  barcode04  barcode05\n",
       "0      42133      70120      30148       2443       4597"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "total_no_of_each_barcode_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "pivot_df = generate_ncbi_taxonomy_pivot(ntblasthit_reads_filtered_barcodes, 'genus','barcode04',10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# % pivot table of dataframe with taxonomic columns: \n",
    "#rows - rank names, columns - barcode, sorted by desired class\n",
    "def generate_ncbi_taxonomy_pivot_pc(tax_df, rank, bcs, num):\n",
    "    \"\"\"From tax_df, generate a pivot table listing num rank %, sorted ascending by bcs\"\"\"\n",
    "    pivot_table_pc = generate_ncbi_taxonomy_pivot(tax_df, rank, bcs, num)\n",
    "    \n",
    "    for ind, brcs in enumerate(pivot_table_pc.columns):\n",
    "        pivot_table_pc[str(brcs) + '%'] = pivot_table_pc[brcs].apply(lambda x: (x / total_no_of_each_barcode_df.iloc[0,ind])*100)\n",
    "    for brcs in pivot_table_pc.columns:\n",
    "        if '%' not in brcs:\n",
    "            pivot_table_pc.drop(brcs, axis=1, inplace=True)\n",
    "    \n",
    "\n",
    "    pivot_table_pc2 = pivot_table_pc.sort_values(str(bcs) + '%', axis=0, ascending=False)\n",
    "    return pivot_table_pc2.applymap(lambda x: '%.3f' % x).head(n=num)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>barcode01%</th>\n",
       "      <th>barcode02%</th>\n",
       "      <th>barcode03%</th>\n",
       "      <th>barcode04%</th>\n",
       "      <th>barcode05%</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>genus</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Triticum</th>\n",
       "      <td>0.000</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.007</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sus</th>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.041</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Alternaria</th>\n",
       "      <td>0.009</td>\n",
       "      <td>0.158</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.041</td>\n",
       "      <td>0.065</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Monopterus</th>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Panthera</th>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Hordeum</th>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Helobdella</th>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Hymenobacter</th>\n",
       "      <td>0.000</td>\n",
       "      <td>0.044</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Exophiala</th>\n",
       "      <td>0.000</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Oryzias</th>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Frankia</th>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sphingomonas</th>\n",
       "      <td>0.000</td>\n",
       "      <td>0.154</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Campylobacter</th>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.041</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gaeumannomyces</th>\n",
       "      <td>0.000</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Plasmodium</th>\n",
       "      <td>0.000</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pectobacterium</th>\n",
       "      <td>0.012</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pseudonocardia</th>\n",
       "      <td>0.000</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Paraphaeosphaeria</th>\n",
       "      <td>0.000</td>\n",
       "      <td>0.006</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pseudomonas</th>\n",
       "      <td>0.185</td>\n",
       "      <td>1.412</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.655</td>\n",
       "      <td>0.044</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pseudodesulfovibrio</th>\n",
       "      <td>0.000</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    barcode01% barcode02% barcode03% barcode04% barcode05%\n",
       "genus                                                                     \n",
       "Triticum                 0.000      0.001      0.007      0.000      0.000\n",
       "Sus                      0.000      0.000      0.003      0.041      0.000\n",
       "Alternaria               0.009      0.158      0.003      0.041      0.065\n",
       "Monopterus               0.000      0.000      0.003      0.000      0.000\n",
       "Panthera                 0.000      0.000      0.003      0.000      0.000\n",
       "Hordeum                  0.000      0.000      0.003      0.000      0.000\n",
       "Helobdella               0.000      0.000      0.003      0.000      0.000\n",
       "Hymenobacter             0.000      0.044      0.003      0.000      0.000\n",
       "Exophiala                0.000      0.001      0.003      0.000      0.000\n",
       "Oryzias                  0.000      0.000      0.003      0.000      0.000\n",
       "Frankia                  0.000      0.000      0.003      0.000      0.000\n",
       "Sphingomonas             0.000      0.154      0.003      0.000      0.000\n",
       "Campylobacter            0.000      0.000      0.003      0.041      0.000\n",
       "Gaeumannomyces           0.000      0.001      0.003      0.000      0.000\n",
       "Plasmodium               0.000      0.001      0.000      0.000      0.000\n",
       "Pectobacterium           0.012      0.000      0.000      0.000      0.000\n",
       "Pseudonocardia           0.000      0.001      0.000      0.000      0.000\n",
       "Paraphaeosphaeria        0.000      0.006      0.000      0.000      0.000\n",
       "Pseudomonas              0.185      1.412      0.000      0.655      0.044\n",
       "Pseudodesulfovibrio      0.000      0.001      0.000      0.000      0.000"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "generate_ncbi_taxonomy_pivot_pc(ntblasthit_reads_filtered_barcodes, 'genus', 'barcode03', 20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "df1=pd.DataFrame([[1, 2],[10, 1], [3, 4], [7, 8]], columns=['A', 'B'])\n",
    "df2=pd.DataFrame([[1, 4],[3, 5]], columns=['A', 'B'])\n",
    "# now replace the location use the index of each df selected from value by .loc function\n",
    "# first select the value from df2 column A\n",
    "for values in df2['A'].unique():\n",
    "    # then locate the position in df1 and replace it by the VALUE of the position in df2\n",
    "    df1.loc[df1[df1['A'] == (values)].index, 'B'] = df2.loc[df2[df2['A'] == (values)].index, 'B'].values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "df1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
