{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This is a notebook to pull down all the information for figure 3 (ncbi hits). The design is to compair between each sample (barcode) instead of each flowcell as they sequenced the same thing. Therefore I just pulled all the ncbi hits for each flowcell for each barcode and merged them together according to the barcode."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import os\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib\n",
    "from itertools import product\n",
    "import argparse\n",
    "from ete3 import NCBITaxa "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "ncbi = NCBITaxa()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# use argparse to do this\n",
    "#BASEDIR = args.BASEDIR\n",
    "BASEDIR = '/home/yiheng/data/20170617_FAH05432'\n",
    "\n",
    "# here we define the folder name of the dataframe it created by capturing the folder from the BASDIR\n",
    "folder_name = os.path.basename(BASEDIR)\n",
    "column_name = folder_name.split('_')[-1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# first check if the analysis folder is there\n",
    "folder_list = 'analysis  basecalled_data  scripts  tracking  workspace'.split(' ')\n",
    "for x in range(0,folder_list.count('')):\n",
    "    folder_list.remove('')\n",
    "#fix this test\n",
    "if not set(os.listdir(os.path.abspath(BASEDIR))) == set (folder_list):\n",
    "    print(\"Something wrong with basefolder. check it please.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/yiheng/anaconda3/lib/python3.6/site-packages/IPython/core/interactiveshell.py:2717: DtypeWarning: Columns (1,2,3,5,27) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  interactivity=interactivity, compiler=compiler, result=result)\n"
     ]
    }
   ],
   "source": [
    "# get the dataframe there\n",
    "dataframe = os.path.join(BASEDIR, 'analysis', 'summary_df_%s.tab' % folder_name)\n",
    "sum_df = pd.read_csv(dataframe, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['read_id', 'passes_filtering', 'sequence_length_template',\n",
       "       'mean_qscore_template', 'barcode_arrangement', 'barcode_score', 'kit',\n",
       "       'variant', 'pc_survived', 'nl_survived', 'qseqid_rg', 'sseqid_rg',\n",
       "       'evalue_rg', 'length_rg', 'pident_rg', 'nident_rg', 'sacc_rg',\n",
       "       'staxids_rg', 'scomnames_rg', 'read_length_pc_x', 'qseqid_nt',\n",
       "       'sseqid_nt', 'evalue_nt', 'length_nt', 'pident_nt', 'nident_nt',\n",
       "       'sacc_nt', 'staxids_nt', 'scomnames_nt', 'read_length_pc_y'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sum_df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# fills the nan with T/F for easier handling\n",
    "sum_df.sseqid_rg.fillna(False, inplace=True)\n",
    "sum_df.sseqid_nt.fillna(False, inplace=True)\n",
    "# filter out the rg blast hit\n",
    "ntblasthit_reads = sum_df[(sum_df.sseqid_rg == False) & (sum_df.sseqid_nt != False) & (sum_df.passes_filtering == True) & (sum_df.pc_survived == True) & (sum_df.nl_survived == True)]\n",
    "pd.set_option('display.max_columns', None)\n",
    "\n",
    "#Removing unncessary columns from joint_NCBI_df_taxonomy\n",
    "#I hard coded to select as it depends on what information you need\n",
    "for columns in ['passes_filtering', 'sequence_length_template',\n",
    "       'mean_qscore_template', 'barcode_score', 'kit',\n",
    "       'variant', 'pc_survived', 'nl_survived', 'qseqid_rg', 'sseqid_rg',\n",
    "       'evalue_rg', 'length_rg', 'pident_rg', 'nident_rg', 'sacc_rg',\n",
    "       'staxids_rg', 'scomnames_rg', 'read_length_pc_x', 'qseqid_nt',\n",
    "       'sseqid_nt', 'evalue_nt', 'length_nt', 'pident_nt', 'nident_nt',\n",
    "       'sacc_nt','scomnames_nt', 'read_length_pc_y']:\n",
    "    del ntblasthit_reads[columns] "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "###########This block for barcodes order\n",
    "# now everything left is what we want to plot out\n",
    "# arrange different barcodes\n",
    "# REMEMBER: This need to be manual check:\n",
    "# barcode01 is Pst infected sample, barcode02 is Zymo sample, 03 is nill sample, 04 is co-infection sample and 05 is Pyre sample\n",
    "new_barcodes = ['barcode01', 'barcode02', 'barcode03', 'barcode04', 'barcode05']\n",
    "if column_name == 'run1':\n",
    "    ntblasthit_barcode01 = ntblasthit_reads[ntblasthit_reads.barcode_arrangement == 'barcode02']\n",
    "    ntblasthit_barcode02 = ntblasthit_reads[ntblasthit_reads.barcode_arrangement == 'barcode06']\n",
    "    ntblasthit_barcode03 = ntblasthit_reads[ntblasthit_reads.barcode_arrangement == 'barcode04']\n",
    "    ntblasthit_barcode04 = ntblasthit_reads[ntblasthit_reads.barcode_arrangement == 'barcode05']\n",
    "    ntblasthit_barcode05 = ntblasthit_reads[ntblasthit_reads.barcode_arrangement == 'barcode03']\n",
    "    ntblasthit_barcode01.barcode_arrangement = ntblasthit_barcode01.barcode_arrangement.replace(to_replace='barcode02', value='barcode01')\n",
    "    ntblasthit_reads.drop(ntblasthit_reads.index[ntblasthit_reads.barcode_arrangement.str.contains('barcode07')], inplace = True)\n",
    "    ntblasthit_barcode01.barcode_arrangement = ntblasthit_barcode01.barcode_arrangement.replace(to_replace='barcode03', value='barcode07')\n",
    "    ntblasthit_barcode01.barcode_arrangement = ntblasthit_barcode01.barcode_arrangement.replace(to_replace='barcode06', value='barcode02')\n",
    "    ntblasthit_barcode01.barcode_arrangement = ntblasthit_barcode01.barcode_arrangement.replace(to_replace='barcode04', value='barcode03')\n",
    "    ntblasthit_barcode01.barcode_arrangement = ntblasthit_barcode01.barcode_arrangement.replace(to_replace='barcode05', value='barcode04')\n",
    "    ntblasthit_barcode01.barcode_arrangement = ntblasthit_barcode01.barcode_arrangement.replace(to_replace='barcode07', value='barcode05')\n",
    "\n",
    "elif column_name == 'run2':\n",
    "    # first delete the rows that contains the barcode we are going to replace\n",
    "    # I think maybe better to first adjust the df to get rid of all the mis/unclassified reads so the df get smaller\n",
    "    # and we do not need those reads for plotting\n",
    "    \n",
    "    ntblasthit_reads.drop(ntblasthit_reads.index[ntblasthit_reads.barcode_arrangement.str.contains('barcode01')], inplace = True)\n",
    "    ntblasthit_reads.drop(ntblasthit_reads.index[ntblasthit_reads.barcode_arrangement.str.contains('barcode02')], inplace = True)\n",
    "    ntblasthit_reads.drop(ntblasthit_reads.index[ntblasthit_reads.barcode_arrangement.str.contains('barcode03')], inplace = True)\n",
    "    ntblasthit_reads.drop(ntblasthit_reads.index[ntblasthit_reads.barcode_arrangement.str.contains('barcode04')], inplace = True)\n",
    "    ntblasthit_reads.drop(ntblasthit_reads.index[ntblasthit_reads.barcode_arrangement.str.contains('barcode05')], inplace = True)\n",
    "    ntblasthit_reads.drop(ntblasthit_reads.index[ntblasthit_reads.barcode_arrangement.str.contains('unclassified')], inplace = True)\n",
    "    old_barcodes = ['barcode07', 'barcode08', 'barcode09', 'barcode10', 'barcode11']\n",
    "    for old_barcode, new_barcode in zip(old_barcodes, new_barcodes):\n",
    "        ntblasthit_reads.loc\\\n",
    "        [ntblasthit_reads[ntblasthit_reads['barcode_arrangement'] == old_barcode].index,'barcode_arrangement']\\\n",
    "        = [new_barcode]\n",
    "    #ntblasthit_reads['barcode_arrangement'] = ntblasthit_reads.barcode_arrangement.replace(to_replace='barcode07', value='barcode01')\n",
    "\n",
    "    \n",
    "elif column_name == 'FAH05432':\n",
    "    ntblasthit_barcode01 = ntblasthit_reads[ntblasthit_reads.barcode_arrangement == 'barcode01']\n",
    "    ntblasthit_barcode02 = ntblasthit_reads[ntblasthit_reads.barcode_arrangement == 'barcode02']\n",
    "    ntblasthit_barcode03 = ntblasthit_reads[ntblasthit_reads.barcode_arrangement == 'barcode03']\n",
    "    ntblasthit_barcode04 = ntblasthit_reads[ntblasthit_reads.barcode_arrangement == 'barcode04']\n",
    "    ntblasthit_barcode05 = ntblasthit_reads[ntblasthit_reads.barcode_arrangement == 'barcode05']\n",
    "elif column_name == 'FAH05731':\n",
    "    ntblasthit_barcode01 = ntblasthit_reads[ntblasthit_reads.barcode_arrangement == 'barcode01']\n",
    "    ntblasthit_barcode02 = ntblasthit_reads[ntblasthit_reads.barcode_arrangement == 'barcode02']\n",
    "    ntblasthit_barcode03 = ntblasthit_reads[ntblasthit_reads.barcode_arrangement == 'barcode03']\n",
    "    ntblasthit_barcode04 = ntblasthit_reads[ntblasthit_reads.barcode_arrangement == 'barcode04']\n",
    "    ntblasthit_barcode05 = ntblasthit_reads[ntblasthit_reads.barcode_arrangement == 'barcode05']\n",
    "else:\n",
    "    pass\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# now just filter out all the reads that are not going to used for ploting\n",
    "ntblasthit_reads_barcode01 = ntblasthit_reads[ntblasthit_reads.barcode_arrangement.str.contains('barcode01')]\n",
    "ntblasthit_reads_barcode02 = ntblasthit_reads[ntblasthit_reads.barcode_arrangement.str.contains('barcode02')]\n",
    "ntblasthit_reads_barcode03 = ntblasthit_reads[ntblasthit_reads.barcode_arrangement.str.contains('barcode03')]\n",
    "ntblasthit_reads_barcode04 = ntblasthit_reads[ntblasthit_reads.barcode_arrangement.str.contains('barcode04')]\n",
    "ntblasthit_reads_barcode05 = ntblasthit_reads[ntblasthit_reads.barcode_arrangement.str.contains('barcode05')]\n",
    "\n",
    "# now concat them together. but remember that the index has to change to add the column for ncbi taxa\n",
    "ntblasthit_reads_filtered_barcodes = pd.concat([ntblasthit_reads_barcode01, \n",
    "                                                ntblasthit_reads_barcode02, \n",
    "                                                ntblasthit_reads_barcode03, \n",
    "                                                ntblasthit_reads_barcode04, \n",
    "                                                ntblasthit_reads_barcode05], ignore_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# as the concate before will result the staxid into a string\n",
    "# so now change the string back to float so it can be recognized by the NCBITaxa\n",
    "for taxid in ntblasthit_reads_filtered_barcodes['staxids_nt']:\n",
    "    if ';' in str(taxid):\n",
    "        taxid = taxid.split(';')[0]\n",
    "    else:\n",
    "        pass\n",
    "    float(taxid)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Function to generate taxonomy columns based on NCBITaxa results for NCBI hit dataframe, \n",
    "def search_rank_output_name_append_column(df, staxid_column, rank_search):\n",
    "    \"\"\"Input df, staxid_column from same df and rank_search (a desired taxonomic rank \n",
    "    from each staxid's lineage), outputs taxonomic name corresponding to rank_search or 'Unclassified' if\n",
    "    unavailable and appends to df row by row\"\"\"\n",
    "    rank_list = []\n",
    "    for read_index in range(0, len(staxid_column)):\n",
    "        taxid = ''\n",
    "        if ';' in str(staxid_column[read_index]):\n",
    "            taxid = staxid_column[read_index].split(';')[0]\n",
    "        else:\n",
    "            taxid = staxid_column[read_index]\n",
    "        \n",
    "        taxid_lineage = ''\n",
    "        taxid_lineage = ncbi.get_lineage(taxid)\n",
    "        \n",
    "        names = ''\n",
    "        names = ncbi.get_taxid_translator(taxid_lineage)\n",
    "        \n",
    "        ranks = ''\n",
    "        ranks = ncbi.get_rank(taxid_lineage) #Dict\n",
    "        \n",
    "        ranks2names = ''\n",
    "        ranks2names = {ranks[k]:names[k] for k in names.keys() & ranks}\n",
    "        \n",
    "        if rank_search in ranks2names.keys():\n",
    "            rank_list.append(ranks2names[rank_search])#if rank in dict, print name\n",
    "        else:\n",
    "            rank_list.append('Unclassified')\n",
    "    df[rank_search] = rank_list\n",
    "# NOTE: Appending is always slow, try and find a better way e.g df.apply to a column based on staxids column\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#ntblasthit_reads_filtered_barcodes_added_TaxaRank = ntblasthit_reads_filtered_barcodes.copy()\n",
    "rank_list = ['superkingdom', 'phylum', 'class', 'order', 'family', 'genus', 'species']\n",
    "for rank in rank_list:\n",
    "    search_rank_output_name_append_column(ntblasthit_reads_filtered_barcodes, \n",
    "                                          ntblasthit_reads_filtered_barcodes.staxids_nt, \n",
    "                                          rank)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>read_id</th>\n",
       "      <th>barcode_arrangement</th>\n",
       "      <th>staxids_nt</th>\n",
       "      <th>superkingdom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>807a8997-b021-4820-b33a-1b9095249dc7</td>\n",
       "      <td>barcode01</td>\n",
       "      <td>200451</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Pseudomonadales</td>\n",
       "      <td>Pseudomonadaceae</td>\n",
       "      <td>Pseudomonas</td>\n",
       "      <td>Pseudomonas poae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>c53fa9c8-f4d5-4876-8526-e04fcea0ac5d</td>\n",
       "      <td>barcode01</td>\n",
       "      <td>200451</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Pseudomonadales</td>\n",
       "      <td>Pseudomonadaceae</td>\n",
       "      <td>Pseudomonas</td>\n",
       "      <td>Pseudomonas poae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>19a62259-0925-41ef-94e4-6fe9ef776c81</td>\n",
       "      <td>barcode01</td>\n",
       "      <td>199201</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Pseudomonadales</td>\n",
       "      <td>Pseudomonadaceae</td>\n",
       "      <td>Pseudomonas</td>\n",
       "      <td>Pseudomonas syringae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>7e4e83ba-26fa-4d0b-b714-724976a9c7a7</td>\n",
       "      <td>barcode01</td>\n",
       "      <td>200451</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Pseudomonadales</td>\n",
       "      <td>Pseudomonadaceae</td>\n",
       "      <td>Pseudomonas</td>\n",
       "      <td>Pseudomonas poae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>c92e3d57-c554-4d73-9e17-a51bc14aa4d4</td>\n",
       "      <td>barcode01</td>\n",
       "      <td>1.35728e+06</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Pseudomonadales</td>\n",
       "      <td>Pseudomonadaceae</td>\n",
       "      <td>Pseudomonas</td>\n",
       "      <td>Pseudomonas syringae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>072bbc16-20a5-45b8-bcbf-e7239f48ed12</td>\n",
       "      <td>barcode01</td>\n",
       "      <td>1.32493e+06</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Pseudomonadales</td>\n",
       "      <td>Pseudomonadaceae</td>\n",
       "      <td>Pseudomonas</td>\n",
       "      <td>Pseudomonas syringae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>7eb68975-bd44-40da-b0e8-2635fcedfbb0</td>\n",
       "      <td>barcode01</td>\n",
       "      <td>716540</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>Erwiniaceae</td>\n",
       "      <td>Erwinia</td>\n",
       "      <td>Erwinia amylovora</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>f876c268-fe76-4fb7-97e7-e3c6fa2d8c8b</td>\n",
       "      <td>barcode01</td>\n",
       "      <td>1.28236e+06</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Pseudomonadales</td>\n",
       "      <td>Pseudomonadaceae</td>\n",
       "      <td>Pseudomonas</td>\n",
       "      <td>Pseudomonas poae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>00f1b4d2-cc23-4231-9b51-543873ddddd2</td>\n",
       "      <td>barcode01</td>\n",
       "      <td>634500</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>Erwiniaceae</td>\n",
       "      <td>Erwinia</td>\n",
       "      <td>Erwinia billingiae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>74885d13-7285-4de6-ae85-89a8927c36ab</td>\n",
       "      <td>barcode01</td>\n",
       "      <td>1.50567e+06</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Basidiomycota</td>\n",
       "      <td>Pucciniomycetes</td>\n",
       "      <td>Pucciniales</td>\n",
       "      <td>Pucciniaceae</td>\n",
       "      <td>Puccinia</td>\n",
       "      <td>Puccinia cf. psidii AE-2014</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>fe662e8d-bb9d-45d5-b699-7891406c5df4</td>\n",
       "      <td>barcode01</td>\n",
       "      <td>1.35728e+06</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Pseudomonadales</td>\n",
       "      <td>Pseudomonadaceae</td>\n",
       "      <td>Pseudomonas</td>\n",
       "      <td>Pseudomonas syringae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>d5533aed-db93-4957-928a-60cd6caa2d0f</td>\n",
       "      <td>barcode01</td>\n",
       "      <td>199201</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Pseudomonadales</td>\n",
       "      <td>Pseudomonadaceae</td>\n",
       "      <td>Pseudomonas</td>\n",
       "      <td>Pseudomonas syringae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>2b27f02a-bd3c-4c0d-a4fb-7e1575c76fe0</td>\n",
       "      <td>barcode01</td>\n",
       "      <td>889211</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>Erwiniaceae</td>\n",
       "      <td>Erwinia</td>\n",
       "      <td>Erwinia amylovora</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>d6a81cd2-3423-4242-9cb3-84f81be34892</td>\n",
       "      <td>barcode01</td>\n",
       "      <td>199201</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Pseudomonadales</td>\n",
       "      <td>Pseudomonadaceae</td>\n",
       "      <td>Pseudomonas</td>\n",
       "      <td>Pseudomonas syringae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>2828d03d-7417-46c5-8382-792e21003d99</td>\n",
       "      <td>barcode01</td>\n",
       "      <td>199201</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Pseudomonadales</td>\n",
       "      <td>Pseudomonadaceae</td>\n",
       "      <td>Pseudomonas</td>\n",
       "      <td>Pseudomonas syringae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>f7760040-5944-4aec-9d3f-b98389bda403</td>\n",
       "      <td>barcode01</td>\n",
       "      <td>587753</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Pseudomonadales</td>\n",
       "      <td>Pseudomonadaceae</td>\n",
       "      <td>Pseudomonas</td>\n",
       "      <td>Pseudomonas chlororaphis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>fa63e5c1-71b2-42e9-906d-879ad1ad4f3b</td>\n",
       "      <td>barcode01</td>\n",
       "      <td>1.81294e+06</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "      <td>Enterobacter</td>\n",
       "      <td>Enterobacter cloacae complex 'Hoffmann cluster...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>be038069-e219-48b3-8352-a70beaa767e0</td>\n",
       "      <td>barcode01</td>\n",
       "      <td>1.32493e+06</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Pseudomonadales</td>\n",
       "      <td>Pseudomonadaceae</td>\n",
       "      <td>Pseudomonas</td>\n",
       "      <td>Pseudomonas syringae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>f31b2c3c-87ba-4957-a72c-dd8126a70476</td>\n",
       "      <td>barcode01</td>\n",
       "      <td>199201</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Pseudomonadales</td>\n",
       "      <td>Pseudomonadaceae</td>\n",
       "      <td>Pseudomonas</td>\n",
       "      <td>Pseudomonas syringae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>52767e8d-6540-439b-a930-bb5f8ece7bb8</td>\n",
       "      <td>barcode01</td>\n",
       "      <td>1.32493e+06</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Pseudomonadales</td>\n",
       "      <td>Pseudomonadaceae</td>\n",
       "      <td>Pseudomonas</td>\n",
       "      <td>Pseudomonas syringae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>b89419aa-7906-4bdb-ad6d-06bd26e65944</td>\n",
       "      <td>barcode01</td>\n",
       "      <td>35703</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "      <td>Citrobacter</td>\n",
       "      <td>Citrobacter amalonaticus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>6047063f-eb37-4ef4-a9b0-304ced98946e</td>\n",
       "      <td>barcode01</td>\n",
       "      <td>418459</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Basidiomycota</td>\n",
       "      <td>Pucciniomycetes</td>\n",
       "      <td>Pucciniales</td>\n",
       "      <td>Pucciniaceae</td>\n",
       "      <td>Puccinia</td>\n",
       "      <td>Puccinia graminis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>af09a516-1d55-4685-8bf6-dba8d13eb50e</td>\n",
       "      <td>barcode01</td>\n",
       "      <td>205918</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Pseudomonadales</td>\n",
       "      <td>Pseudomonadaceae</td>\n",
       "      <td>Pseudomonas</td>\n",
       "      <td>Pseudomonas syringae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>b0ded9cf-3b68-4dbb-b12c-c7e8b18dd35e</td>\n",
       "      <td>barcode01</td>\n",
       "      <td>1.32493e+06</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Pseudomonadales</td>\n",
       "      <td>Pseudomonadaceae</td>\n",
       "      <td>Pseudomonas</td>\n",
       "      <td>Pseudomonas syringae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>8e937eda-6ba6-4f1c-b5ea-fd9f3690ac2c</td>\n",
       "      <td>barcode01</td>\n",
       "      <td>465817</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>Erwiniaceae</td>\n",
       "      <td>Erwinia</td>\n",
       "      <td>Erwinia tasmaniensis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>ec446ef2-0236-4dc3-a8db-29dcf2a4d5f8</td>\n",
       "      <td>barcode01</td>\n",
       "      <td>199201</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Pseudomonadales</td>\n",
       "      <td>Pseudomonadaceae</td>\n",
       "      <td>Pseudomonas</td>\n",
       "      <td>Pseudomonas syringae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>502dab27-225d-4462-ab53-5572a4ff7d46</td>\n",
       "      <td>barcode01</td>\n",
       "      <td>1.28236e+06</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Pseudomonadales</td>\n",
       "      <td>Pseudomonadaceae</td>\n",
       "      <td>Pseudomonas</td>\n",
       "      <td>Pseudomonas poae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>0d733956-7720-40c7-9ab0-19842a016005</td>\n",
       "      <td>barcode01</td>\n",
       "      <td>199201</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Pseudomonadales</td>\n",
       "      <td>Pseudomonadaceae</td>\n",
       "      <td>Pseudomonas</td>\n",
       "      <td>Pseudomonas syringae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>c77016d6-1b7f-49f7-8c4b-6c7ca171b291</td>\n",
       "      <td>barcode01</td>\n",
       "      <td>1.32493e+06</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Pseudomonadales</td>\n",
       "      <td>Pseudomonadaceae</td>\n",
       "      <td>Pseudomonas</td>\n",
       "      <td>Pseudomonas syringae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>bbdb95b5-2066-44d1-97e0-a49fcb9ce0a0</td>\n",
       "      <td>barcode01</td>\n",
       "      <td>199201</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Pseudomonadales</td>\n",
       "      <td>Pseudomonadaceae</td>\n",
       "      <td>Pseudomonas</td>\n",
       "      <td>Pseudomonas syringae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5823</th>\n",
       "      <td>f0a73265-3eaf-41e1-9ea7-5b810f177025</td>\n",
       "      <td>barcode05</td>\n",
       "      <td>930090</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Ascomycota</td>\n",
       "      <td>Dothideomycetes</td>\n",
       "      <td>Pleosporales</td>\n",
       "      <td>Pleosporaceae</td>\n",
       "      <td>Bipolaris</td>\n",
       "      <td>Bipolaris oryzae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5824</th>\n",
       "      <td>036f16cc-2e7a-4414-94a2-55b80e502345</td>\n",
       "      <td>barcode05</td>\n",
       "      <td>1.46066e+06</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Ascomycota</td>\n",
       "      <td>Dothideomycetes</td>\n",
       "      <td>Pleosporales</td>\n",
       "      <td>Didymosphaeriaceae</td>\n",
       "      <td>Paraphaeosphaeria</td>\n",
       "      <td>Paraphaeosphaeria sporulosa</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5825</th>\n",
       "      <td>3febc7c5-f308-4f64-9b37-bc2357608d43</td>\n",
       "      <td>barcode05</td>\n",
       "      <td>317</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Pseudomonadales</td>\n",
       "      <td>Pseudomonadaceae</td>\n",
       "      <td>Pseudomonas</td>\n",
       "      <td>Pseudomonas syringae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5826</th>\n",
       "      <td>2de9f0c0-1338-4978-9f88-f34336e3c16d</td>\n",
       "      <td>barcode05</td>\n",
       "      <td>341663</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Ascomycota</td>\n",
       "      <td>Eurotiomycetes</td>\n",
       "      <td>Eurotiales</td>\n",
       "      <td>Aspergillaceae</td>\n",
       "      <td>Aspergillus</td>\n",
       "      <td>Aspergillus terreus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5827</th>\n",
       "      <td>1c4cf73a-df25-4eb8-a96d-dfa117243cb0</td>\n",
       "      <td>barcode05</td>\n",
       "      <td>398673</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Ascomycota</td>\n",
       "      <td>Sordariomycetes</td>\n",
       "      <td>Hypocreales</td>\n",
       "      <td>Hypocreaceae</td>\n",
       "      <td>Trichoderma</td>\n",
       "      <td>Trichoderma gamsii</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5828</th>\n",
       "      <td>f0f4e05a-7bf0-4eeb-9a0a-5b84c6590bc2</td>\n",
       "      <td>barcode05</td>\n",
       "      <td>76761</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Pseudomonadales</td>\n",
       "      <td>Pseudomonadaceae</td>\n",
       "      <td>Pseudomonas</td>\n",
       "      <td>Pseudomonas veronii</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5829</th>\n",
       "      <td>b109ccf5-048d-45e6-a0c1-e568d4498b66</td>\n",
       "      <td>barcode05</td>\n",
       "      <td>1.6036e+06</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Insecta</td>\n",
       "      <td>Hemiptera</td>\n",
       "      <td>Pentatomidae</td>\n",
       "      <td>Cressona</td>\n",
       "      <td>Cressona rufa</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5830</th>\n",
       "      <td>00cb05e4-9dcc-4562-97b6-1d34db28b2ce</td>\n",
       "      <td>barcode05</td>\n",
       "      <td>169297</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Liliopsida</td>\n",
       "      <td>Poales</td>\n",
       "      <td>Poaceae</td>\n",
       "      <td>Aegilops</td>\n",
       "      <td>Aegilops tauschii</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5831</th>\n",
       "      <td>0be8c7f8-9da2-4d6c-84bc-4852d29a23ff</td>\n",
       "      <td>barcode05</td>\n",
       "      <td>5599</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Ascomycota</td>\n",
       "      <td>Dothideomycetes</td>\n",
       "      <td>Pleosporales</td>\n",
       "      <td>Pleosporaceae</td>\n",
       "      <td>Alternaria</td>\n",
       "      <td>Alternaria alternata</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5832</th>\n",
       "      <td>4af043da-d8a4-48d0-bb61-9130131e08b7</td>\n",
       "      <td>barcode05</td>\n",
       "      <td>562</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>Escherichia coli</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5833</th>\n",
       "      <td>4e1bc3af-c37e-44be-82e4-765000b17887</td>\n",
       "      <td>barcode05</td>\n",
       "      <td>45151</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Ascomycota</td>\n",
       "      <td>Dothideomycetes</td>\n",
       "      <td>Pleosporales</td>\n",
       "      <td>Pleosporaceae</td>\n",
       "      <td>Pyrenophora</td>\n",
       "      <td>Pyrenophora tritici-repentis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5834</th>\n",
       "      <td>54b4c516-a392-4c67-bc3f-20d9cb08dfd1</td>\n",
       "      <td>barcode05</td>\n",
       "      <td>1.32493e+06</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Pseudomonadales</td>\n",
       "      <td>Pseudomonadaceae</td>\n",
       "      <td>Pseudomonas</td>\n",
       "      <td>Pseudomonas syringae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5835</th>\n",
       "      <td>5dfeeb74-f577-4706-bc10-4173ea8c13d4</td>\n",
       "      <td>barcode05</td>\n",
       "      <td>671987</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Ascomycota</td>\n",
       "      <td>Dothideomycetes</td>\n",
       "      <td>Pleosporales</td>\n",
       "      <td>Pleosporaceae</td>\n",
       "      <td>Setosphaeria</td>\n",
       "      <td>Setosphaeria turcica</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5836</th>\n",
       "      <td>248cff4b-734e-402d-95ba-aeda3dfd5a78</td>\n",
       "      <td>barcode05</td>\n",
       "      <td>29760</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Vitales</td>\n",
       "      <td>Vitaceae</td>\n",
       "      <td>Vitis</td>\n",
       "      <td>Vitis vinifera</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5837</th>\n",
       "      <td>434bf795-fa15-4709-872e-fea2f0bf28fa</td>\n",
       "      <td>barcode05</td>\n",
       "      <td>4565</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Liliopsida</td>\n",
       "      <td>Poales</td>\n",
       "      <td>Poaceae</td>\n",
       "      <td>Triticum</td>\n",
       "      <td>Triticum aestivum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5838</th>\n",
       "      <td>0e196992-8736-4a23-a766-b6a48ac53f92</td>\n",
       "      <td>barcode05</td>\n",
       "      <td>7165</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Insecta</td>\n",
       "      <td>Diptera</td>\n",
       "      <td>Culicidae</td>\n",
       "      <td>Anopheles</td>\n",
       "      <td>Anopheles gambiae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5839</th>\n",
       "      <td>3422c128-320c-40a4-8f54-8df22e42117e</td>\n",
       "      <td>barcode05</td>\n",
       "      <td>51953</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Liliopsida</td>\n",
       "      <td>Arecales</td>\n",
       "      <td>Arecaceae</td>\n",
       "      <td>Elaeis</td>\n",
       "      <td>Elaeis guineensis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5840</th>\n",
       "      <td>afe0dea5-5835-41ad-a3d5-0111885fadb4</td>\n",
       "      <td>barcode05</td>\n",
       "      <td>562</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>Escherichia coli</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5841</th>\n",
       "      <td>74af4e06-3343-446c-8d5d-4b0904a192fe</td>\n",
       "      <td>barcode05</td>\n",
       "      <td>39946</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Liliopsida</td>\n",
       "      <td>Poales</td>\n",
       "      <td>Poaceae</td>\n",
       "      <td>Oryza</td>\n",
       "      <td>Oryza sativa</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5842</th>\n",
       "      <td>b734a7f7-1c8f-4dcf-a175-702a74f07473</td>\n",
       "      <td>barcode05</td>\n",
       "      <td>5599</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Ascomycota</td>\n",
       "      <td>Dothideomycetes</td>\n",
       "      <td>Pleosporales</td>\n",
       "      <td>Pleosporaceae</td>\n",
       "      <td>Alternaria</td>\n",
       "      <td>Alternaria alternata</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5843</th>\n",
       "      <td>7e83c505-957c-43a7-bd2f-8921b8bacede</td>\n",
       "      <td>barcode05</td>\n",
       "      <td>930090</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Ascomycota</td>\n",
       "      <td>Dothideomycetes</td>\n",
       "      <td>Pleosporales</td>\n",
       "      <td>Pleosporaceae</td>\n",
       "      <td>Bipolaris</td>\n",
       "      <td>Bipolaris oryzae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5844</th>\n",
       "      <td>9607245c-061a-4f58-8d18-a93d2b3e8cdc</td>\n",
       "      <td>barcode05</td>\n",
       "      <td>7955</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Actinopteri</td>\n",
       "      <td>Cypriniformes</td>\n",
       "      <td>Cyprinidae</td>\n",
       "      <td>Danio</td>\n",
       "      <td>Danio rerio</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5845</th>\n",
       "      <td>f6d6b97e-9204-41ef-9e43-2126f7dd8557</td>\n",
       "      <td>barcode05</td>\n",
       "      <td>7955</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Actinopteri</td>\n",
       "      <td>Cypriniformes</td>\n",
       "      <td>Cyprinidae</td>\n",
       "      <td>Danio</td>\n",
       "      <td>Danio rerio</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5846</th>\n",
       "      <td>a9dab824-a355-4c7d-bad8-9f46ad5db179</td>\n",
       "      <td>barcode05</td>\n",
       "      <td>353152</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Apicomplexa</td>\n",
       "      <td>Coccidia</td>\n",
       "      <td>Eucoccidiorida</td>\n",
       "      <td>Cryptosporidiidae</td>\n",
       "      <td>Cryptosporidium</td>\n",
       "      <td>Cryptosporidium parvum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5847</th>\n",
       "      <td>032c2734-9254-49df-b949-f21606858835</td>\n",
       "      <td>barcode05</td>\n",
       "      <td>741155</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Pseudomonadales</td>\n",
       "      <td>Pseudomonadaceae</td>\n",
       "      <td>Pseudomonas</td>\n",
       "      <td>Pseudomonas alcaliphila</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5848</th>\n",
       "      <td>59353f77-87ee-40a0-930f-328719e2bfe4</td>\n",
       "      <td>barcode05</td>\n",
       "      <td>5599</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Ascomycota</td>\n",
       "      <td>Dothideomycetes</td>\n",
       "      <td>Pleosporales</td>\n",
       "      <td>Pleosporaceae</td>\n",
       "      <td>Alternaria</td>\n",
       "      <td>Alternaria alternata</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5849</th>\n",
       "      <td>86bf9dba-4728-463d-b767-231f0863ebea</td>\n",
       "      <td>barcode05</td>\n",
       "      <td>5599</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Ascomycota</td>\n",
       "      <td>Dothideomycetes</td>\n",
       "      <td>Pleosporales</td>\n",
       "      <td>Pleosporaceae</td>\n",
       "      <td>Alternaria</td>\n",
       "      <td>Alternaria alternata</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5850</th>\n",
       "      <td>88013815-813d-4f85-833c-2d5d0f647a94</td>\n",
       "      <td>barcode05</td>\n",
       "      <td>930090</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Ascomycota</td>\n",
       "      <td>Dothideomycetes</td>\n",
       "      <td>Pleosporales</td>\n",
       "      <td>Pleosporaceae</td>\n",
       "      <td>Bipolaris</td>\n",
       "      <td>Bipolaris oryzae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5851</th>\n",
       "      <td>62375b9c-54db-4097-95af-4bd20c723822</td>\n",
       "      <td>barcode05</td>\n",
       "      <td>33069</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Pseudomonadales</td>\n",
       "      <td>Pseudomonadaceae</td>\n",
       "      <td>Pseudomonas</td>\n",
       "      <td>Pseudomonas viridiflava</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5852</th>\n",
       "      <td>7f21fa02-74cf-4d6e-896d-76ebc4773042</td>\n",
       "      <td>barcode05</td>\n",
       "      <td>1785145</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Pseudomonadales</td>\n",
       "      <td>Pseudomonadaceae</td>\n",
       "      <td>Pseudomonas</td>\n",
       "      <td>Pseudomonas sp. MS586</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5853 rows Ã— 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   read_id barcode_arrangement   staxids_nt  \\\n",
       "0     807a8997-b021-4820-b33a-1b9095249dc7           barcode01       200451   \n",
       "1     c53fa9c8-f4d5-4876-8526-e04fcea0ac5d           barcode01       200451   \n",
       "2     19a62259-0925-41ef-94e4-6fe9ef776c81           barcode01       199201   \n",
       "3     7e4e83ba-26fa-4d0b-b714-724976a9c7a7           barcode01       200451   \n",
       "4     c92e3d57-c554-4d73-9e17-a51bc14aa4d4           barcode01  1.35728e+06   \n",
       "5     072bbc16-20a5-45b8-bcbf-e7239f48ed12           barcode01  1.32493e+06   \n",
       "6     7eb68975-bd44-40da-b0e8-2635fcedfbb0           barcode01       716540   \n",
       "7     f876c268-fe76-4fb7-97e7-e3c6fa2d8c8b           barcode01  1.28236e+06   \n",
       "8     00f1b4d2-cc23-4231-9b51-543873ddddd2           barcode01       634500   \n",
       "9     74885d13-7285-4de6-ae85-89a8927c36ab           barcode01  1.50567e+06   \n",
       "10    fe662e8d-bb9d-45d5-b699-7891406c5df4           barcode01  1.35728e+06   \n",
       "11    d5533aed-db93-4957-928a-60cd6caa2d0f           barcode01       199201   \n",
       "12    2b27f02a-bd3c-4c0d-a4fb-7e1575c76fe0           barcode01       889211   \n",
       "13    d6a81cd2-3423-4242-9cb3-84f81be34892           barcode01       199201   \n",
       "14    2828d03d-7417-46c5-8382-792e21003d99           barcode01       199201   \n",
       "15    f7760040-5944-4aec-9d3f-b98389bda403           barcode01       587753   \n",
       "16    fa63e5c1-71b2-42e9-906d-879ad1ad4f3b           barcode01  1.81294e+06   \n",
       "17    be038069-e219-48b3-8352-a70beaa767e0           barcode01  1.32493e+06   \n",
       "18    f31b2c3c-87ba-4957-a72c-dd8126a70476           barcode01       199201   \n",
       "19    52767e8d-6540-439b-a930-bb5f8ece7bb8           barcode01  1.32493e+06   \n",
       "20    b89419aa-7906-4bdb-ad6d-06bd26e65944           barcode01        35703   \n",
       "21    6047063f-eb37-4ef4-a9b0-304ced98946e           barcode01       418459   \n",
       "22    af09a516-1d55-4685-8bf6-dba8d13eb50e           barcode01       205918   \n",
       "23    b0ded9cf-3b68-4dbb-b12c-c7e8b18dd35e           barcode01  1.32493e+06   \n",
       "24    8e937eda-6ba6-4f1c-b5ea-fd9f3690ac2c           barcode01       465817   \n",
       "25    ec446ef2-0236-4dc3-a8db-29dcf2a4d5f8           barcode01       199201   \n",
       "26    502dab27-225d-4462-ab53-5572a4ff7d46           barcode01  1.28236e+06   \n",
       "27    0d733956-7720-40c7-9ab0-19842a016005           barcode01       199201   \n",
       "28    c77016d6-1b7f-49f7-8c4b-6c7ca171b291           barcode01  1.32493e+06   \n",
       "29    bbdb95b5-2066-44d1-97e0-a49fcb9ce0a0           barcode01       199201   \n",
       "...                                    ...                 ...          ...   \n",
       "5823  f0a73265-3eaf-41e1-9ea7-5b810f177025           barcode05       930090   \n",
       "5824  036f16cc-2e7a-4414-94a2-55b80e502345           barcode05  1.46066e+06   \n",
       "5825  3febc7c5-f308-4f64-9b37-bc2357608d43           barcode05          317   \n",
       "5826  2de9f0c0-1338-4978-9f88-f34336e3c16d           barcode05       341663   \n",
       "5827  1c4cf73a-df25-4eb8-a96d-dfa117243cb0           barcode05       398673   \n",
       "5828  f0f4e05a-7bf0-4eeb-9a0a-5b84c6590bc2           barcode05        76761   \n",
       "5829  b109ccf5-048d-45e6-a0c1-e568d4498b66           barcode05   1.6036e+06   \n",
       "5830  00cb05e4-9dcc-4562-97b6-1d34db28b2ce           barcode05       169297   \n",
       "5831  0be8c7f8-9da2-4d6c-84bc-4852d29a23ff           barcode05         5599   \n",
       "5832  4af043da-d8a4-48d0-bb61-9130131e08b7           barcode05          562   \n",
       "5833  4e1bc3af-c37e-44be-82e4-765000b17887           barcode05        45151   \n",
       "5834  54b4c516-a392-4c67-bc3f-20d9cb08dfd1           barcode05  1.32493e+06   \n",
       "5835  5dfeeb74-f577-4706-bc10-4173ea8c13d4           barcode05       671987   \n",
       "5836  248cff4b-734e-402d-95ba-aeda3dfd5a78           barcode05        29760   \n",
       "5837  434bf795-fa15-4709-872e-fea2f0bf28fa           barcode05         4565   \n",
       "5838  0e196992-8736-4a23-a766-b6a48ac53f92           barcode05         7165   \n",
       "5839  3422c128-320c-40a4-8f54-8df22e42117e           barcode05        51953   \n",
       "5840  afe0dea5-5835-41ad-a3d5-0111885fadb4           barcode05          562   \n",
       "5841  74af4e06-3343-446c-8d5d-4b0904a192fe           barcode05        39946   \n",
       "5842  b734a7f7-1c8f-4dcf-a175-702a74f07473           barcode05         5599   \n",
       "5843  7e83c505-957c-43a7-bd2f-8921b8bacede           barcode05       930090   \n",
       "5844  9607245c-061a-4f58-8d18-a93d2b3e8cdc           barcode05         7955   \n",
       "5845  f6d6b97e-9204-41ef-9e43-2126f7dd8557           barcode05         7955   \n",
       "5846  a9dab824-a355-4c7d-bad8-9f46ad5db179           barcode05       353152   \n",
       "5847  032c2734-9254-49df-b949-f21606858835           barcode05       741155   \n",
       "5848  59353f77-87ee-40a0-930f-328719e2bfe4           barcode05         5599   \n",
       "5849  86bf9dba-4728-463d-b767-231f0863ebea           barcode05         5599   \n",
       "5850  88013815-813d-4f85-833c-2d5d0f647a94           barcode05       930090   \n",
       "5851  62375b9c-54db-4097-95af-4bd20c723822           barcode05        33069   \n",
       "5852  7f21fa02-74cf-4d6e-896d-76ebc4773042           barcode05      1785145   \n",
       "\n",
       "     superkingdom          phylum                class             order  \\\n",
       "0        Bacteria  Proteobacteria  Gammaproteobacteria   Pseudomonadales   \n",
       "1        Bacteria  Proteobacteria  Gammaproteobacteria   Pseudomonadales   \n",
       "2        Bacteria  Proteobacteria  Gammaproteobacteria   Pseudomonadales   \n",
       "3        Bacteria  Proteobacteria  Gammaproteobacteria   Pseudomonadales   \n",
       "4        Bacteria  Proteobacteria  Gammaproteobacteria   Pseudomonadales   \n",
       "5        Bacteria  Proteobacteria  Gammaproteobacteria   Pseudomonadales   \n",
       "6        Bacteria  Proteobacteria  Gammaproteobacteria  Enterobacterales   \n",
       "7        Bacteria  Proteobacteria  Gammaproteobacteria   Pseudomonadales   \n",
       "8        Bacteria  Proteobacteria  Gammaproteobacteria  Enterobacterales   \n",
       "9       Eukaryota   Basidiomycota      Pucciniomycetes       Pucciniales   \n",
       "10       Bacteria  Proteobacteria  Gammaproteobacteria   Pseudomonadales   \n",
       "11       Bacteria  Proteobacteria  Gammaproteobacteria   Pseudomonadales   \n",
       "12       Bacteria  Proteobacteria  Gammaproteobacteria  Enterobacterales   \n",
       "13       Bacteria  Proteobacteria  Gammaproteobacteria   Pseudomonadales   \n",
       "14       Bacteria  Proteobacteria  Gammaproteobacteria   Pseudomonadales   \n",
       "15       Bacteria  Proteobacteria  Gammaproteobacteria   Pseudomonadales   \n",
       "16       Bacteria  Proteobacteria  Gammaproteobacteria  Enterobacterales   \n",
       "17       Bacteria  Proteobacteria  Gammaproteobacteria   Pseudomonadales   \n",
       "18       Bacteria  Proteobacteria  Gammaproteobacteria   Pseudomonadales   \n",
       "19       Bacteria  Proteobacteria  Gammaproteobacteria   Pseudomonadales   \n",
       "20       Bacteria  Proteobacteria  Gammaproteobacteria  Enterobacterales   \n",
       "21      Eukaryota   Basidiomycota      Pucciniomycetes       Pucciniales   \n",
       "22       Bacteria  Proteobacteria  Gammaproteobacteria   Pseudomonadales   \n",
       "23       Bacteria  Proteobacteria  Gammaproteobacteria   Pseudomonadales   \n",
       "24       Bacteria  Proteobacteria  Gammaproteobacteria  Enterobacterales   \n",
       "25       Bacteria  Proteobacteria  Gammaproteobacteria   Pseudomonadales   \n",
       "26       Bacteria  Proteobacteria  Gammaproteobacteria   Pseudomonadales   \n",
       "27       Bacteria  Proteobacteria  Gammaproteobacteria   Pseudomonadales   \n",
       "28       Bacteria  Proteobacteria  Gammaproteobacteria   Pseudomonadales   \n",
       "29       Bacteria  Proteobacteria  Gammaproteobacteria   Pseudomonadales   \n",
       "...           ...             ...                  ...               ...   \n",
       "5823    Eukaryota      Ascomycota      Dothideomycetes      Pleosporales   \n",
       "5824    Eukaryota      Ascomycota      Dothideomycetes      Pleosporales   \n",
       "5825     Bacteria  Proteobacteria  Gammaproteobacteria   Pseudomonadales   \n",
       "5826    Eukaryota      Ascomycota       Eurotiomycetes        Eurotiales   \n",
       "5827    Eukaryota      Ascomycota      Sordariomycetes       Hypocreales   \n",
       "5828     Bacteria  Proteobacteria  Gammaproteobacteria   Pseudomonadales   \n",
       "5829    Eukaryota      Arthropoda              Insecta         Hemiptera   \n",
       "5830    Eukaryota    Streptophyta           Liliopsida            Poales   \n",
       "5831    Eukaryota      Ascomycota      Dothideomycetes      Pleosporales   \n",
       "5832     Bacteria  Proteobacteria  Gammaproteobacteria  Enterobacterales   \n",
       "5833    Eukaryota      Ascomycota      Dothideomycetes      Pleosporales   \n",
       "5834     Bacteria  Proteobacteria  Gammaproteobacteria   Pseudomonadales   \n",
       "5835    Eukaryota      Ascomycota      Dothideomycetes      Pleosporales   \n",
       "5836    Eukaryota    Streptophyta         Unclassified           Vitales   \n",
       "5837    Eukaryota    Streptophyta           Liliopsida            Poales   \n",
       "5838    Eukaryota      Arthropoda              Insecta           Diptera   \n",
       "5839    Eukaryota    Streptophyta           Liliopsida          Arecales   \n",
       "5840     Bacteria  Proteobacteria  Gammaproteobacteria  Enterobacterales   \n",
       "5841    Eukaryota    Streptophyta           Liliopsida            Poales   \n",
       "5842    Eukaryota      Ascomycota      Dothideomycetes      Pleosporales   \n",
       "5843    Eukaryota      Ascomycota      Dothideomycetes      Pleosporales   \n",
       "5844    Eukaryota        Chordata          Actinopteri     Cypriniformes   \n",
       "5845    Eukaryota        Chordata          Actinopteri     Cypriniformes   \n",
       "5846    Eukaryota     Apicomplexa             Coccidia    Eucoccidiorida   \n",
       "5847     Bacteria  Proteobacteria  Gammaproteobacteria   Pseudomonadales   \n",
       "5848    Eukaryota      Ascomycota      Dothideomycetes      Pleosporales   \n",
       "5849    Eukaryota      Ascomycota      Dothideomycetes      Pleosporales   \n",
       "5850    Eukaryota      Ascomycota      Dothideomycetes      Pleosporales   \n",
       "5851     Bacteria  Proteobacteria  Gammaproteobacteria   Pseudomonadales   \n",
       "5852     Bacteria  Proteobacteria  Gammaproteobacteria   Pseudomonadales   \n",
       "\n",
       "                  family              genus  \\\n",
       "0       Pseudomonadaceae        Pseudomonas   \n",
       "1       Pseudomonadaceae        Pseudomonas   \n",
       "2       Pseudomonadaceae        Pseudomonas   \n",
       "3       Pseudomonadaceae        Pseudomonas   \n",
       "4       Pseudomonadaceae        Pseudomonas   \n",
       "5       Pseudomonadaceae        Pseudomonas   \n",
       "6            Erwiniaceae            Erwinia   \n",
       "7       Pseudomonadaceae        Pseudomonas   \n",
       "8            Erwiniaceae            Erwinia   \n",
       "9           Pucciniaceae           Puccinia   \n",
       "10      Pseudomonadaceae        Pseudomonas   \n",
       "11      Pseudomonadaceae        Pseudomonas   \n",
       "12           Erwiniaceae            Erwinia   \n",
       "13      Pseudomonadaceae        Pseudomonas   \n",
       "14      Pseudomonadaceae        Pseudomonas   \n",
       "15      Pseudomonadaceae        Pseudomonas   \n",
       "16    Enterobacteriaceae       Enterobacter   \n",
       "17      Pseudomonadaceae        Pseudomonas   \n",
       "18      Pseudomonadaceae        Pseudomonas   \n",
       "19      Pseudomonadaceae        Pseudomonas   \n",
       "20    Enterobacteriaceae        Citrobacter   \n",
       "21          Pucciniaceae           Puccinia   \n",
       "22      Pseudomonadaceae        Pseudomonas   \n",
       "23      Pseudomonadaceae        Pseudomonas   \n",
       "24           Erwiniaceae            Erwinia   \n",
       "25      Pseudomonadaceae        Pseudomonas   \n",
       "26      Pseudomonadaceae        Pseudomonas   \n",
       "27      Pseudomonadaceae        Pseudomonas   \n",
       "28      Pseudomonadaceae        Pseudomonas   \n",
       "29      Pseudomonadaceae        Pseudomonas   \n",
       "...                  ...                ...   \n",
       "5823       Pleosporaceae          Bipolaris   \n",
       "5824  Didymosphaeriaceae  Paraphaeosphaeria   \n",
       "5825    Pseudomonadaceae        Pseudomonas   \n",
       "5826      Aspergillaceae        Aspergillus   \n",
       "5827        Hypocreaceae        Trichoderma   \n",
       "5828    Pseudomonadaceae        Pseudomonas   \n",
       "5829        Pentatomidae           Cressona   \n",
       "5830             Poaceae           Aegilops   \n",
       "5831       Pleosporaceae         Alternaria   \n",
       "5832  Enterobacteriaceae        Escherichia   \n",
       "5833       Pleosporaceae        Pyrenophora   \n",
       "5834    Pseudomonadaceae        Pseudomonas   \n",
       "5835       Pleosporaceae       Setosphaeria   \n",
       "5836            Vitaceae              Vitis   \n",
       "5837             Poaceae           Triticum   \n",
       "5838           Culicidae          Anopheles   \n",
       "5839           Arecaceae             Elaeis   \n",
       "5840  Enterobacteriaceae        Escherichia   \n",
       "5841             Poaceae              Oryza   \n",
       "5842       Pleosporaceae         Alternaria   \n",
       "5843       Pleosporaceae          Bipolaris   \n",
       "5844          Cyprinidae              Danio   \n",
       "5845          Cyprinidae              Danio   \n",
       "5846   Cryptosporidiidae    Cryptosporidium   \n",
       "5847    Pseudomonadaceae        Pseudomonas   \n",
       "5848       Pleosporaceae         Alternaria   \n",
       "5849       Pleosporaceae         Alternaria   \n",
       "5850       Pleosporaceae          Bipolaris   \n",
       "5851    Pseudomonadaceae        Pseudomonas   \n",
       "5852    Pseudomonadaceae        Pseudomonas   \n",
       "\n",
       "                                                species  \n",
       "0                                      Pseudomonas poae  \n",
       "1                                      Pseudomonas poae  \n",
       "2                                  Pseudomonas syringae  \n",
       "3                                      Pseudomonas poae  \n",
       "4                                  Pseudomonas syringae  \n",
       "5                                  Pseudomonas syringae  \n",
       "6                                     Erwinia amylovora  \n",
       "7                                      Pseudomonas poae  \n",
       "8                                    Erwinia billingiae  \n",
       "9                           Puccinia cf. psidii AE-2014  \n",
       "10                                 Pseudomonas syringae  \n",
       "11                                 Pseudomonas syringae  \n",
       "12                                    Erwinia amylovora  \n",
       "13                                 Pseudomonas syringae  \n",
       "14                                 Pseudomonas syringae  \n",
       "15                             Pseudomonas chlororaphis  \n",
       "16    Enterobacter cloacae complex 'Hoffmann cluster...  \n",
       "17                                 Pseudomonas syringae  \n",
       "18                                 Pseudomonas syringae  \n",
       "19                                 Pseudomonas syringae  \n",
       "20                             Citrobacter amalonaticus  \n",
       "21                                    Puccinia graminis  \n",
       "22                                 Pseudomonas syringae  \n",
       "23                                 Pseudomonas syringae  \n",
       "24                                 Erwinia tasmaniensis  \n",
       "25                                 Pseudomonas syringae  \n",
       "26                                     Pseudomonas poae  \n",
       "27                                 Pseudomonas syringae  \n",
       "28                                 Pseudomonas syringae  \n",
       "29                                 Pseudomonas syringae  \n",
       "...                                                 ...  \n",
       "5823                                   Bipolaris oryzae  \n",
       "5824                        Paraphaeosphaeria sporulosa  \n",
       "5825                               Pseudomonas syringae  \n",
       "5826                                Aspergillus terreus  \n",
       "5827                                 Trichoderma gamsii  \n",
       "5828                                Pseudomonas veronii  \n",
       "5829                                      Cressona rufa  \n",
       "5830                                  Aegilops tauschii  \n",
       "5831                               Alternaria alternata  \n",
       "5832                                   Escherichia coli  \n",
       "5833                       Pyrenophora tritici-repentis  \n",
       "5834                               Pseudomonas syringae  \n",
       "5835                               Setosphaeria turcica  \n",
       "5836                                     Vitis vinifera  \n",
       "5837                                  Triticum aestivum  \n",
       "5838                                  Anopheles gambiae  \n",
       "5839                                  Elaeis guineensis  \n",
       "5840                                   Escherichia coli  \n",
       "5841                                       Oryza sativa  \n",
       "5842                               Alternaria alternata  \n",
       "5843                                   Bipolaris oryzae  \n",
       "5844                                        Danio rerio  \n",
       "5845                                        Danio rerio  \n",
       "5846                             Cryptosporidium parvum  \n",
       "5847                            Pseudomonas alcaliphila  \n",
       "5848                               Alternaria alternata  \n",
       "5849                               Alternaria alternata  \n",
       "5850                                   Bipolaris oryzae  \n",
       "5851                            Pseudomonas viridiflava  \n",
       "5852                              Pseudomonas sp. MS586  \n",
       "\n",
       "[5853 rows x 10 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ntblasthit_reads_filtered_barcodes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# add the flowcell column just in case\n",
    "ntblasthit_reads_filtered_barcodes = ntblasthit_reads_filtered_barcodes.assign(Flowcell = column_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "ntblasthit_reads_filtered_barcodes.to_csv(r'/home/yiheng/analysis/WGS/%s_taxa.tab' % column_name, header=column_name, index=None, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Defines a function to make manual adjustments to the NCBI taxonomic sorting\n",
    "def manually_adjust_ncbi_taxonomy(rank_select, entry_select, ranks_to_edit, new_entry):\n",
    "    \"\"\"Select dataframe entries to edit by a specific rank (rank_select) \n",
    "    and entry related to that rank (entry_select), \n",
    "    provide a list of ranks to edit for those selected entries (ranks_to_edit),\n",
    "    replace these ranks with a new entry (new_entry)\"\"\"\n",
    "    if entry_select in list(ntblasthit_reads_filtered_barcodes[rank_select].unique()):\n",
    "        selection = (ntblasthit_reads_filtered_barcodes[rank_select] == entry_select)\n",
    "        selected_tax_ranks = ranks_to_edit\n",
    "        for ranks in selected_tax_ranks:\n",
    "            ntblasthit_reads_filtered_barcodes[ranks][selection] = new_entry"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "manually_adjust_ncbi_taxonomy('species', \n",
    "                            'Cloning vector lambda TXF97', \n",
    "                            ['superkingdom', 'phylum', 'class', 'order', 'family', 'genus'], \n",
    "                            'Cloning vector lambda TXF97')\n",
    "\n",
    "manually_adjust_ncbi_taxonomy('superkingdom', \n",
    "                            'Viruses', \n",
    "                            ['phylum', 'class'], \n",
    "                            'dsDNA viruses, no RNA stage')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#count pivot table of dataframe with taxonomic columns: \n",
    "#rows - rank names, columns - barcode, sorted by desired class\n",
    "def generate_ncbi_taxonomy_pivot(tax_df, rank, bcs, num):\n",
    "    \"\"\"From tax_df, generate a pivot table listing num rank counts, sorted by bcs\"\"\"\n",
    "    pivot_table = tax_df.pivot_table(values='order', \n",
    "                                            index=rank, \n",
    "                                            columns='barcode_arrangement', \n",
    "                                            aggfunc='count', \n",
    "                                            fill_value=0, \n",
    "                                            margins=True)\n",
    "    pivot_table.columns.name = None\n",
    "    return pivot_table.sort_values(bcs, axis=0, ascending=False).head(n=num)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>barcode01</th>\n",
       "      <th>barcode02</th>\n",
       "      <th>barcode03</th>\n",
       "      <th>barcode04</th>\n",
       "      <th>barcode05</th>\n",
       "      <th>All</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>family</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>All</th>\n",
       "      <td>1942</td>\n",
       "      <td>1855</td>\n",
       "      <td>37</td>\n",
       "      <td>1748</td>\n",
       "      <td>271</td>\n",
       "      <td>5853</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pseudomonadaceae</th>\n",
       "      <td>1470</td>\n",
       "      <td>814</td>\n",
       "      <td>0</td>\n",
       "      <td>1104</td>\n",
       "      <td>58</td>\n",
       "      <td>3446</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pleosporaceae</th>\n",
       "      <td>44</td>\n",
       "      <td>263</td>\n",
       "      <td>1</td>\n",
       "      <td>160</td>\n",
       "      <td>77</td>\n",
       "      <td>545</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sphingomonadaceae</th>\n",
       "      <td>1</td>\n",
       "      <td>161</td>\n",
       "      <td>0</td>\n",
       "      <td>83</td>\n",
       "      <td>1</td>\n",
       "      <td>246</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mycosphaerellaceae</th>\n",
       "      <td>1</td>\n",
       "      <td>103</td>\n",
       "      <td>0</td>\n",
       "      <td>43</td>\n",
       "      <td>2</td>\n",
       "      <td>149</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Hymenobacteraceae</th>\n",
       "      <td>1</td>\n",
       "      <td>82</td>\n",
       "      <td>0</td>\n",
       "      <td>39</td>\n",
       "      <td>0</td>\n",
       "      <td>122</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Enterobacteriaceae</th>\n",
       "      <td>51</td>\n",
       "      <td>56</td>\n",
       "      <td>10</td>\n",
       "      <td>35</td>\n",
       "      <td>5</td>\n",
       "      <td>157</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Microbacteriaceae</th>\n",
       "      <td>3</td>\n",
       "      <td>37</td>\n",
       "      <td>0</td>\n",
       "      <td>31</td>\n",
       "      <td>0</td>\n",
       "      <td>71</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Poaceae</th>\n",
       "      <td>14</td>\n",
       "      <td>8</td>\n",
       "      <td>13</td>\n",
       "      <td>29</td>\n",
       "      <td>41</td>\n",
       "      <td>105</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Leptosphaeriaceae</th>\n",
       "      <td>7</td>\n",
       "      <td>34</td>\n",
       "      <td>1</td>\n",
       "      <td>18</td>\n",
       "      <td>17</td>\n",
       "      <td>77</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Oxalobacteraceae</th>\n",
       "      <td>16</td>\n",
       "      <td>18</td>\n",
       "      <td>0</td>\n",
       "      <td>15</td>\n",
       "      <td>0</td>\n",
       "      <td>49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Erwiniaceae</th>\n",
       "      <td>165</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>185</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Unclassified</th>\n",
       "      <td>6</td>\n",
       "      <td>20</td>\n",
       "      <td>0</td>\n",
       "      <td>10</td>\n",
       "      <td>1</td>\n",
       "      <td>37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cryptococcaceae</th>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>8</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Burkholderiaceae</th>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pucciniaceae</th>\n",
       "      <td>56</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>62</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Nocardiaceae</th>\n",
       "      <td>2</td>\n",
       "      <td>33</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sporidiobolaceae</th>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Paenibacillaceae</th>\n",
       "      <td>1</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sphingobacteriaceae</th>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     barcode01  barcode02  barcode03  barcode04  barcode05  \\\n",
       "family                                                                       \n",
       "All                       1942       1855         37       1748        271   \n",
       "Pseudomonadaceae          1470        814          0       1104         58   \n",
       "Pleosporaceae               44        263          1        160         77   \n",
       "Sphingomonadaceae            1        161          0         83          1   \n",
       "Mycosphaerellaceae           1        103          0         43          2   \n",
       "Hymenobacteraceae            1         82          0         39          0   \n",
       "Enterobacteriaceae          51         56         10         35          5   \n",
       "Microbacteriaceae            3         37          0         31          0   \n",
       "Poaceae                     14          8         13         29         41   \n",
       "Leptosphaeriaceae            7         34          1         18         17   \n",
       "Oxalobacteraceae            16         18          0         15          0   \n",
       "Erwiniaceae                165          7          0         13          0   \n",
       "Unclassified                 6         20          0         10          1   \n",
       "Cryptococcaceae              2          6          0          7          8   \n",
       "Burkholderiaceae             2          5          0          7          1   \n",
       "Pucciniaceae                56          0          0          6          0   \n",
       "Nocardiaceae                 2         33          0          6          0   \n",
       "Sporidiobolaceae             2          1          0          5          3   \n",
       "Paenibacillaceae             1          7          0          5          0   \n",
       "Sphingobacteriaceae          0          2          0          5          0   \n",
       "\n",
       "                      All  \n",
       "family                     \n",
       "All                  5853  \n",
       "Pseudomonadaceae     3446  \n",
       "Pleosporaceae         545  \n",
       "Sphingomonadaceae     246  \n",
       "Mycosphaerellaceae    149  \n",
       "Hymenobacteraceae     122  \n",
       "Enterobacteriaceae    157  \n",
       "Microbacteriaceae      71  \n",
       "Poaceae               105  \n",
       "Leptosphaeriaceae      77  \n",
       "Oxalobacteraceae       49  \n",
       "Erwiniaceae           185  \n",
       "Unclassified           37  \n",
       "Cryptococcaceae        23  \n",
       "Burkholderiaceae       15  \n",
       "Pucciniaceae           62  \n",
       "Nocardiaceae           41  \n",
       "Sporidiobolaceae       11  \n",
       "Paenibacillaceae       13  \n",
       "Sphingobacteriaceae     7  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "generate_ncbi_taxonomy_pivot(ntblasthit_reads_filtered_barcodes, 'family','barcode04',20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# % pivot table of dataframe with taxonomic columns: \n",
    "#rows - rank names, columns - barcode, sorted by desired class\n",
    "def generate_ncbi_taxonomy_pivot_pc(tax_df, rank, bcs, num):\n",
    "    \"\"\"From tax_df, generate a pivot table listing num rank %, sorted ascending by bcs\"\"\"\n",
    "    pivot_table_pc = tax_df.pivot_table(values='order', \n",
    "                                            index=rank, \n",
    "                                            columns='barcode_arrangement', \n",
    "                                            aggfunc='count', \n",
    "                                            fill_value=0, \n",
    "                                            margins=True)\n",
    "    pivot_table_pc.columns.name = None\n",
    "    \n",
    "    for brcs in pivot_table_pc.columns:\n",
    "        pivot_table_pc[str(brcs) + '%'] = pivot_table_pc[brcs].apply(lambda x: (x / pivot_table_pc[brcs].loc['All'])*100)\n",
    "    for brcs in pivot_table_pc.columns:\n",
    "        if '%' not in brcs:\n",
    "            pivot_table_pc.drop(brcs, axis=1, inplace=True)\n",
    "    pivot_table_pc.drop('All', axis=0, inplace=True)\n",
    "\n",
    "    pivot_table_pc2 = pivot_table_pc.sort_values(str(bcs) + '%', axis=0, ascending=False)\n",
    "    return pivot_table_pc2.applymap(lambda x: '%.3f' % x).head(n=num)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>barcode01%</th>\n",
       "      <th>barcode02%</th>\n",
       "      <th>barcode03%</th>\n",
       "      <th>barcode04%</th>\n",
       "      <th>barcode05%</th>\n",
       "      <th>All%</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>genus</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Pseudomonas</th>\n",
       "      <td>75.695</td>\n",
       "      <td>43.881</td>\n",
       "      <td>0.000</td>\n",
       "      <td>63.158</td>\n",
       "      <td>21.402</td>\n",
       "      <td>58.876</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Alternaria</th>\n",
       "      <td>1.442</td>\n",
       "      <td>11.321</td>\n",
       "      <td>2.703</td>\n",
       "      <td>7.037</td>\n",
       "      <td>20.295</td>\n",
       "      <td>7.125</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sphingomonas</th>\n",
       "      <td>0.051</td>\n",
       "      <td>7.925</td>\n",
       "      <td>0.000</td>\n",
       "      <td>4.291</td>\n",
       "      <td>0.369</td>\n",
       "      <td>3.827</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Zymoseptoria</th>\n",
       "      <td>0.051</td>\n",
       "      <td>5.337</td>\n",
       "      <td>0.000</td>\n",
       "      <td>2.346</td>\n",
       "      <td>0.000</td>\n",
       "      <td>2.409</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Hymenobacter</th>\n",
       "      <td>0.051</td>\n",
       "      <td>4.420</td>\n",
       "      <td>0.000</td>\n",
       "      <td>2.231</td>\n",
       "      <td>0.000</td>\n",
       "      <td>2.084</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Escherichia</th>\n",
       "      <td>0.566</td>\n",
       "      <td>2.426</td>\n",
       "      <td>24.324</td>\n",
       "      <td>1.716</td>\n",
       "      <td>1.476</td>\n",
       "      <td>1.691</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bipolaris</th>\n",
       "      <td>0.412</td>\n",
       "      <td>1.456</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1.201</td>\n",
       "      <td>4.059</td>\n",
       "      <td>1.145</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Curtobacterium</th>\n",
       "      <td>0.051</td>\n",
       "      <td>0.755</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1.144</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.598</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Leptosphaeria</th>\n",
       "      <td>0.360</td>\n",
       "      <td>1.833</td>\n",
       "      <td>2.703</td>\n",
       "      <td>1.030</td>\n",
       "      <td>6.273</td>\n",
       "      <td>1.316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Triticum</th>\n",
       "      <td>0.309</td>\n",
       "      <td>0.108</td>\n",
       "      <td>29.730</td>\n",
       "      <td>0.973</td>\n",
       "      <td>14.022</td>\n",
       "      <td>1.264</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Unclassified</th>\n",
       "      <td>0.463</td>\n",
       "      <td>0.485</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.858</td>\n",
       "      <td>0.369</td>\n",
       "      <td>0.581</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pantoea</th>\n",
       "      <td>1.493</td>\n",
       "      <td>0.216</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.744</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.786</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Setosphaeria</th>\n",
       "      <td>0.206</td>\n",
       "      <td>0.431</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.629</td>\n",
       "      <td>1.107</td>\n",
       "      <td>0.444</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Massilia</th>\n",
       "      <td>0.360</td>\n",
       "      <td>0.431</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.458</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.393</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Frondihabitans</th>\n",
       "      <td>0.000</td>\n",
       "      <td>0.270</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.400</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.205</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Aegilops</th>\n",
       "      <td>0.154</td>\n",
       "      <td>0.108</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.400</td>\n",
       "      <td>0.738</td>\n",
       "      <td>0.239</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Puccinia</th>\n",
       "      <td>2.781</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.343</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1.025</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhodococcus</th>\n",
       "      <td>0.103</td>\n",
       "      <td>1.725</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.343</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.683</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Janthinobacterium</th>\n",
       "      <td>0.309</td>\n",
       "      <td>0.377</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.343</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.325</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhodotorula</th>\n",
       "      <td>0.103</td>\n",
       "      <td>0.054</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.286</td>\n",
       "      <td>1.107</td>\n",
       "      <td>0.188</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  barcode01% barcode02% barcode03% barcode04% barcode05%  \\\n",
       "genus                                                                      \n",
       "Pseudomonas           75.695     43.881      0.000     63.158     21.402   \n",
       "Alternaria             1.442     11.321      2.703      7.037     20.295   \n",
       "Sphingomonas           0.051      7.925      0.000      4.291      0.369   \n",
       "Zymoseptoria           0.051      5.337      0.000      2.346      0.000   \n",
       "Hymenobacter           0.051      4.420      0.000      2.231      0.000   \n",
       "Escherichia            0.566      2.426     24.324      1.716      1.476   \n",
       "Bipolaris              0.412      1.456      0.000      1.201      4.059   \n",
       "Curtobacterium         0.051      0.755      0.000      1.144      0.000   \n",
       "Leptosphaeria          0.360      1.833      2.703      1.030      6.273   \n",
       "Triticum               0.309      0.108     29.730      0.973     14.022   \n",
       "Unclassified           0.463      0.485      0.000      0.858      0.369   \n",
       "Pantoea                1.493      0.216      0.000      0.744      0.000   \n",
       "Setosphaeria           0.206      0.431      0.000      0.629      1.107   \n",
       "Massilia               0.360      0.431      0.000      0.458      0.000   \n",
       "Frondihabitans         0.000      0.270      0.000      0.400      0.000   \n",
       "Aegilops               0.154      0.108      0.000      0.400      0.738   \n",
       "Puccinia               2.781      0.000      0.000      0.343      0.000   \n",
       "Rhodococcus            0.103      1.725      0.000      0.343      0.000   \n",
       "Janthinobacterium      0.309      0.377      0.000      0.343      0.000   \n",
       "Rhodotorula            0.103      0.054      0.000      0.286      1.107   \n",
       "\n",
       "                     All%  \n",
       "genus                      \n",
       "Pseudomonas        58.876  \n",
       "Alternaria          7.125  \n",
       "Sphingomonas        3.827  \n",
       "Zymoseptoria        2.409  \n",
       "Hymenobacter        2.084  \n",
       "Escherichia         1.691  \n",
       "Bipolaris           1.145  \n",
       "Curtobacterium      0.598  \n",
       "Leptosphaeria       1.316  \n",
       "Triticum            1.264  \n",
       "Unclassified        0.581  \n",
       "Pantoea             0.786  \n",
       "Setosphaeria        0.444  \n",
       "Massilia            0.393  \n",
       "Frondihabitans      0.205  \n",
       "Aegilops            0.239  \n",
       "Puccinia            1.025  \n",
       "Rhodococcus         0.683  \n",
       "Janthinobacterium   0.325  \n",
       "Rhodotorula         0.188  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "generate_ncbi_taxonomy_pivot_pc(ntblasthit_reads_filtered_barcodes, 'genus', 'barcode04', 20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Assesses number of barcodes, generates necessary number of plots (even) and relevant indexing system\n",
    "ncbiblast_barcodes = list(ntblasthit_reads_filtered_barcodes['barcode_arrangement'].unique())\n",
    "no_of_subplots = len(ncbiblast_barcodes)+1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Always 2 columns, bc-count/2 rows \n",
    "no_of_subplots_pair = [int(no_of_subplots/2), 2]\n",
    "\n",
    "#Produce pairs of indices correlating to the coordinates of the subplots\n",
    "subplot_coordinates = list(product(range(no_of_subplots_pair[0]), range(no_of_subplots_pair[1])))\n",
    "subplot_coordinates_list = [list(l) for l in subplot_coordinates]\n",
    "subplot_coordinates_list_rows = [i[0] for i in subplot_coordinates_list]\n",
    "subplot_coordinates_list_columns = [i[1] for i in subplot_coordinates_list]\n",
    "\n",
    "fig, ax = plt.subplots(no_of_subplots_pair[1], no_of_subplots_pair[0], figsize=(20, 6))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "name_list = ['Stripe rust', 'Septoria tritici blotch', 'Healthy wheat', 'Stripe rust and sptoria tritici blotch', 'Yellow spot']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def get_color(x):\n",
    "    x = x%10\n",
    "    return ('#%02X%02X%02X' % (int(plt.cm.tab10.colors[x][0]*255),int(plt.cm.tab10.colors[x][1]*255),int(plt.cm.tab10.colors[x][2]*255)))\n",
    "\n",
    "def applyTaxPlotStyle(ax_ind1, ax_ind2, bcs, rank, name):\n",
    "    labels = ''\n",
    "    labels = tuple(generate_ncbi_taxonomy_pivot_pc(ntblasthit_reads_filtered_barcodes, rank, bcs, 10)[bcs + '%'].index)\n",
    "    y_pos = ''\n",
    "    y_pos = np.arange(len(labels))\n",
    "    sizes = ''\n",
    "    sizes = tuple([float(i) for i in generate_ncbi_taxonomy_pivot_pc(ntblasthit_reads_filtered_barcodes, rank, bcs, 10)[bcs + '%']])\n",
    "    \n",
    "    ax[ax_ind1, ax_ind2].barh(y_pos, sizes, align='center', alpha=0.4, color=get_color(2*ax_ind2 + ax_ind1))\n",
    "    ax[ax_ind1, ax_ind2].set_yticks(y_pos)\n",
    "    ax[ax_ind1, ax_ind2].set_yticklabels(labels)\n",
    "    ax[ax_ind1, ax_ind2].tick_params(top=\"off\", right=\"off\", left=\"off\")\n",
    "    ax[ax_ind1, ax_ind2].set_xlabel('% Total Reads Hitting NCBI')\n",
    "    ax[ax_ind1, ax_ind2].set_xticks(range(0, 110, 10))\n",
    "    ax[ax_ind1, ax_ind2].set_title(name)\n",
    "    ax[ax_ind1, ax_ind2].invert_yaxis()\n",
    "    ax[ax_ind1, ax_ind2].grid(b=None, which='major', axis='x', linestyle='dashed', linewidth=0.2)\n",
    "    ax[ax_ind1, ax_ind2].annotate('E = 0.00001', \n",
    "             xy=(1.06, 0.3), xycoords='axes fraction', fontsize=10, xytext=(-20, -30), textcoords='offset points', \n",
    "             ha='right', va='bottom')\n",
    "\n",
    "    \n",
    "def generateTaxSubplots(search_rank):\n",
    "    for ax_ind1, ax_ind2, bcs, name in zip(subplot_coordinates_list_columns, subplot_coordinates_list_rows, ncbiblast_barcodes, name_list):\n",
    "        applyTaxPlotStyle(ax_ind1, ax_ind2, bcs, search_rank, name)    \n",
    "\n",
    "    if len(list(ntblasthit_reads_filtered_barcodes['barcode_arrangement'].unique())) != 0:\n",
    "        plt.delaxes(ax[subplot_coordinates_list_columns[-1], subplot_coordinates_list_rows[-1]])\n",
    "    \n",
    "    plt.suptitle('Reads Hitting NCBI Database - % By ' + str(search_rank).title() + ' By Sample', \n",
    "             fontsize='x-large',\n",
    "             y=1.02,\n",
    "             fontweight=\"bold\")\n",
    "    \n",
    "    plt.subplots_adjust(left=None, bottom=None, right=None, top=None, wspace=1.4, hspace=0.4)\n",
    "    plt.tight_layout()\n",
    "    \n",
    "    plt.savefig('/home/yiheng/analysis/WGS/figures' + 'figure3_%s_%s.png' % (column_name, search_rank), bbox_inches='tight', dpi=1000)\n",
    "    plt.show()\n",
    "            \n",
    "            \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(no_of_subplots_pair[1], no_of_subplots_pair[0], figsize=(20, 6))\n",
    "generateTaxSubplots('genus')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "df1=pd.DataFrame([[1, 2],[10, 1], [3, 4], [7, 8]], columns=['A', 'B'])\n",
    "df2=pd.DataFrame([[1, 4],[3, 5]], columns=['A', 'B'])\n",
    "# now replace the location use the index of each df selected from value by .loc function\n",
    "# first select the value from df2 column A\n",
    "for values in df2['A'].unique():\n",
    "    # then locate the position in df1 and replace it by the VALUE of the position in df2\n",
    "    df1.loc[df1[df1['A'] == (values)].index, 'B'] = df2.loc[df2[df2['A'] == (values)].index, 'B'].values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "df1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
